<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Convenience functions for bioinformatics • coriell</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Convenience functions for bioinformatics">
<meta property="og:description" content="Functions for plotting and transforming data.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">coriell</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/REdiscoverTE-workflow.html">REdiscoverTE Workflow</a>
    </li>
    <li>
      <a href="articles/RNA-seq-workflow.html">RNA-seq-workflow</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/coriell-research/coriell/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="coriell">coriell<a class="anchor" aria-label="anchor" href="#coriell"></a>
</h1></div>
<p>This package contains helper functions for common bioinformatics tasks (and some not-so-common tasks). If you find yourself reusing old code over and over, let me know and we’ll create a function and put it in this package.</p>
<p>I will be adding tests, however, <strong>please test functions and let me know when things are broken</strong> by either raising an issue on Github or contacting me.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Since this package will be constantly changing be sure to install the latest version from Github using:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Make sure you have devtools installed</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Then install using devtools::install_github</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"coriell-research/coriell"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h2>
<ul>
<li><a href="https://coriell-research.github.io/coriell/#summarize-results-from-differential-expression-test" class="external-link">Summarize results from differential expression analysis</a></li>
<li><a href="https://coriell-research.github.io/coriell/#create-volcano-plot-from-differential-expression-results" class="external-link">Create volcano plot from differential expression results</a></li>
<li><a href="https://coriell-research.github.io/coriell/#create-md-plot-from-differential-expression-results" class="external-link">Create md plot from differential expression results</a></li>
<li><a href="https://coriell-research.github.io/coriell/#heatmap-with-sensible-defaults" class="external-link">Heatmap with sensible defaults</a></li>
<li><a href="https://coriell-research.github.io/coriell/#parallel-coordinates-plot-of-expression-matrix" class="external-link">Parallel coordinates plot of expression matrix</a></li>
<li><a href="https://coriell-research.github.io/coriell/#boxplot-of-expression-matrix" class="external-link">Boxplot of expression matrix</a></li>
<li><a href="https://coriell-research.github.io/coriell/#density-plot-of-expression-matrix" class="external-link">Density plot of expression matrix</a></li>
<li><a href="https://coriell-research.github.io/coriell/#perform-gene-ontology-analysis-with-panther" class="external-link">Perform gene ontology analysis with PANTHER</a></li>
<li><a href="https://coriell-research.github.io/coriell/#perform-correlation-permutation-test-using-multiple-cores" class="external-link">Correlate methylation data with age</a></li>
</ul>
<p>For an overview of how to use some of these functions in a typical analysis check out the <a href="https://coriell-research.github.io/coriell/articles/RNA-seq-workflow.html" class="external-link">RNA-seq vignette</a> (more vignettes to come)</p>
<p>The package also contains many other convenience functions so be sure to check out the <a href="https://coriell-research.github.io/coriell/reference/index.html" class="external-link">reference</a> page as well.</p>
<div class="section level3">
<h3 id="built-in-datasets">Built-in datasets<a class="anchor" aria-label="anchor" href="#built-in-datasets"></a>
</h3>
<p>These datasets are built into the package for testing purposes and are used below to illustrate some functionality.</p>
<ul>
<li>
<code>GSE161650_de</code> : Differential expression results of THZ1 vs DMSO from GSE161650</li>
<li>
<code>GSE161650_lc</code> : Normalized log2 counts from THZ1 vs DMSO replicates from GSE161650</li>
</ul>
<p>The head of <code>GSE161650_de</code> looks like:</p>
<pre><code>&gt;   feature_id     logFC unshrunk.logFC    logCPM       PValue          FDR
&gt; 1        JUN  5.759233       5.759908  9.079350 2.919097e-14 1.990666e-10
&gt; 2       IER5  3.931325       3.931420 10.158336 3.365738e-14 1.990666e-10
&gt; 3    GADD45B  5.813030       5.814071  8.432432 6.435666e-14 2.537583e-10
&gt; 4       IER2  4.457981       4.458016 12.223835 1.528890e-13 4.521309e-10
&gt; 5     PIK3R3 -4.018325      -4.019603  7.124408 2.122484e-13 4.752097e-10
&gt; 6     HEXIM1  4.497345       4.497844  8.275561 2.696985e-13 4.752097e-10</code></pre>
<p>And <code>GSE161650_lc</code>:</p>
<pre><code>&gt;          DMSO.1     DMSO.2     DMSO.3    THZ1.1    THZ1.2     THZ1.3
&gt; A1BG  5.3323512  5.4576081  5.2876011  6.703752 6.8090471  6.7908595
&gt; AAAS  3.8738768  3.8839857  3.5242625  3.768811 4.1406003  3.7454773
&gt; AACS  2.1381539  2.3748462  2.2761971  3.769163 3.4352003  3.4064114
&gt; AADAT 0.8240013 -0.2391457 -0.8699138 -1.342273 0.1967574 -0.8355924
&gt; AAED1 1.2814586  1.5332476  2.0735095  2.443188 2.1270932  0.9707206
&gt; AAGAB 6.8747238  6.7396922  6.6670762  6.756664 6.5840839  6.7662317</code></pre>
<p>See the package documentation <code><a href="reference/GSE161650_de.html">?GSE161650_de</a></code> and <code><a href="reference/GSE161650_lc.html">?GSE161650_lc</a></code> for citation information.</p>
</div>
<div class="section level3">
<h3 id="summarize-results-from-differential-expression-test">Summarize results from differential expression test<a class="anchor" aria-label="anchor" href="#summarize-results-from-differential-expression-test"></a>
</h3>
<p>Return a table of up/down/non-de counts and their percentages.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>de <span class="ot">&lt;-</span> GSE161650_de</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summarize_dge</span>(de, <span class="at">fdr =</span> <span class="fl">0.1</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span>     Direction    N Percent</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">1</span>          Up <span class="dv">2588</span>   <span class="fl">23.52</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">2</span>        Down <span class="dv">3365</span>   <span class="fl">30.59</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">3</span> Unperturbed <span class="dv">5049</span>   <span class="fl">45.89</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-volcano-plot-from-differential-expression-results">Create volcano plot from differential expression results<a class="anchor" aria-label="anchor" href="#create-volcano-plot-from-differential-expression-results"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">de</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/volcano1.png"></p>
<p>Different significance levels can be used to filter the plotted points. For example, significance levels can be set by specifying the <code>fdr</code> and <code>lfc</code> values.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">de</span>, fdr <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/volcano2.png"></p>
<p>Labels for the counts will be displayed by default. To remove them set <code>annotate_counts = FALSE</code></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">de</span>, fdr <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>, annotate_counts <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/volcano3.png"></p>
<p>Text labels can also be added for the DE genes by setting <code>label_sig = TRUE</code>. If <code>label_sig = TRUE</code> then the <code>lab</code> argument must also designate a column containing the names of the items to plot. Additional arguments can be passed to <code><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">ggrepel::geom_text_repel()</a></code> via <code>...</code> if you need to modify the behavior of the labels.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">de</span>, fdr <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>, label_sig <span class="op">=</span> <span class="cn">TRUE</span>, lab <span class="op">=</span> <span class="st">"feature_id"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/volcano4.png"></p>
<p>All aesthetics of the points can also be changed</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">de</span>,</span>
<span>  fdr <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>  lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>  up_color <span class="op">=</span> <span class="st">"firebrick"</span>,</span>
<span>  down_color <span class="op">=</span> <span class="st">"midnightblue"</span>,</span>
<span>  nonde_color <span class="op">=</span> <span class="st">"grey80"</span>,</span>
<span>  up_alpha <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  down_alpha <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  nonde_alpha <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  up_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  down_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  nonde_size <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/volcano5.png"></p>
</div>
<div class="section level3">
<h3 id="create-md-plot-from-differential-expression-results">Create md plot from differential expression results<a class="anchor" aria-label="anchor" href="#create-md-plot-from-differential-expression-results"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">de</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/md1.png"></p>
<p>Different significance levels can be used to filter the plotted points. For example, significance levels can be set by specifying the <code>fdr</code> and <code>lfc</code> values.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">de</span>, fdr <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/md2.png"></p>
<p>Labels for the counts will be displayed by default. To remove them set <code>annotate_counts = FALSE</code></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">de</span>, fdr <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>, annotate_counts <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/md3.png"></p>
<p>Text labels can also be added for the DE genes by setting <code>label_sig = TRUE</code>. If <code>label_sig = TRUE</code> then the <code>lab</code> argument must also designate a column containing the names of the items to plot. Additional arguments can be passed to <code><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">ggrepel::geom_text_repel()</a></code> via <code>...</code> if you need to modify the behavior of the labels.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">de</span>, fdr <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>, label_sig <span class="op">=</span> <span class="cn">TRUE</span>, lab <span class="op">=</span> <span class="st">"feature_id"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/md5.png"></p>
<p>All aesthetics of the points can also be changed</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">de</span>,</span>
<span>  fdr <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>  lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>  up_color <span class="op">=</span> <span class="st">"firebrick"</span>,</span>
<span>  down_color <span class="op">=</span> <span class="st">"midnightblue"</span>,</span>
<span>  nonde_color <span class="op">=</span> <span class="st">"grey80"</span>,</span>
<span>  up_alpha <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  down_alpha <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  nonde_alpha <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  up_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  down_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  nonde_size <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/md4.png"></p>
</div>
<div class="section level3">
<h3 id="heatmap-with-sensible-defaults">Heatmap with sensible defaults<a class="anchor" aria-label="anchor" href="#heatmap-with-sensible-defaults"></a>
</h3>
<p>We often use the same settings when making calls to <code>pheatmap</code>. This function is a wrapper around <code>pheatmap</code> which uses sensible default values for expression data.</p>
<p>Any of these default values can be overridden by simply supplying the arguments to <code>quickmap</code> as you would <code>pheatmap</code>. This also allows for additional arguments to be passed to the <code>quickmap</code> function for creating row and column annotations.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="va">GSE161650_lc</span></span>
<span></span>
<span><span class="co"># plot a heatmap of the logCPM values</span></span>
<span><span class="fu"><a href="reference/quickmap.html">quickmap</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/quickmap1.png"></p>
<p>Other <code>pheatmap</code> arguments can be passed to the <code>quickmap</code> function as well.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create annotation for columns</span></span>
<span><span class="va">col_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Treatment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DMSO"</span>, <span class="st">"THZ1"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">col_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create color scheme for treatment conditions</span></span>
<span><span class="va">ann_colors</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Treatment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DMSO"</span> <span class="op">=</span> <span class="st">"steelblue"</span>, <span class="st">"THZ1"</span> <span class="op">=</span> <span class="st">"firebrick"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the heatmap, passing additional args to pheatmap</span></span>
<span><span class="fu"><a href="reference/quickmap.html">quickmap</a></span><span class="op">(</span></span>
<span>  <span class="va">logcounts</span>,</span>
<span>  annotation_col <span class="op">=</span> <span class="va">col_df</span>,</span>
<span>  annotation_colors <span class="op">=</span> <span class="va">ann_colors</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Treatment vs Control"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/quickmap2.png"></p>
<p>Additional arguments can be set to limit the scales of the heatmap as well as remove low variance features prior to plotting. To remove low variance features set the <code>removeVar</code> argument to the desired proportion of features to drop.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/quickmap.html">quickmap</a></span><span class="op">(</span></span>
<span>  <span class="va">logcounts</span>,</span>
<span>  removeVar <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  annotation_col <span class="op">=</span> <span class="va">col_df</span>,</span>
<span>  annotation_colors <span class="op">=</span> <span class="va">ann_colors</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"THZ1 vs DMSO"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/quickmap3.png"></p>
<p>The colors of the heatmap scale can also be ‘fixed’ above and below a certain threshold with the <code>fix_extreme</code> and <code>thresh</code> arguments</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/quickmap.html">quickmap</a></span><span class="op">(</span></span>
<span>    <span class="va">logcounts</span>,</span>
<span>    removeVar <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>    fix_extreme <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    thresh <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>    annotation_col <span class="op">=</span> <span class="va">col_df</span>,</span>
<span>    annotation_colors <span class="op">=</span> <span class="va">ann_colors</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"THZ1 vs DMSO"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/quickmap4.png"></p>
</div>
<div class="section level3">
<h3 id="parallel-coordinates-plot-of-expression-matrix">Parallel coordinates plot of expression matrix<a class="anchor" aria-label="anchor" href="#parallel-coordinates-plot-of-expression-matrix"></a>
</h3>
<p>Parallel coordinates plots (PCP), as well as boxplots and density plots (below), can be useful tools for examining expression patterns across samples before and after normalization, for example. To create a PCP with <code>coriell</code> use the <code><a href="reference/plot_parallel.html">plot_parallel()</a></code> function.</p>
<p>Using <code>logcounts</code> define above:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create PCP plot -- passing additional alpha value to geom_line()</span></span>
<span><span class="fu"><a href="reference/plot_parallel.html">plot_parallel</a></span><span class="op">(</span><span class="va">logcounts</span>, alpha <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/parallel.png"></p>
<p>If metadata is supplied then the PCP can be colored by the supplied variable name. Using <code>col_df</code> defined above:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_parallel.html">plot_parallel</a></span><span class="op">(</span><span class="va">logcounts</span>, <span class="va">col_df</span>, colBy <span class="op">=</span> <span class="st">"Treatment"</span>, alpha <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/parallel2.png"></p>
</div>
<div class="section level3">
<h3 id="boxplot-of-expression-matrix">Boxplot of expression matrix<a class="anchor" aria-label="anchor" href="#boxplot-of-expression-matrix"></a>
</h3>
<p>Likewise, boxplots of the expression values can be visualized in a similar fashion with <code><a href="reference/plot_boxplot.html">plot_boxplot()</a></code></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_boxplot.html">plot_boxplot</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/boxplot.png"></p>
<p>If metadata are supplied the boxes can be colored by the desired metadata column and additional arguments can be passed to <code><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot()</a></code> to modify the boxplots.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_boxplot.html">plot_boxplot</a></span><span class="op">(</span><span class="va">logcounts</span>, <span class="va">col_df</span>, fillBy <span class="op">=</span> <span class="st">"Treatment"</span>, outlier.shape <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/boxplot2.png"></p>
</div>
<div class="section level3">
<h3 id="density-plot-of-expression-matrix">Density plot of expression matrix<a class="anchor" aria-label="anchor" href="#density-plot-of-expression-matrix"></a>
</h3>
<p>Density plots showing the expression values for samples or groups can be plotted using the <code><a href="reference/plot_density.html">plot_density()</a></code> function.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_density.html">plot_density</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/density.png"></p>
<p>If metadata are supplied, the lines can be colored by the desired metadata column and additional arguments can be passed to <code><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density()</a></code> to modify the density layer.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_density.html">plot_density</a></span><span class="op">(</span><span class="va">logcounts</span>, <span class="va">col_df</span>, colBy <span class="op">=</span> <span class="st">"Treatment"</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/density2.png"></p>
</div>
<div class="section level3">
<h3 id="perform-gene-ontology-analysis-with-panther">Perform Gene Ontology Analysis with PANTHER<a class="anchor" aria-label="anchor" href="#perform-gene-ontology-analysis-with-panther"></a>
</h3>
<p>This function sends a request to the PANTHER REST API for GO over-representation analysis on the user specified vector of genes. The API should accept Ensembl gene identifiers, Ensembl protein identifiers, Ensembl transcript identifiers, Entrez gene ids, gene symbols, NCBI GIs, HGNC Ids, International protein index ids, NCBI UniGene ids, UniProt accessions and UniProt ids. The function will return a tibble of the GO results for the user specified organism, annotation dataset, test type and test correction method.</p>
<p>NOTE: the <code>organism</code> parameter is given by the Entrez Taxon ID. Typical values for this parameter are “9606” for HUMAN, “10090” for MOUSE, “10116” for RAT. Any other Taxon ID could also be used. see <code><a href="reference/panther_go.html">?coriell::panther_go</a></code> for more information.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"CTNNB1"</span>, <span class="st">"ADAM17"</span>, <span class="st">"AXIN1"</span>, <span class="st">"AXIN2"</span>, <span class="st">"CCND2"</span>, <span class="st">"CSNK1E"</span>, <span class="st">"CTNNB1"</span>, </span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>           <span class="st">"CUL1"</span>, <span class="st">"DKK1"</span>, <span class="st">"DKK4"</span>, <span class="st">"DLL1"</span>, <span class="st">"DVL2"</span>, <span class="st">"FRAT1"</span>, <span class="st">"FZD1"</span>, <span class="st">"FZD8"</span>, </span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>           <span class="st">"GNAI1"</span>, <span class="st">"HDAC11"</span>, <span class="st">"HDAC2"</span>, <span class="st">"HDAC5"</span>, <span class="st">"HEY1"</span>, <span class="st">"HEY2"</span>, <span class="st">"JAG1"</span>, </span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>           <span class="st">"JAG2"</span>, <span class="st">"KAT2A"</span>, <span class="st">"LEF1"</span>, <span class="st">"MAML1"</span>, <span class="st">"MYC"</span>, <span class="st">"NCOR2"</span>, <span class="st">"NCSTN"</span>, </span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>           <span class="st">"NKD1"</span>, <span class="st">"NOTCH1"</span>, <span class="st">"NOTCH4"</span>, <span class="st">"NUMB"</span>, <span class="st">"PPARD"</span>, <span class="st">"PSEN2"</span>, <span class="st">"PTCH1"</span>, </span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>           <span class="st">"RBPJ"</span>, <span class="st">"SKP2"</span>, <span class="st">"TCF7"</span>, <span class="st">"TP53"</span>, <span class="st">"WNT1"</span>, <span class="st">"WNT5B"</span>, <span class="st">"WNT6"</span>)</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>go_results <span class="ot">&lt;-</span> <span class="fu">panther_go</span>(genes, <span class="at">organism =</span> <span class="st">"9606"</span>, <span class="at">annot_dataset =</span> <span class="st">"biological_process"</span>)</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(go_results, <span class="at">n =</span> <span class="dv">10</span>)</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> result_number number_in_list fold_enrichment      fdr expected number_in_reference   pValue plus_minus GO_term    description                            </span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                  <span class="er">&lt;</span>int<span class="sc">&gt;</span>           <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>    <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>    <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>               <span class="er">&lt;</span>int<span class="sc">&gt;</span>    <span class="er">&lt;</span>dbl<span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span>      <span class="er">&lt;</span>chr<span class="sc">&gt;</span>      <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                                  </span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="dv">1</span>                         <span class="dv">32</span>           <span class="fl">6.29</span>  <span class="fl">3.63e-17</span>    <span class="fl">5.09</span>                 <span class="dv">2525</span> <span class="fl">2.28e-21</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0007166</span> cell surface receptor signaling pathway</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">2</span>                         <span class="dv">24</span>           <span class="fl">10.6</span>  <span class="fl">5.46e-16</span>    <span class="fl">2.27</span>                 <span class="dv">1129</span> <span class="fl">6.87e-20</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0060429</span> epithelium development                 </span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">3</span>                         <span class="dv">17</span>           <span class="fl">24.3</span>  <span class="fl">8.23e-16</span>    <span class="fl">0.701</span>                 <span class="dv">348</span> <span class="fl">1.55e-19</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0198738</span> cell<span class="sc">-</span>cell signaling by wnt             </span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">4</span>                         <span class="dv">17</span>           <span class="fl">24.3</span>  <span class="fl">6.18e-16</span>    <span class="fl">0.701</span>                 <span class="dv">348</span> <span class="fl">1.55e-19</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0016055</span> Wnt signaling pathway                  </span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">5</span>                         <span class="dv">34</span>           <span class="fl">4.72</span>  <span class="fl">9.14e-16</span>    <span class="fl">7.20</span>                 <span class="dv">3576</span> <span class="fl">2.87e-19</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0010646</span> regulation of cell communication       </span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">6</span>                         <span class="dv">33</span>           <span class="fl">5.00</span>  <span class="fl">9.16e-16</span>    <span class="fl">6.60</span>                 <span class="dv">3277</span> <span class="fl">3.46e-19</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0048513</span> animal organ development               </span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">7</span>                         <span class="dv">19</span>           <span class="fl">17.0</span>  <span class="fl">9.19e-16</span>    <span class="fl">1.12</span>                  <span class="dv">555</span> <span class="fl">4.04e-19</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0048729</span> tissue morphogenesis                   </span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">8</span>                         <span class="dv">34</span>           <span class="fl">4.67</span>  <span class="fl">8.11e-16</span>    <span class="fl">7.28</span>                 <span class="dv">3615</span> <span class="fl">4.08e-19</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0023051</span> regulation of signaling                </span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">9</span>                         <span class="dv">12</span>           <span class="fl">70.9</span>  <span class="fl">7.58e-16</span>    <span class="fl">0.169</span>                  <span class="dv">84</span> <span class="fl">4.29e-19</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0060070</span> canonical Wnt signaling pathway        </span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">10</span>                        <span class="dv">13</span>           <span class="fl">52.5</span>  <span class="fl">7.24e-16</span>    <span class="fl">0.248</span>                 <span class="dv">123</span> <span class="fl">4.55e-19</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0007219</span> Notch signaling pathway</span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a><span class="co"># using ensembl ids ------------------------------------------------------------</span></span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a><span class="co"># all unique ensembl IDs for the above gene set</span></span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a>ensembl_ids <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ENSG00000162736"</span>, <span class="st">"ENSG00000143801"</span>, <span class="st">"ENSG00000177283"</span>, </span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"ENSG00000107984"</span>, <span class="st">"ENSG00000165879"</span>, <span class="st">"ENSG00000111186"</span>, </span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"ENSG00000118971"</span>, <span class="st">"ENSG00000125084"</span>, <span class="st">"ENSG00000196498"</span>, </span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"ENSG00000133961"</span>, <span class="st">"ENSG00000103126"</span>, <span class="st">"ENSG00000004975"</span>, </span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"ENSG00000141510"</span>, <span class="st">"ENSG00000108840"</span>, <span class="st">"ENSG00000168646"</span>, </span>
<span id="cb25-31"><a href="#cb25-31" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"ENSG00000151694"</span>, <span class="st">"ENSG00000115596"</span>, <span class="st">"ENSG00000101384"</span>, </span>
<span id="cb25-32"><a href="#cb25-32" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"ENSG00000213923"</span>, <span class="st">"ENSG00000163517"</span>, <span class="st">"ENSG00000168036"</span>, </span>
<span id="cb25-33"><a href="#cb25-33" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"ENSG00000138795"</span>, <span class="st">"ENSG00000145604"</span>, <span class="st">"ENSG00000081059"</span>, </span>
<span id="cb25-34"><a href="#cb25-34" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"ENSG00000161021"</span>, <span class="st">"ENSG00000204301"</span>, <span class="st">"ENSG00000112033"</span>, </span>
<span id="cb25-35"><a href="#cb25-35" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"ENSG00000196591"</span>, <span class="st">"ENSG00000135547"</span>, <span class="st">"ENSG00000198719"</span>, </span>
<span id="cb25-36"><a href="#cb25-36" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"ENSG00000127955"</span>, <span class="st">"ENSG00000157240"</span>, <span class="st">"ENSG00000055130"</span>, </span>
<span id="cb25-37"><a href="#cb25-37" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"ENSG00000104371"</span>, <span class="st">"ENSG00000164683"</span>, <span class="st">"ENSG00000136997"</span>, </span>
<span id="cb25-38"><a href="#cb25-38" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"ENSG00000185920"</span>, <span class="st">"ENSG00000148400"</span>, <span class="st">"ENSG00000283780"</span>, </span>
<span id="cb25-39"><a href="#cb25-39" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"ENSG00000275555"</span>, <span class="st">"ENSG00000238196"</span>, <span class="st">"ENSG00000235396"</span>, </span>
<span id="cb25-40"><a href="#cb25-40" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"ENSG00000232339"</span>, <span class="st">"ENSG00000223355"</span>, <span class="st">"ENSG00000206312"</span>, </span>
<span id="cb25-41"><a href="#cb25-41" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"ENSG00000234876"</span>)</span>
<span id="cb25-42"><a href="#cb25-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-43"><a href="#cb25-43" aria-hidden="true" tabindex="-1"></a>ensembl_results <span class="ot">&lt;-</span> <span class="fu">panther_go</span>(ensembl_ids, <span class="st">"9606"</span>, <span class="st">"biological_process"</span>)</span>
<span id="cb25-44"><a href="#cb25-44" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ensembl_results, <span class="at">n =</span> <span class="dv">10</span>)</span>
<span id="cb25-45"><a href="#cb25-45" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> result_number number_in_list fold_enrichment      fdr expected number_in_reference   pValue plus_minus GO_term   description                                               </span>
<span id="cb25-46"><a href="#cb25-46" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                  <span class="er">&lt;</span>int<span class="sc">&gt;</span>           <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>    <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>    <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>               <span class="er">&lt;</span>int<span class="sc">&gt;</span>    <span class="er">&lt;</span>dbl<span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span>      <span class="er">&lt;</span>chr<span class="sc">&gt;</span>     <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                                                     </span>
<span id="cb25-47"><a href="#cb25-47" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="dv">1</span>                         <span class="dv">12</span>           <span class="fl">78.4</span>  <span class="fl">1.70e-15</span>    <span class="fl">0.153</span>                  <span class="dv">84</span> <span class="fl">1.07e-19</span> <span class="sc">+</span>          GO<span class="sc">:</span>00600… canonical Wnt signaling pathway                           </span>
<span id="cb25-48"><a href="#cb25-48" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">2</span>                         <span class="dv">29</span>           <span class="fl">6.30</span>  <span class="fl">1.25e-15</span>    <span class="fl">4.60</span>                 <span class="dv">2525</span> <span class="fl">1.57e-19</span> <span class="sc">+</span>          GO<span class="sc">:</span>00071… cell surface receptor signaling pathway                   </span>
<span id="cb25-49"><a href="#cb25-49" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">3</span>                         <span class="dv">16</span>           <span class="fl">25.2</span>  <span class="fl">4.33e-15</span>    <span class="fl">0.634</span>                 <span class="dv">348</span> <span class="fl">8.17e-19</span> <span class="sc">+</span>          GO<span class="sc">:</span>01987… cell<span class="sc">-</span>cell signaling by wnt                                </span>
<span id="cb25-50"><a href="#cb25-50" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">4</span>                         <span class="dv">16</span>           <span class="fl">25.2</span>  <span class="fl">3.25e-15</span>    <span class="fl">0.634</span>                 <span class="dv">348</span> <span class="fl">8.17e-19</span> <span class="sc">+</span>          GO<span class="sc">:</span>00160… Wnt signaling pathway                                     </span>
<span id="cb25-51"><a href="#cb25-51" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">5</span>                         <span class="dv">16</span>           <span class="fl">20.9</span>  <span class="fl">4.76e-14</span>    <span class="fl">0.767</span>                 <span class="dv">421</span> <span class="fl">1.50e-17</span> <span class="sc">+</span>          GO<span class="sc">:</span>19051… cell surface receptor signaling pathway involved <span class="cf">in</span> cell<span class="sc">-</span>…</span>
<span id="cb25-52"><a href="#cb25-52" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">6</span>                         <span class="dv">21</span>           <span class="fl">10.2</span>  <span class="fl">9.38e-14</span>    <span class="fl">2.06</span>                 <span class="dv">1129</span> <span class="fl">3.54e-17</span> <span class="sc">+</span>          GO<span class="sc">:</span>00604… epithelium development                                    </span>
<span id="cb25-53"><a href="#cb25-53" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">7</span>                         <span class="dv">29</span>           <span class="fl">5.01</span>  <span class="fl">1.92e-13</span>    <span class="fl">5.78</span>                 <span class="dv">3174</span> <span class="fl">8.46e-17</span> <span class="sc">+</span>          GO<span class="sc">:</span>00099… regulation of signal transduction                         </span>
<span id="cb25-54"><a href="#cb25-54" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">8</span>                         <span class="dv">30</span>           <span class="fl">4.60</span>  <span class="fl">2.68e-13</span>    <span class="fl">6.52</span>                 <span class="dv">3576</span> <span class="fl">1.35e-16</span> <span class="sc">+</span>          GO<span class="sc">:</span>00106… regulation of cell communication                          </span>
<span id="cb25-55"><a href="#cb25-55" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">9</span>                         <span class="dv">30</span>           <span class="fl">4.55</span>  <span class="fl">3.23e-13</span>    <span class="fl">6.59</span>                 <span class="dv">3615</span> <span class="fl">1.83e-16</span> <span class="sc">+</span>          GO<span class="sc">:</span>00230… regulation of signaling                                   </span>
<span id="cb25-56"><a href="#cb25-56" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">10</span>                        <span class="dv">29</span>           <span class="fl">4.86</span>  <span class="fl">3.22e-13</span>    <span class="fl">5.97</span>                 <span class="dv">3277</span> <span class="fl">2.02e-16</span> <span class="sc">+</span>          GO<span class="sc">:</span>00485… animal organ development </span>
<span id="cb25-57"><a href="#cb25-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-58"><a href="#cb25-58" aria-hidden="true" tabindex="-1"></a><span class="co"># using mouse genes ------------------------------------------------------------</span></span>
<span id="cb25-59"><a href="#cb25-59" aria-hidden="true" tabindex="-1"></a>mouse_genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Adam17"</span>, <span class="st">"Axin1"</span>, <span class="st">"Axin2"</span>, <span class="st">"Ccnd2"</span>, <span class="st">"Csnk1e"</span>, <span class="st">"Ctnnb1"</span>, </span>
<span id="cb25-60"><a href="#cb25-60" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"Cul1"</span>, <span class="st">"Dkk1"</span>, <span class="st">"Dkk4"</span>, <span class="st">"Dll1"</span>, <span class="st">"Dvl2"</span>, <span class="st">"Frat1"</span>, <span class="st">"Fzd1"</span>, </span>
<span id="cb25-61"><a href="#cb25-61" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"Fzd8"</span>, <span class="st">"Gnai1"</span>, <span class="st">"Hdac11"</span>, <span class="st">"Hdac2"</span>, <span class="st">"Hdac5"</span>, <span class="st">"Hey1"</span>, <span class="st">"Hey2"</span>, </span>
<span id="cb25-62"><a href="#cb25-62" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"Jag1"</span>, <span class="st">"Lef1"</span>, <span class="st">"Maml1"</span>, <span class="st">"Myc"</span>, <span class="st">"Ncor2"</span>, <span class="st">"Ncstn"</span>, <span class="st">"Notch1"</span>, </span>
<span id="cb25-63"><a href="#cb25-63" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"Notch4"</span>, <span class="st">"Numb"</span>, <span class="st">"Ppard"</span>, <span class="st">"Psen2"</span>, <span class="st">"Ptch1"</span>, <span class="st">"Skp2"</span>, <span class="st">"Tcf7"</span>, </span>
<span id="cb25-64"><a href="#cb25-64" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"Wnt1"</span>, <span class="st">"Wnt5b"</span>, <span class="st">"Wnt6"</span>)</span>
<span id="cb25-65"><a href="#cb25-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-66"><a href="#cb25-66" aria-hidden="true" tabindex="-1"></a>mouse_results <span class="ot">&lt;-</span> <span class="fu">panther_go</span>(mouse_genes, <span class="at">organism =</span> <span class="st">"10090"</span>, <span class="at">annot_dataset =</span> <span class="st">"biological_process"</span>)</span>
<span id="cb25-67"><a href="#cb25-67" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mouse_results, <span class="at">n =</span> <span class="dv">10</span>)</span>
<span id="cb25-68"><a href="#cb25-68" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> result_number number_in_list fold_enrichment      fdr expected number_in_reference   pValue plus_minus GO_term    description                                                     </span>
<span id="cb25-69"><a href="#cb25-69" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                  <span class="er">&lt;</span>int<span class="sc">&gt;</span>           <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>    <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>    <span class="er">&lt;</span>dbl<span class="sc">&gt;</span>               <span class="er">&lt;</span>int<span class="sc">&gt;</span>    <span class="er">&lt;</span>dbl<span class="sc">&gt;</span> <span class="er">&lt;</span>chr<span class="sc">&gt;</span>      <span class="er">&lt;</span>chr<span class="sc">&gt;</span>      <span class="er">&lt;</span>chr<span class="sc">&gt;</span>                                                           </span>
<span id="cb25-70"><a href="#cb25-70" aria-hidden="true" tabindex="-1"></a><span class="er">&gt;</span> <span class="dv">1</span>                         <span class="dv">13</span>           <span class="fl">97.8</span>  <span class="fl">2.50e-18</span>    <span class="fl">0.133</span>                  <span class="dv">80</span> <span class="fl">1.58e-22</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0060070</span> canonical Wnt signaling pathway                                 </span>
<span id="cb25-71"><a href="#cb25-71" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">2</span>                         <span class="dv">16</span>           <span class="fl">40.5</span>  <span class="fl">4.10e-18</span>    <span class="fl">0.396</span>                 <span class="dv">238</span> <span class="fl">5.18e-22</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0198738</span> cell<span class="sc">-</span>cell signaling by wnt                                      </span>
<span id="cb25-72"><a href="#cb25-72" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">3</span>                         <span class="dv">16</span>           <span class="fl">40.5</span>  <span class="fl">2.73e-18</span>    <span class="fl">0.396</span>                 <span class="dv">238</span> <span class="fl">5.18e-22</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0016055</span> Wnt signaling pathway                                           </span>
<span id="cb25-73"><a href="#cb25-73" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">4</span>                         <span class="dv">23</span>           <span class="fl">13.3</span>  <span class="fl">3.77e-18</span>    <span class="fl">1.72</span>                 <span class="dv">1037</span> <span class="fl">9.53e-22</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0060429</span> epithelium development                                          </span>
<span id="cb25-74"><a href="#cb25-74" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">5</span>                         <span class="dv">27</span>           <span class="fl">8.52</span>  <span class="fl">8.41e-18</span>    <span class="fl">3.17</span>                 <span class="dv">1907</span> <span class="fl">2.66e-21</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0007166</span> cell surface receptor signaling pathway                         </span>
<span id="cb25-75"><a href="#cb25-75" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">6</span>                         <span class="dv">16</span>           <span class="fl">33.0</span>  <span class="fl">3.10e-17</span>    <span class="fl">0.485</span>                 <span class="dv">292</span> <span class="fl">1.18e-20</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">1905114</span> cell surface receptor signaling pathway involved <span class="cf">in</span> cell<span class="sc">-</span>cell s…</span>
<span id="cb25-76"><a href="#cb25-76" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">7</span>                         <span class="dv">25</span>           <span class="fl">9.04</span>  <span class="fl">1.39e-16</span>    <span class="fl">2.77</span>                 <span class="dv">1664</span> <span class="fl">6.17e-20</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0009888</span> tissue development                                              </span>
<span id="cb25-77"><a href="#cb25-77" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">8</span>                         <span class="dv">27</span>           <span class="fl">7.21</span>  <span class="fl">3.95e-16</span>    <span class="fl">3.75</span>                 <span class="dv">2254</span> <span class="fl">2.00e-19</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0009653</span> anatomical structure morphogenesis                              </span>
<span id="cb25-78"><a href="#cb25-78" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">9</span>                         <span class="dv">29</span>           <span class="fl">6.06</span>  <span class="fl">4.50e-16</span>    <span class="fl">4.78</span>                 <span class="dv">2878</span> <span class="fl">2.56e-19</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0009966</span> regulation of signal transduction                               </span>
<span id="cb25-79"><a href="#cb25-79" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">10</span>                        <span class="dv">21</span>           <span class="fl">12.3</span>  <span class="fl">1.04e-15</span>    <span class="fl">1.71</span>                 <span class="dv">1026</span> <span class="fl">6.57e-19</span> <span class="sc">+</span>          GO<span class="sc">:</span><span class="dv">0009887</span> animal organ morphogenesis </span></code></pre></div>
</div>
<div class="section level3">
<h3 id="perform-correlation-permutation-test-using-multiple-cores">Perform correlation permutation test using multiple cores<a class="anchor" aria-label="anchor" href="#perform-correlation-permutation-test-using-multiple-cores"></a>
</h3>
<p>Note: if the number of possible permutations of your data is less than the desired number of permutations given by the parameter <code>n_perm</code> then an exact test on all of the real permutations will be performed instead of random sampling. For example, if you only have 6 samples (720 possible permutations) but set <code>n_perm = 1000</code> only 720 permutations (i.e. the exact test) will be tested. The <code>permutation_correlation_test</code> function will display a message if this occurs. The function accepts a numeric matrix as input or a data.frame that can be converted to a numeric matrix. In the case a data.frame is passed to argument <code>X</code>, the function call will display a message confirming the conversion. In most cases this conversion is inconsequential.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(coriell)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(methylKit)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co"># define age dataframe -- ages matching column order</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>ages <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">age =</span> <span class="fu">c</span>(<span class="dv">30</span>, <span class="dv">80</span>, <span class="dv">34</span>, <span class="dv">30</span>, <span class="dv">80</span>, <span class="dv">40</span>, <span class="dv">35</span>, <span class="dv">80</span>))</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a><span class="co"># simulate a methylation dataset</span></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>sim_meth <span class="ot">&lt;-</span> <span class="fu">dataSim</span>(</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">replicates =</span> <span class="dv">8</span>,</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">sites =</span> <span class="dv">1000</span>,</span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">treatment =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">1</span>, <span class="dv">4</span>), <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">4</span>)),</span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">covariates =</span> ages,</span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample.ids =</span> <span class="fu">c</span>(<span class="fu">paste0</span>(<span class="st">"test"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>), <span class="fu">paste0</span>(<span class="st">"ctrl"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>))</span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a><span class="co"># extract the methylation as percentages and coerce to data.frame</span></span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a>perc_meth <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">percMethylation</span>(sim_meth))</span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(perc_meth)</span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span>       test1     test2     test3     test4    ctrl1     ctrl2     ctrl3      ctrl4</span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">1</span>  <span class="fl">48.14815</span>  <span class="fl">54.05405</span>  <span class="fl">45.45455</span>  <span class="fl">57.89474</span> <span class="fl">32.46753</span> <span class="fl">24.418605</span>  <span class="fl">4.687500</span>  <span class="fl">30.769231</span></span>
<span id="cb26-22"><a href="#cb26-22" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">2</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>  <span class="fl">0.00000</span>  <span class="fl">0.000000</span>  <span class="fl">0.000000</span>   <span class="fl">0.000000</span></span>
<span id="cb26-23"><a href="#cb26-23" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">3</span>  <span class="fl">38.20598</span>  <span class="fl">16.34615</span>  <span class="fl">12.50000</span>  <span class="fl">60.00000</span> <span class="fl">25.42373</span>  <span class="fl">3.636364</span> <span class="fl">18.421053</span>  <span class="fl">32.335329</span></span>
<span id="cb26-24"><a href="#cb26-24" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">4</span> <span class="fl">100.00000</span> <span class="fl">100.00000</span> <span class="fl">100.00000</span> <span class="fl">100.00000</span> <span class="fl">86.15385</span> <span class="fl">72.340426</span> <span class="fl">86.666667</span> <span class="fl">100.000000</span></span>
<span id="cb26-25"><a href="#cb26-25" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">5</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>  <span class="fl">0.00000</span>  <span class="fl">0.000000</span>  <span class="fl">0.000000</span>  <span class="fl">10.526316</span></span>
<span id="cb26-26"><a href="#cb26-26" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">6</span>  <span class="fl">26.98413</span>  <span class="fl">12.63158</span>  <span class="fl">23.07692</span>  <span class="fl">10.00000</span> <span class="fl">22.22222</span>  <span class="fl">0.000000</span>  <span class="fl">3.174603</span>   <span class="fl">4.166667</span></span>
<span id="cb26-27"><a href="#cb26-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-28"><a href="#cb26-28" aria-hidden="true" tabindex="-1"></a><span class="co"># permutation testing -----------------------------------------------------------</span></span>
<span id="cb26-29"><a href="#cb26-29" aria-hidden="true" tabindex="-1"></a><span class="co"># perform permutation testing using 4 cores and 10000 permutations</span></span>
<span id="cb26-30"><a href="#cb26-30" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">permutation_correlation_test</span>(</span>
<span id="cb26-31"><a href="#cb26-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">X =</span> perc_meth, </span>
<span id="cb26-32"><a href="#cb26-32" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> ages<span class="sc">$</span>age, </span>
<span id="cb26-33"><a href="#cb26-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">n_cores =</span> <span class="dv">4</span>, </span>
<span id="cb26-34"><a href="#cb26-34" aria-hidden="true" tabindex="-1"></a>  <span class="at">n_perm =</span> <span class="dv">10000</span>,</span>
<span id="cb26-35"><a href="#cb26-35" aria-hidden="true" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">"spearman"</span> </span>
<span id="cb26-36"><a href="#cb26-36" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb26-37"><a href="#cb26-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-38"><a href="#cb26-38" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res)</span>
<span id="cb26-39"><a href="#cb26-39" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span>       test1     test2     test3     test4    ctrl1     ctrl2     ctrl3      ctrl4        cor empirical.p       fdr</span>
<span id="cb26-40"><a href="#cb26-40" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">1</span>  <span class="fl">48.14815</span>  <span class="fl">54.05405</span>  <span class="fl">45.45455</span>  <span class="fl">57.89474</span> <span class="fl">32.46753</span> <span class="fl">24.418605</span>  <span class="fl">4.687500</span>  <span class="fl">30.769231</span> <span class="sc">-</span><span class="fl">0.3437200</span>      <span class="fl">0.1925</span> <span class="fl">0.4812073</span></span>
<span id="cb26-41"><a href="#cb26-41" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">2</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>  <span class="fl">0.00000</span>  <span class="fl">0.000000</span>  <span class="fl">0.000000</span>   <span class="fl">0.000000</span>         <span class="cn">NA</span>          <span class="cn">NA</span>        <span class="cn">NA</span></span>
<span id="cb26-42"><a href="#cb26-42" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">3</span>  <span class="fl">38.20598</span>  <span class="fl">16.34615</span>  <span class="fl">12.50000</span>  <span class="fl">60.00000</span> <span class="fl">25.42373</span>  <span class="fl">3.636364</span> <span class="fl">18.421053</span>  <span class="fl">32.335329</span> <span class="sc">-</span><span class="fl">0.3559957</span>      <span class="fl">0.1903</span> <span class="fl">0.4812073</span></span>
<span id="cb26-43"><a href="#cb26-43" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">4</span> <span class="fl">100.00000</span> <span class="fl">100.00000</span> <span class="fl">100.00000</span> <span class="fl">100.00000</span> <span class="fl">86.15385</span> <span class="fl">72.340426</span> <span class="fl">86.666667</span> <span class="fl">100.000000</span> <span class="sc">-</span><span class="fl">0.3093992</span>      <span class="fl">0.2195</span> <span class="fl">0.4812073</span></span>
<span id="cb26-44"><a href="#cb26-44" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">5</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>  <span class="fl">0.00000</span>  <span class="fl">0.000000</span>  <span class="fl">0.000000</span>  <span class="fl">10.526316</span>  <span class="fl">0.4252433</span>      <span class="fl">0.3769</span> <span class="fl">0.4812073</span></span>
<span id="cb26-45"><a href="#cb26-45" aria-hidden="true" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">6</span>  <span class="fl">26.98413</span>  <span class="fl">12.63158</span>  <span class="fl">23.07692</span>  <span class="fl">10.00000</span> <span class="fl">22.22222</span>  <span class="fl">0.000000</span>  <span class="fl">3.174603</span>   <span class="fl">4.166667</span> <span class="sc">-</span><span class="fl">0.2946172</span>      <span class="fl">0.2334</span> <span class="fl">0.4812073</span></span></code></pre></div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/coriell-research/coriell/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/coriell-research/coriell/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing coriell</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Gennaro Calendo <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Gennaro Calendo.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
