<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="This package provides many helper functions for common bioinformatics tasks such as plotting and summarizing the results of differential expression analyses, creating heatmaps from expression data, performing permutation tests, selecting outliers, analyzing PCA plots, and many other common tasks.">
<title>Convenience Functions for Bioinformatics • coriell</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Convenience Functions for Bioinformatics">
<meta property="og:description" content="This package provides many helper functions for common bioinformatics tasks such as plotting and summarizing the results of differential expression analyses, creating heatmaps from expression data, performing permutation tests, selecting outliers, analyzing PCA plots, and many other common tasks.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">coriell</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.15.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="articles/Practical-analysis-with-SummarizedExperiments.html">Practical analysis with SummarizedExperiments</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/coriell-research/coriell/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="coriell">coriell<a class="anchor" aria-label="anchor" href="#coriell"></a>
</h1></div>
<p>This package contains helper functions for common bioinformatics tasks (and some not-so-common tasks). If you find yourself reusing old code over and over, let me know and we’ll create a function and put it in this package.</p>
<p>I will be adding tests, however, <strong>please test functions and let me know when things are broken</strong> by either raising an issue on Github or contacting me.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The latest version can be installed from Github using:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"coriell-research/coriell"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h2>
<ul>
<li><a href="https://coriell-research.github.io/coriell/#summarize-results-from-differential-expression-test" class="external-link">Summarize results from differential expression analysis</a></li>
<li><a href="https://coriell-research.github.io/coriell/#create-volcano-plot-from-differential-expression-results" class="external-link">Create volcano plot from differential expression results</a></li>
<li><a href="https://coriell-research.github.io/coriell/#create-md-plot-from-differential-expression-results" class="external-link">Create md plot from differential expression results</a></li>
<li><a href="https://coriell-research.github.io/coriell/#heatmap-with-sensible-defaults" class="external-link">Heatmap with sensible defaults</a></li>
<li><a href="https://coriell-research.github.io/coriell/#parallel-coordinates-plot-of-expression-matrix" class="external-link">Parallel coordinates plot of expression matrix</a></li>
<li><a href="https://coriell-research.github.io/coriell/#boxplot-of-expression-matrix" class="external-link">Boxplot of expression matrix</a></li>
<li><a href="https://coriell-research.github.io/coriell/#density-plot-of-expression-matrix" class="external-link">Density plot of expression matrix</a></li>
</ul>
<p>The package also contains many other convenience functions so be sure to check out the <a href="https://coriell-research.github.io/coriell/reference/index.html" class="external-link">reference</a> page as well.</p>
<div class="section level3">
<h3 id="built-in-datasets">Built-in datasets<a class="anchor" aria-label="anchor" href="#built-in-datasets"></a>
</h3>
<p>These datasets are built into the package for testing purposes and are used below to illustrate some functionality.</p>
<ul>
<li>
<code>GSE161650_de</code> : Differential expression results of THZ1 vs DMSO from GSE161650</li>
<li>
<code>GSE161650_lc</code> : Normalized log2 counts from THZ1 vs DMSO replicates from GSE161650</li>
</ul>
<p>The head of <code>GSE161650_de</code> looks like:</p>
<pre><code>&gt;   feature_id     logFC unshrunk.logFC    logCPM       PValue          FDR
&gt; 1        JUN  5.759233       5.759908  9.079350 2.919097e-14 1.990666e-10
&gt; 2       IER5  3.931325       3.931420 10.158336 3.365738e-14 1.990666e-10
&gt; 3    GADD45B  5.813030       5.814071  8.432432 6.435666e-14 2.537583e-10
&gt; 4       IER2  4.457981       4.458016 12.223835 1.528890e-13 4.521309e-10
&gt; 5     PIK3R3 -4.018325      -4.019603  7.124408 2.122484e-13 4.752097e-10
&gt; 6     HEXIM1  4.497345       4.497844  8.275561 2.696985e-13 4.752097e-10</code></pre>
<p>And <code>GSE161650_lc</code>:</p>
<pre><code>&gt;          DMSO.1     DMSO.2     DMSO.3    THZ1.1    THZ1.2     THZ1.3
&gt; A1BG  5.3323512  5.4576081  5.2876011  6.703752 6.8090471  6.7908595
&gt; AAAS  3.8738768  3.8839857  3.5242625  3.768811 4.1406003  3.7454773
&gt; AACS  2.1381539  2.3748462  2.2761971  3.769163 3.4352003  3.4064114
&gt; AADAT 0.8240013 -0.2391457 -0.8699138 -1.342273 0.1967574 -0.8355924
&gt; AAED1 1.2814586  1.5332476  2.0735095  2.443188 2.1270932  0.9707206
&gt; AAGAB 6.8747238  6.7396922  6.6670762  6.756664 6.5840839  6.7662317</code></pre>
<p>See the package documentation <code><a href="reference/GSE161650_de.html">?GSE161650_de</a></code> and <code><a href="reference/GSE161650_lc.html">?GSE161650_lc</a></code> for citation information.</p>
</div>
<div class="section level3">
<h3 id="summarize-results-from-differential-expression-test">Summarize results from differential expression test<a class="anchor" aria-label="anchor" href="#summarize-results-from-differential-expression-test"></a>
</h3>
<p>Return a table of up/down/non-de counts and their percentages.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>de <span class="ot">&lt;-</span> GSE161650_de</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="fu">summarize_dge</span>(de, <span class="at">fdr =</span> <span class="fl">0.1</span>)</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="sc">&gt;</span>     Direction    N Percent</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">1</span>          Up <span class="dv">2588</span>   <span class="fl">23.52</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">2</span>        Down <span class="dv">3365</span>   <span class="fl">30.59</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="sc">&gt;</span> <span class="dv">3</span> Unperturbed <span class="dv">5049</span>   <span class="fl">45.89</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-volcano-plot-from-differential-expression-results">Create volcano plot from differential expression results<a class="anchor" aria-label="anchor" href="#create-volcano-plot-from-differential-expression-results"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">de</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/volcano1.png"></p>
<p>Different significance levels can be used to filter the plotted points. For example, significance levels can be set by specifying the <code>fdr</code> and <code>lfc</code> values.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">de</span>, fdr <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/volcano2.png"></p>
<p>Labels for the counts will be displayed by default. To remove them set <code>annotate_counts = FALSE</code></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">de</span>, fdr <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>, annotate_counts <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/volcano3.png"></p>
<p>Text labels can also be added for the DE genes by setting <code>label_sig = TRUE</code>. If <code>label_sig = TRUE</code> then the <code>lab</code> argument must also designate a column containing the names of the items to plot. Additional arguments can be passed to <code><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">ggrepel::geom_text_repel()</a></code> via <code>...</code> if you need to modify the behavior of the labels.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">de</span>, fdr <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>, label_sig <span class="op">=</span> <span class="cn">TRUE</span>, lab <span class="op">=</span> <span class="st">"feature_id"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/volcano4.png"></p>
<p>All aesthetics of the points can also be changed</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">de</span>,</span>
<span>  fdr <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>  lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>  up_color <span class="op">=</span> <span class="st">"firebrick"</span>,</span>
<span>  down_color <span class="op">=</span> <span class="st">"midnightblue"</span>,</span>
<span>  nonde_color <span class="op">=</span> <span class="st">"grey80"</span>,</span>
<span>  up_alpha <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  down_alpha <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  nonde_alpha <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  up_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  down_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  nonde_size <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/volcano5.png"></p>
</div>
<div class="section level3">
<h3 id="create-md-plot-from-differential-expression-results">Create md plot from differential expression results<a class="anchor" aria-label="anchor" href="#create-md-plot-from-differential-expression-results"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">de</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/md1.png"></p>
<p>Different significance levels can be used to filter the plotted points. For example, significance levels can be set by specifying the <code>fdr</code> and <code>lfc</code> values.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">de</span>, fdr <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/md2.png"></p>
<p>Labels for the counts will be displayed by default. To remove them set <code>annotate_counts = FALSE</code></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">de</span>, fdr <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>, annotate_counts <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/md3.png"></p>
<p>Text labels can also be added for the DE genes by setting <code>label_sig = TRUE</code>. If <code>label_sig = TRUE</code> then the <code>lab</code> argument must also designate a column containing the names of the items to plot. Additional arguments can be passed to <code><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">ggrepel::geom_text_repel()</a></code> via <code>...</code> if you need to modify the behavior of the labels.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">de</span>, fdr <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>, label_sig <span class="op">=</span> <span class="cn">TRUE</span>, lab <span class="op">=</span> <span class="st">"feature_id"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/md5.png"></p>
<p>All aesthetics of the points can also be changed</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">de</span>,</span>
<span>  fdr <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>  lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>  up_color <span class="op">=</span> <span class="st">"firebrick"</span>,</span>
<span>  down_color <span class="op">=</span> <span class="st">"midnightblue"</span>,</span>
<span>  nonde_color <span class="op">=</span> <span class="st">"grey80"</span>,</span>
<span>  up_alpha <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  down_alpha <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  nonde_alpha <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  up_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  down_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  nonde_size <span class="op">=</span> <span class="fl">0.5</span></span>
<span>  <span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"THZ1 vs Control"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/md4.png"></p>
</div>
<div class="section level3">
<h3 id="heatmap-with-sensible-defaults">Heatmap with sensible defaults<a class="anchor" aria-label="anchor" href="#heatmap-with-sensible-defaults"></a>
</h3>
<p>We often use the same settings when making calls to <code>pheatmap</code>. This function is a wrapper around <code>pheatmap</code> which uses sensible default values for expression data.</p>
<p>Any of these default values can be overridden by simply supplying the arguments to <code>quickmap</code> as you would <code>pheatmap</code>. This also allows for additional arguments to be passed to the <code>quickmap</code> function for creating row and column annotations.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="va">GSE161650_lc</span></span>
<span></span>
<span><span class="co"># plot a heatmap of the logCPM values</span></span>
<span><span class="fu"><a href="reference/quickmap.html">quickmap</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/quickmap1.png"></p>
<p>Other <code>pheatmap</code> arguments can be passed to the <code>quickmap</code> function as well.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create annotation for columns</span></span>
<span><span class="va">col_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Treatment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DMSO"</span>, <span class="st">"THZ1"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">col_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create color scheme for treatment conditions</span></span>
<span><span class="va">ann_colors</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Treatment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DMSO"</span> <span class="op">=</span> <span class="st">"steelblue"</span>, <span class="st">"THZ1"</span> <span class="op">=</span> <span class="st">"firebrick"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot the heatmap, passing additional args to pheatmap</span></span>
<span><span class="fu"><a href="reference/quickmap.html">quickmap</a></span><span class="op">(</span></span>
<span>  <span class="va">logcounts</span>,</span>
<span>  annotation_col <span class="op">=</span> <span class="va">col_df</span>,</span>
<span>  annotation_colors <span class="op">=</span> <span class="va">ann_colors</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Treatment vs Control"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/quickmap2.png"></p>
<p>Additional arguments can be set to limit the scales of the heatmap as well as remove low variance features prior to plotting. To remove low variance features set the <code>removeVar</code> argument to the desired proportion of features to drop.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/quickmap.html">quickmap</a></span><span class="op">(</span></span>
<span>  <span class="va">logcounts</span>,</span>
<span>  removeVar <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  annotation_col <span class="op">=</span> <span class="va">col_df</span>,</span>
<span>  annotation_colors <span class="op">=</span> <span class="va">ann_colors</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"THZ1 vs DMSO"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/quickmap3.png"></p>
<p>The colors of the heatmap scale can also be ‘fixed’ above and below a certain threshold with the <code>fix_extreme</code> and <code>thresh</code> arguments</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/quickmap.html">quickmap</a></span><span class="op">(</span></span>
<span>    <span class="va">logcounts</span>,</span>
<span>    removeVar <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>    fix_extreme <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    thresh <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>    annotation_col <span class="op">=</span> <span class="va">col_df</span>,</span>
<span>    annotation_colors <span class="op">=</span> <span class="va">ann_colors</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"THZ1 vs DMSO"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/quickmap4.png"></p>
</div>
<div class="section level3">
<h3 id="parallel-coordinates-plot-of-expression-matrix">Parallel coordinates plot of expression matrix<a class="anchor" aria-label="anchor" href="#parallel-coordinates-plot-of-expression-matrix"></a>
</h3>
<p>Parallel coordinates plots (PCP), as well as boxplots and density plots (below), can be useful tools for examining expression patterns across samples before and after normalization, for example. To create a PCP with <code>coriell</code> use the <code><a href="reference/plot_parallel.html">plot_parallel()</a></code> function.</p>
<p>Using <code>logcounts</code> define above:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create PCP plot -- passing additional alpha value to geom_line()</span></span>
<span><span class="fu"><a href="reference/plot_parallel.html">plot_parallel</a></span><span class="op">(</span><span class="va">logcounts</span>, alpha <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/parallel.png"></p>
<p>If metadata is supplied then the PCP can be colored by the supplied variable name. Using <code>col_df</code> defined above:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_parallel.html">plot_parallel</a></span><span class="op">(</span><span class="va">logcounts</span>, <span class="va">col_df</span>, colBy <span class="op">=</span> <span class="st">"Treatment"</span>, alpha <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/parallel2.png"></p>
</div>
<div class="section level3">
<h3 id="boxplot-of-expression-matrix">Boxplot of expression matrix<a class="anchor" aria-label="anchor" href="#boxplot-of-expression-matrix"></a>
</h3>
<p>Likewise, boxplots of the expression values can be visualized in a similar fashion with <code><a href="reference/plot_boxplot.html">plot_boxplot()</a></code></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_boxplot.html">plot_boxplot</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/boxplot.png"></p>
<p>If metadata are supplied the boxes can be colored by the desired metadata column and additional arguments can be passed to <code><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot()</a></code> to modify the boxplots.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_boxplot.html">plot_boxplot</a></span><span class="op">(</span><span class="va">logcounts</span>, <span class="va">col_df</span>, fillBy <span class="op">=</span> <span class="st">"Treatment"</span>, outlier.shape <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/boxplot2.png"></p>
</div>
<div class="section level3">
<h3 id="density-plot-of-expression-matrix">Density plot of expression matrix<a class="anchor" aria-label="anchor" href="#density-plot-of-expression-matrix"></a>
</h3>
<p>Density plots showing the expression values for samples or groups can be plotted using the <code><a href="reference/plot_density.html">plot_density()</a></code> function.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_density.html">plot_density</a></span><span class="op">(</span><span class="va">logcounts</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/density.png"></p>
<p>If metadata are supplied, the lines can be colored by the desired metadata column and additional arguments can be passed to <code><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density()</a></code> to modify the density layer.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_density.html">plot_density</a></span><span class="op">(</span><span class="va">logcounts</span>, <span class="va">col_df</span>, colBy <span class="op">=</span> <span class="st">"Treatment"</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/density2.png"></p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/coriell-research/coriell/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/coriell-research/coriell/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing coriell</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Gennaro Calendo <br><small class="roles"> Author, maintainer </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gennaro Calendo.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
