<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Convenience functions for bioinformatics • coriell</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Convenience functions for bioinformatics">
<meta property="og:description" content="Functions for plotting and transforming data.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">coriell</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/RNA-seq-workflow.html">RNA-seq-workflow</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="coriell" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#coriell" class="anchor"></a>coriell</h1></div>
<p>Helper functions for common bioinformatics tasks. If you find yourself reusing old code over and over, let me know and we’ll create a function and put it in this package.</p>
<p>I will be adding tests, however, <strong>please test functions and let me know when things are broken</strong> by either raising an issue on Github or contacting me.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>Since this package will be constantly changing be sure to install the latest version from Github using:</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="co"># Make sure you have devtools installed</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>

<span class="co"># Then install using devtools::install_github</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"coriell-research/coriell"</span><span class="op">)</span></pre></div>
</div>
<div id="examples" class="section level2">
<h2 class="hasAnchor">
<a href="#examples" class="anchor"></a>Examples</h2>
<p>For an overview of how to use some of these functions in a typical analysis check out the RNA-seq vignette (more vignettes to come)</p>
<ul>
<li><a href="https://coriell-research.github.io/coriell/#perform-correlation-permutation-test-using-multiple-cores">Correlate methylation data with age</a></li>
<li><a href="https://coriell-research.github.io/coriell/#extract-random-correlations-from-a-dataframe">Create a null distribution of correlations</a></li>
<li><a href="https://coriell-research.github.io/coriell/#convert-edger-results-into-a-tidy-dataframe">Convert edgeR results into tidy dataframes</a></li>
<li><a href="https://coriell-research.github.io/coriell/#summarize-results-from-differential-expression-test">Summarize results from differential expression analysis</a></li>
<li><a href="https://coriell-research.github.io/coriell/#create-volcano-plot-from-differential-expression-results">Create volcano plot from differential expression results</a></li>
<li><a href="https://coriell-research.github.io/coriell/#create-md-plot-from-differential-expression-results">Create md plot from differential expression results</a></li>
<li><a href="https://coriell-research.github.io/coriell/#heatmap-with-sensible-defaults">Heatmap with sensible defaults</a></li>
<li><a href="https://coriell-research.github.io/coriell/#z-score-a-dataframe">Z-score a dataframe</a></li>
<li><a href="https://coriell-research.github.io/coriell/#convert-a-list-of-sets-into-a-binary-matrix">Convert a list of sets into a binary matrix</a></li>
<li><a href="https://coriell-research.github.io/coriell/#get-statistics-for-all-pairwise-combinations-of-a-list-of-sets">Get statistics for all pairwise combinations of a list of sets</a></li>
<li><a href="https://coriell-research.github.io/coriell/#perform-gene-ontology-analysis-with-panther">Perform Gene Ontology Analysis with PANTHER</a></li>
</ul>
<div id="perform-correlation-permutation-test-using-multiple-cores" class="section level3">
<h3 class="hasAnchor">
<a href="#perform-correlation-permutation-test-using-multiple-cores" class="anchor"></a>Perform correlation permutation test using multiple cores</h3>
<p>Note: if the number of possible permutations of your data is less than the desired number of permutations given by the parameter <code>n_perm</code> then an exact test on all of the real permutations will be performed instead of random sampling. For example, if you only have 6 samples (720 possible permutations) but set <code>n_perm = 1000</code> only 720 permutations (i.e. the exact test) will be tested. The <code>permutation_correlation_test</code> function will display a message if this occurs.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw">library</span>(coriell)</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="kw">library</span>(methylKit)</span>
<span id="cb2-3"><a href="#cb2-3"></a></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="co"># define age dataframe -- ages matching column order</span></span>
<span id="cb2-5"><a href="#cb2-5"></a>ages =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">age =</span> <span class="kw">c</span>(<span class="dv">30</span>, <span class="dv">80</span>, <span class="dv">34</span>, <span class="dv">30</span>, <span class="dv">80</span>, <span class="dv">40</span>, <span class="dv">35</span>, <span class="dv">80</span>))</span>
<span id="cb2-6"><a href="#cb2-6"></a></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="co"># simulate a methylation dataset</span></span>
<span id="cb2-8"><a href="#cb2-8"></a>sim_meth &lt;-<span class="st"> </span><span class="kw">dataSim</span>(</span>
<span id="cb2-9"><a href="#cb2-9"></a>  <span class="dt">replicates =</span> <span class="dv">8</span>,</span>
<span id="cb2-10"><a href="#cb2-10"></a>  <span class="dt">sites =</span> <span class="dv">1000</span>,</span>
<span id="cb2-11"><a href="#cb2-11"></a>  <span class="dt">treatment =</span> <span class="kw">c</span>(<span class="kw">rep</span>(<span class="dv">1</span>, <span class="dv">4</span>), <span class="kw">rep</span>(<span class="dv">0</span>, <span class="dv">4</span>)),</span>
<span id="cb2-12"><a href="#cb2-12"></a>  <span class="dt">covariates =</span> ages,</span>
<span id="cb2-13"><a href="#cb2-13"></a>  <span class="dt">sample.ids =</span> <span class="kw">c</span>(<span class="kw">paste0</span>(<span class="st">"test"</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>), <span class="kw">paste0</span>(<span class="st">"ctrl"</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">4</span>))</span>
<span id="cb2-14"><a href="#cb2-14"></a>)</span>
<span id="cb2-15"><a href="#cb2-15"></a></span>
<span id="cb2-16"><a href="#cb2-16"></a><span class="co"># extract the methylation as percentages and coerce to data.frame</span></span>
<span id="cb2-17"><a href="#cb2-17"></a>perc_meth &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">percMethylation</span>(sim_meth))</span>
<span id="cb2-18"><a href="#cb2-18"></a></span>
<span id="cb2-19"><a href="#cb2-19"></a><span class="kw">head</span>(perc_meth)</span>
<span id="cb2-20"><a href="#cb2-20"></a><span class="op">&gt;</span><span class="st">       </span>test1     test2     test3     test4    ctrl1     ctrl2     ctrl3      ctrl4</span>
<span id="cb2-21"><a href="#cb2-21"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>  <span class="fl">48.14815</span>  <span class="fl">54.05405</span>  <span class="fl">45.45455</span>  <span class="fl">57.89474</span> <span class="fl">32.46753</span> <span class="fl">24.418605</span>  <span class="fl">4.687500</span>  <span class="fl">30.769231</span></span>
<span id="cb2-22"><a href="#cb2-22"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>  <span class="fl">0.00000</span>  <span class="fl">0.000000</span>  <span class="fl">0.000000</span>   <span class="fl">0.000000</span></span>
<span id="cb2-23"><a href="#cb2-23"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">3</span>  <span class="fl">38.20598</span>  <span class="fl">16.34615</span>  <span class="fl">12.50000</span>  <span class="fl">60.00000</span> <span class="fl">25.42373</span>  <span class="fl">3.636364</span> <span class="fl">18.421053</span>  <span class="fl">32.335329</span></span>
<span id="cb2-24"><a href="#cb2-24"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">4</span> <span class="fl">100.00000</span> <span class="fl">100.00000</span> <span class="fl">100.00000</span> <span class="fl">100.00000</span> <span class="fl">86.15385</span> <span class="fl">72.340426</span> <span class="fl">86.666667</span> <span class="fl">100.000000</span></span>
<span id="cb2-25"><a href="#cb2-25"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>  <span class="fl">0.00000</span>  <span class="fl">0.000000</span>  <span class="fl">0.000000</span>  <span class="fl">10.526316</span></span>
<span id="cb2-26"><a href="#cb2-26"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">6</span>  <span class="fl">26.98413</span>  <span class="fl">12.63158</span>  <span class="fl">23.07692</span>  <span class="fl">10.00000</span> <span class="fl">22.22222</span>  <span class="fl">0.000000</span>  <span class="fl">3.174603</span>   <span class="fl">4.166667</span></span>
<span id="cb2-27"><a href="#cb2-27"></a></span>
<span id="cb2-28"><a href="#cb2-28"></a><span class="co"># permutation testing -----------------------------------------------------------</span></span>
<span id="cb2-29"><a href="#cb2-29"></a><span class="co"># perform permutation testing using 4 cores and 10000 permutations</span></span>
<span id="cb2-30"><a href="#cb2-30"></a>res &lt;-<span class="st"> </span><span class="kw">permutation_correlation_test</span>(perc_meth, </span>
<span id="cb2-31"><a href="#cb2-31"></a>                                    <span class="dt">y =</span> ages<span class="op">$</span>age, </span>
<span id="cb2-32"><a href="#cb2-32"></a>                                    <span class="dt">n_cores =</span> <span class="dv">4</span>, </span>
<span id="cb2-33"><a href="#cb2-33"></a>                                    <span class="dt">n_perm =</span> <span class="dv">10000</span>,</span>
<span id="cb2-34"><a href="#cb2-34"></a>                                    <span class="dt">cor_method =</span> <span class="st">"spearman"</span>, </span>
<span id="cb2-35"><a href="#cb2-35"></a>                                    <span class="dt">p_adjust_method =</span> <span class="st">"fdr"</span>)</span>
<span id="cb2-36"><a href="#cb2-36"></a></span>
<span id="cb2-37"><a href="#cb2-37"></a><span class="kw">head</span>(res)</span>
<span id="cb2-38"><a href="#cb2-38"></a><span class="op">&gt;</span><span class="st">       </span>test1     test2     test3     test4    ctrl1     ctrl2     ctrl3      ctrl4   spearman empirical_p       fdr</span>
<span id="cb2-39"><a href="#cb2-39"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>  <span class="fl">48.14815</span>  <span class="fl">54.05405</span>  <span class="fl">45.45455</span>  <span class="fl">57.89474</span> <span class="fl">32.46753</span> <span class="fl">24.418605</span>  <span class="fl">4.687500</span>  <span class="fl">30.769231</span> <span class="fl">-0.3437200</span>      <span class="fl">0.1925</span> <span class="fl">0.4812073</span></span>
<span id="cb2-40"><a href="#cb2-40"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>  <span class="fl">0.00000</span>  <span class="fl">0.000000</span>  <span class="fl">0.000000</span>   <span class="fl">0.000000</span>         <span class="ot">NA</span>          <span class="ot">NA</span>        <span class="ot">NA</span></span>
<span id="cb2-41"><a href="#cb2-41"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">3</span>  <span class="fl">38.20598</span>  <span class="fl">16.34615</span>  <span class="fl">12.50000</span>  <span class="fl">60.00000</span> <span class="fl">25.42373</span>  <span class="fl">3.636364</span> <span class="fl">18.421053</span>  <span class="fl">32.335329</span> <span class="fl">-0.3559957</span>      <span class="fl">0.1903</span> <span class="fl">0.4812073</span></span>
<span id="cb2-42"><a href="#cb2-42"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">4</span> <span class="fl">100.00000</span> <span class="fl">100.00000</span> <span class="fl">100.00000</span> <span class="fl">100.00000</span> <span class="fl">86.15385</span> <span class="fl">72.340426</span> <span class="fl">86.666667</span> <span class="fl">100.000000</span> <span class="fl">-0.3093992</span>      <span class="fl">0.2195</span> <span class="fl">0.4812073</span></span>
<span id="cb2-43"><a href="#cb2-43"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>   <span class="fl">0.00000</span>  <span class="fl">0.00000</span>  <span class="fl">0.000000</span>  <span class="fl">0.000000</span>  <span class="fl">10.526316</span>  <span class="fl">0.4252433</span>      <span class="fl">0.3769</span> <span class="fl">0.4812073</span></span>
<span id="cb2-44"><a href="#cb2-44"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">6</span>  <span class="fl">26.98413</span>  <span class="fl">12.63158</span>  <span class="fl">23.07692</span>  <span class="fl">10.00000</span> <span class="fl">22.22222</span>  <span class="fl">0.000000</span>  <span class="fl">3.174603</span>   <span class="fl">4.166667</span> <span class="fl">-0.2946172</span>      <span class="fl">0.2334</span> <span class="fl">0.4812073</span></span></code></pre></div>
</div>
<div id="extract-random-correlations-from-a-dataframe" class="section level3">
<h3 class="hasAnchor">
<a href="#extract-random-correlations-from-a-dataframe" class="anchor"></a>Extract random correlations from a data.frame</h3>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="co"># using the same perc_meth data.frame and ages as defined above</span>
<span class="co"># get 1,000,000 random correlations from the dataset</span>
<span class="va">cors</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/sample_n_random_cor.html">sample_n_random_cor</a></span><span class="op">(</span>df <span class="op">=</span> <span class="va">perc_meth</span>, 
                            y <span class="op">=</span> <span class="va">ages</span><span class="op">$</span><span class="va">age</span>,
                            n <span class="op">=</span> <span class="fl">1000000</span>,
                            cor_method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span>

<span class="co"># simple histogram of correlation values -- drop NAs if present</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">cors</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">cors</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></pre></div>
</div>
<div id="convert-edger-results-into-a-tidy-dataframe" class="section level3">
<h3 class="hasAnchor">
<a href="#convert-edger-results-into-a-tidy-dataframe" class="anchor"></a>Convert edgeR results into a tidy dataframe</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw">library</span>(edgeR)</span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="kw">library</span>(coriell)</span>
<span id="cb4-3"><a href="#cb4-3"></a></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co"># simulate expression data using coriell::simulate_counts()</span></span>
<span id="cb4-5"><a href="#cb4-5"></a>x &lt;-<span class="st"> </span><span class="kw">simulate_counts</span>()</span>
<span id="cb4-6"><a href="#cb4-6"></a></span>
<span id="cb4-7"><a href="#cb4-7"></a><span class="co"># run edger pipeline</span></span>
<span id="cb4-8"><a href="#cb4-8"></a>group &lt;-<span class="st">  </span><span class="kw">factor</span>(<span class="kw">rep</span>(<span class="kw">c</span>(<span class="st">"ctl"</span>, <span class="st">"trt"</span>), <span class="dt">each =</span> <span class="dv">3</span>))</span>
<span id="cb4-9"><a href="#cb4-9"></a>y &lt;-<span class="st"> </span><span class="kw">DGEList</span>(<span class="dt">counts =</span> x<span class="op">$</span>table, <span class="dt">group =</span> group)</span>
<span id="cb4-10"><a href="#cb4-10"></a>y &lt;-<span class="st"> </span><span class="kw">calcNormFactors</span>(y)</span>
<span id="cb4-11"><a href="#cb4-11"></a>design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(<span class="op">~</span>group)</span>
<span id="cb4-12"><a href="#cb4-12"></a>y &lt;-<span class="st"> </span><span class="kw">estimateDisp</span>(y, design)</span>
<span id="cb4-13"><a href="#cb4-13"></a></span>
<span id="cb4-14"><a href="#cb4-14"></a><span class="co"># To perform quasi-likelihood F-tests</span></span>
<span id="cb4-15"><a href="#cb4-15"></a>fit &lt;-<span class="st"> </span><span class="kw">glmQLFit</span>(y, design)</span>
<span id="cb4-16"><a href="#cb4-16"></a>qlf &lt;-<span class="st"> </span><span class="kw">glmQLFTest</span>(fit, <span class="dt">coef =</span> <span class="dv">2</span>)</span>
<span id="cb4-17"><a href="#cb4-17"></a></span>
<span id="cb4-18"><a href="#cb4-18"></a><span class="co"># -----------------------------------------------------------------------------</span></span>
<span id="cb4-19"><a href="#cb4-19"></a><span class="co"># convert to tidy dataframe</span></span>
<span id="cb4-20"><a href="#cb4-20"></a>res_df &lt;-<span class="st"> </span><span class="kw">edger_to_df</span>(qlf, <span class="dt">fdr =</span> <span class="dv">1</span>, <span class="dt">lfc =</span> <span class="dv">0</span>)</span>
<span id="cb4-21"><a href="#cb4-21"></a></span>
<span id="cb4-22"><a href="#cb4-22"></a><span class="kw">head</span>(res_df)</span>
<span id="cb4-23"><a href="#cb4-23"></a><span class="op">&gt;</span><span class="st"> </span><span class="co"># A tibble: 6 x 6</span></span>
<span id="cb4-24"><a href="#cb4-24"></a><span class="er">&gt;</span><span class="st">   </span>feature_id logFC unshrunk.logFC logCPM   PValue        FDR</span>
<span id="cb4-25"><a href="#cb4-25"></a><span class="op">&gt;</span><span class="st">   </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">      </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">          </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">  </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">    </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">      </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span></span>
<span id="cb4-26"><a href="#cb4-26"></a><span class="er">&gt;</span><span class="st"> </span><span class="dv">1</span> gene<span class="fl">.51</span>    <span class="fl">-4.43</span>          <span class="fl">-4.43</span>   <span class="fl">6.37</span> <span class="fl">2.93e-10</span> <span class="fl">0.00000561</span></span>
<span id="cb4-27"><a href="#cb4-27"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span> gene<span class="fl">.26</span>    <span class="fl">-6.63</span>          <span class="fl">-6.68</span>   <span class="fl">3.45</span> <span class="fl">8.37e-9</span>   <span class="fl">0.0000801</span> </span>
<span id="cb4-28"><a href="#cb4-28"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">3</span> gene<span class="fl">.19</span>    <span class="fl">-7.01</span>          <span class="fl">-7.05</span>   <span class="fl">3.41</span> <span class="fl">1.51e-8</span>   <span class="fl">0.0000962</span> </span>
<span id="cb4-29"><a href="#cb4-29"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">4</span> gene<span class="fl">.100</span>   <span class="fl">-4.62</span>          <span class="fl">-4.63</span>   <span class="fl">4.40</span> <span class="fl">3.24e-8</span>    <span class="fl">0.000132</span>  </span>
<span id="cb4-30"><a href="#cb4-30"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span> gene<span class="fl">.74</span>    <span class="fl">-6.05</span>          <span class="fl">-6.07</span>   <span class="fl">7.04</span> <span class="fl">4.06e-8</span>    <span class="fl">0.000132</span>  </span>
<span id="cb4-31"><a href="#cb4-31"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">6</span> gene<span class="fl">.53</span>    <span class="fl">-4.67</span>          <span class="fl">-4.68</span>   <span class="fl">3.91</span> <span class="fl">4.14e-8</span>    <span class="fl">0.000132</span> </span></code></pre></div>
</div>
<div id="summarize-results-from-differential-expression-test" class="section level3">
<h3 class="hasAnchor">
<a href="#summarize-results-from-differential-expression-test" class="anchor"></a>Summarize results from differential expression test</h3>
<p>Return a table of up/down/non-de counts and their percentages.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co"># for edgeR results default values can be used</span></span>
<span id="cb5-2"><a href="#cb5-2"></a>res_df &lt;-<span class="st"> </span><span class="kw">edger_to_df</span>(qlf)</span>
<span id="cb5-3"><a href="#cb5-3"></a></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="kw">summarize_dge</span>(res_df, <span class="dt">fdr =</span> <span class="fl">0.05</span>)</span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="op">&gt;</span><span class="st"> </span><span class="co"># A tibble: 3 x 3</span></span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="er">&gt;</span><span class="st">   </span>dge         n   perc</span>
<span id="cb5-7"><a href="#cb5-7"></a><span class="op">&gt;</span><span class="st">   </span><span class="er">&lt;</span>fct<span class="op">&gt;</span><span class="st">   </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st">  </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span></span>
<span id="cb5-8"><a href="#cb5-8"></a><span class="er">&gt;</span><span class="st"> </span><span class="dv">1</span> up        <span class="dv">162</span>  <span class="fl">0.847</span></span>
<span id="cb5-9"><a href="#cb5-9"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span> down      <span class="dv">266</span>   <span class="fl">1.39</span> </span>
<span id="cb5-10"><a href="#cb5-10"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">3</span> non<span class="op">-</span>dge <span class="dv">18709</span>   <span class="fl">97.8</span>  </span>
<span id="cb5-11"><a href="#cb5-11"></a></span>
<span id="cb5-12"><a href="#cb5-12"></a><span class="co"># For DESeq2 results you must specify the column names</span></span>
<span id="cb5-13"><a href="#cb5-13"></a><span class="kw">summarize_dge</span>(deseq_res_df, </span>
<span id="cb5-14"><a href="#cb5-14"></a>              <span class="dt">fdr_col =</span> padj,</span>
<span id="cb5-15"><a href="#cb5-15"></a>              <span class="dt">lfc_col =</span> log2FoldChange,</span>
<span id="cb5-16"><a href="#cb5-16"></a>              <span class="dt">fdr =</span> <span class="fl">0.05</span>)</span></code></pre></div>
</div>
<div id="create-volcano-plot-from-differential-expression-results" class="section level3">
<h3 class="hasAnchor">
<a href="#create-volcano-plot-from-differential-expression-results" class="anchor"></a>Create volcano plot from differential expression results</h3>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="co"># For edgeR results default values can be used</span>
<span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">res_df</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Treatment vs Control"</span><span class="op">)</span>

<span class="co"># For DESeq2 results you must specify the column names</span>
<span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">deseq_res_df</span>,
             x <span class="op">=</span> <span class="va">log2FoldChange</span>,
             y <span class="op">=</span> <span class="va">padj</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/volcano1.png"></p>
<p>Different significance levels can be used to filter the plotted points. For example, significance levels can be set by specifying the <code>fdr</code> and <code>lfc</code> values.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">res_df</span>, fdr <span class="op">=</span> <span class="fl">0.01</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span>  <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Treatment vs Control"</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/volcano2.png"></p>
<p>Labels for the counts will be displayed by default. To remove them set <code>annotate_counts = FALSE</code></p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">res_df</span>, 
             fdr <span class="op">=</span> <span class="fl">0.01</span>, 
             lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>, 
             annotate_counts <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>  <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Treatment vs Control"</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/volcano3.png"></p>
<p>Positions of the count labels can be adjusted by setting the <code>xmax_label_offset</code>, <code>xmin_label_offset</code> and <code>ymax_label_offset</code> values. Setting the values closer to 1 moves the labels away from the origin. <code>xmax_label_offset</code> controls the ‘up’ gene label whereas <code>xmin_label_offset</code> controls the ‘down’ genes label. <code>ymax_label_offset</code> controls the vertical position of the labels.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">res_df</span>,
             fdr <span class="op">=</span> <span class="fl">0.01</span>,
             lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,
             xmax_label_offset <span class="op">=</span> <span class="fl">0.2</span>, 
             xmin_label_offset <span class="op">=</span> <span class="fl">0.7</span>, 
             ymax_label_offset <span class="op">=</span> <span class="fl">0.9</span><span class="op">)</span>  <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Treatment vs Control"</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/volcano4.png"></p>
<p>Text labels can also be added for the DE genes by setting <code>label_sig = TRUE</code>. Caution, if there are many DE genes this will be overplotted</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">res_df</span>,
             fdr <span class="op">=</span> <span class="fl">1e-4</span>,
             lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,
             annotate_counts <span class="op">=</span> <span class="cn">FALSE</span>,
             label_sig <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>  <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Treatment vs Control"</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/volcano5.png"></p>
</div>
<div id="create-md-plot-from-differential-expression-results" class="section level3">
<h3 class="hasAnchor">
<a href="#create-md-plot-from-differential-expression-results" class="anchor"></a>Create md plot from differential expression results</h3>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="co"># For edgeR results default values can be used</span>
<span class="fu"><a href="reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">res_df</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Treatment vs Control"</span><span class="op">)</span>

<span class="co"># For DESeq2 results you must specify the column names</span>
<span class="fu"><a href="reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">deseq_res_df</span>,
        x <span class="op">=</span> <span class="va">baseMean</span>,
        y <span class="op">=</span> <span class="va">log2FoldChange</span>,
        sig_col <span class="op">=</span> <span class="va">padj</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/md1.png"></p>
<p>Different significance levels can be used to filter the plotted points. For example, significance levels can be set by specifying the <code>fdr</code> and <code>lfc</code> values.</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="fu"><a href="reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">res_df</span>, 
        fdr <span class="op">=</span> <span class="fl">0.01</span>, 
        lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Treatment vs Control"</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/md2.png"></p>
<p>Labels for the counts will be displayed by default. To remove them set <code>annotate_counts = FALSE</code></p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="fu"><a href="reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">res_df</span>, 
        fdr <span class="op">=</span> <span class="fl">0.01</span>, 
        lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,
        annotate_counts <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Treatment vs Control"</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/md3.png"></p>
<p>Positions of the count labels can be adjusted by setting the <code>xmax_label_offset</code>, <code>ymin_label_offset</code> and <code>ymax_label_offset</code> values. Setting the values closer to 1 moves the labels away from the origin. <code>xmax_label_offset</code> controls the horizontal position of the labels. <code>ymin_label_offset</code> controls the ‘down’ gene label. <code>ymax_label_offset</code> controls the ‘up’ genes label.</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="fu"><a href="reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">res_df</span>, 
        fdr <span class="op">=</span> <span class="fl">0.01</span>,
        lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span>,
        xmax_label_offset <span class="op">=</span> <span class="fl">0.6</span>, 
        ymin_label_offset <span class="op">=</span> <span class="fl">0.25</span>, 
        ymax_label_offset <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Treatment vs Control"</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/md4.png"></p>
</div>
<div id="heatmap-with-sensible-defaults" class="section level3">
<h3 class="hasAnchor">
<a href="#heatmap-with-sensible-defaults" class="anchor"></a>Heatmap with sensible defaults</h3>
<p>We often use the same settings when making calls to <code>pheatmap</code>. This function is a wrapper around <code>pheatmap</code> which uses sensible default values for expression data. It changes the default color scale to a diverging blue to white to red scale, modifies the clustering parameters (row-wise euclidean, col-wise correlation) and clustering method (complete), angles the column labels, removes border colors and rownames.</p>
<p>Any of these options can be overridden by simply supplying the arguments to <code>quickmap</code> as you would <code>pheatmap</code>. This also allows for additional arguments to be passed to the <code>quickmap</code> function for creating row and column annotations.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="co"># generate some example data and log-scale it</span>
<span class="va">lcpms</span> <span class="op">&lt;-</span> <span class="fu">coriell</span><span class="fu">::</span><span class="fu"><a href="reference/simulate_counts.html">simulate_counts</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span><span class="op">$</span><span class="va">table</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log1p</a></span><span class="op">(</span><span class="op">)</span>

<span class="co"># plot a heatmap of the logCPM values</span>
<span class="fu"><a href="reference/quickmap.html">quickmap</a></span><span class="op">(</span><span class="va">lcpms</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/quickmap1.png"></p>
<p>Other <code>pheatmap</code> arguments can be passed to the <code>quickmap</code> function as well.</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="co"># create annotation for columns</span>
<span class="va">col_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>treatment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ctl"</span>, <span class="st">"trt"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">col_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">lcpms</span><span class="op">)</span>

<span class="co"># create color scheme for treatment conditions</span>
<span class="va">ann_colors</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>treatment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ctl"</span> <span class="op">=</span> <span class="st">"steelblue"</span>, <span class="st">"trt"</span> <span class="op">=</span> <span class="st">"firebrick"</span><span class="op">)</span><span class="op">)</span>

<span class="co"># plot the heatmap, passing additional args to pheatmap</span>
<span class="fu"><a href="reference/quickmap.html">quickmap</a></span><span class="op">(</span><span class="va">lcpms</span>,
         annotation_col <span class="op">=</span> <span class="va">col_df</span>,
         annotation_colors <span class="op">=</span> <span class="va">ann_colors</span>,
         main <span class="op">=</span> <span class="st">"Treatment vs Control"</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/quickmap2.png"></p>
<p>There are two default color scales included which can be specified by setting the <code>diverging_palette</code> argument. By default <code>diverging_palette = TRUE</code> which sets the color scale the same as the above heatmaps. This is useful for scaled data. However, if you are plotting unscaled data such as normalized expression values then a continuous color palette is more appropriate. Setting <code>diverging_palette = FALSE</code> will set the color palette to a continuous (<code><a href="https://rdrr.io/pkg/viridisLite/man/viridis.html">viridis::magma(50)</a></code>) palette.</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="co"># NOTE: different lcpms data than above</span>
<span class="fu"><a href="reference/quickmap.html">quickmap</a></span><span class="op">(</span><span class="va">lcpms</span>, diverging_palette <span class="op">=</span> <span class="cn">FALSE</span>, scale <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></pre></div>
<p><img src="reference/figures/quickmap3.png"></p>
</div>
<div id="z-score-a-dataframe" class="section level3">
<h3 class="hasAnchor">
<a href="#z-score-a-dataframe" class="anchor"></a>Z-score a dataframe</h3>
<p>Z-score a dataframe by row or column</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="co"># create some example data</span></span>
<span id="cb18-2"><a href="#cb18-2"></a>cpms &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">a =</span> <span class="kw">runif</span>(<span class="dv">100</span>, <span class="dt">min =</span> <span class="dv">0</span>, <span class="dt">max =</span> <span class="dv">100</span>),</span>
<span id="cb18-3"><a href="#cb18-3"></a>                   <span class="dt">b =</span> <span class="kw">runif</span>(<span class="dv">100</span>, <span class="dt">min =</span> <span class="dv">0</span>, <span class="dt">max =</span> <span class="dv">100</span>),</span>
<span id="cb18-4"><a href="#cb18-4"></a>                   <span class="dt">c =</span> <span class="kw">runif</span>(<span class="dv">100</span>, <span class="dt">min =</span> <span class="dv">0</span>, <span class="dt">max =</span> <span class="dv">100</span>),</span>
<span id="cb18-5"><a href="#cb18-5"></a>                   <span class="dt">d =</span> <span class="kw">runif</span>(<span class="dv">100</span>, <span class="dt">min =</span> <span class="dv">0</span>, <span class="dt">max =</span> <span class="dv">100</span>))</span>
<span id="cb18-6"><a href="#cb18-6"></a></span>
<span id="cb18-7"><a href="#cb18-7"></a><span class="op">&gt;</span><span class="st"> </span><span class="kw">head</span>(cpms)</span>
<span id="cb18-8"><a href="#cb18-8"></a><span class="op">&gt;</span><span class="st">           </span>a        b        c         d</span>
<span id="cb18-9"><a href="#cb18-9"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span> <span class="fl">93.586737</span> <span class="fl">41.79316</span> <span class="fl">58.59588</span> <span class="fl">73.082215</span></span>
<span id="cb18-10"><a href="#cb18-10"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span> <span class="fl">70.009822</span> <span class="fl">25.84383</span> <span class="fl">57.03569</span> <span class="fl">40.512135</span></span>
<span id="cb18-11"><a href="#cb18-11"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">3</span> <span class="fl">60.053908</span> <span class="fl">14.68176</span> <span class="fl">82.66302</span> <span class="fl">60.842900</span></span>
<span id="cb18-12"><a href="#cb18-12"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">4</span> <span class="fl">95.711441</span> <span class="fl">86.76281</span> <span class="fl">58.07523</span> <span class="fl">22.571323</span></span>
<span id="cb18-13"><a href="#cb18-13"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span>  <span class="fl">2.833631</span> <span class="fl">25.04612</span> <span class="fl">72.85270</span>  <span class="fl">5.417795</span></span>
<span id="cb18-14"><a href="#cb18-14"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">6</span> <span class="fl">24.596068</span> <span class="fl">85.46398</span> <span class="fl">16.12987</span> <span class="fl">33.810050</span></span>
<span id="cb18-15"><a href="#cb18-15"></a></span>
<span id="cb18-16"><a href="#cb18-16"></a><span class="co"># scale the data by rows</span></span>
<span id="cb18-17"><a href="#cb18-17"></a>cpms_st &lt;-<span class="st"> </span><span class="kw">zscore_df</span>(cpms)</span>
<span id="cb18-18"><a href="#cb18-18"></a></span>
<span id="cb18-19"><a href="#cb18-19"></a><span class="op">&gt;</span><span class="st"> </span><span class="kw">head</span>(cpms_st)</span>
<span id="cb18-20"><a href="#cb18-20"></a><span class="op">&gt;</span><span class="st">            </span>a           b          c          d</span>
<span id="cb18-21"><a href="#cb18-21"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>  <span class="fl">1.2201846</span> <span class="fl">-1.13598436</span> <span class="fl">-0.3716029</span>  <span class="fl">0.2874026</span></span>
<span id="cb18-22"><a href="#cb18-22"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>  <span class="fl">1.1247307</span> <span class="fl">-1.16871823</span>  <span class="fl">0.4510108</span> <span class="fl">-0.4070232</span></span>
<span id="cb18-23"><a href="#cb18-23"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">3</span>  <span class="fl">0.1922442</span> <span class="fl">-1.39554374</span>  <span class="fl">0.9834448</span>  <span class="fl">0.2198548</span></span>
<span id="cb18-24"><a href="#cb18-24"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">4</span>  <span class="fl">0.9076295</span>  <span class="fl">0.63627273</span> <span class="fl">-0.2336442</span> <span class="fl">-1.3102580</span></span>
<span id="cb18-25"><a href="#cb18-25"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span> <span class="fl">-0.7309118</span> <span class="fl">-0.04598875</span>  <span class="fl">1.4281295</span> <span class="fl">-0.6512290</span></span>
<span id="cb18-26"><a href="#cb18-26"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">6</span> <span class="fl">-0.4943903</span>  <span class="fl">1.45917080</span> <span class="fl">-0.7661138</span> <span class="fl">-0.1986667</span></span>
<span id="cb18-27"><a href="#cb18-27"></a></span>
<span id="cb18-28"><a href="#cb18-28"></a></span>
<span id="cb18-29"><a href="#cb18-29"></a><span class="co"># default is to scale by row, scaling by columns can also be performed by</span></span>
<span id="cb18-30"><a href="#cb18-30"></a><span class="co"># setting the by = "column"</span></span>
<span id="cb18-31"><a href="#cb18-31"></a>cpms_st_by_col &lt;-<span class="st"> </span><span class="kw">zscore_df</span>(cpms, <span class="dt">by =</span> <span class="st">"column"</span>)</span></code></pre></div>
</div>
<div id="convert-a-list-of-sets-into-a-binary-matrix" class="section level3">
<h3 class="hasAnchor">
<a href="#convert-a-list-of-sets-into-a-binary-matrix" class="anchor"></a>Convert a list of sets into a binary matrix</h3>
<p>This function is useful for comparing if a given gene is present across all or a certain proportion of conditions.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>sets &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">"set1"</span> =<span class="st"> </span>letters[<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>],</span>
<span id="cb19-2"><a href="#cb19-2"></a>             <span class="st">"set2"</span> =<span class="st"> </span>letters[<span class="dv">3</span><span class="op">:</span><span class="dv">8</span>],</span>
<span id="cb19-3"><a href="#cb19-3"></a>             <span class="st">"set3"</span> =<span class="st"> </span>letters[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>],</span>
<span id="cb19-4"><a href="#cb19-4"></a>             <span class="st">"set4"</span> =<span class="st"> </span>letters[<span class="dv">4</span><span class="op">:</span><span class="dv">6</span>])</span>
<span id="cb19-5"><a href="#cb19-5"></a></span>
<span id="cb19-6"><a href="#cb19-6"></a><span class="co"># convert the sets to a binary matrix</span></span>
<span id="cb19-7"><a href="#cb19-7"></a>mat &lt;-<span class="st"> </span><span class="kw">list_to_matrix</span>(sets)</span>
<span id="cb19-8"><a href="#cb19-8"></a></span>
<span id="cb19-9"><a href="#cb19-9"></a>mat</span>
<span id="cb19-10"><a href="#cb19-10"></a><span class="op">&gt;</span><span class="st">   </span>set1 set2 set3 set4</span>
<span id="cb19-11"><a href="#cb19-11"></a><span class="op">&gt;</span><span class="st"> </span>a    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">0</span></span>
<span id="cb19-12"><a href="#cb19-12"></a><span class="op">&gt;</span><span class="st"> </span>b    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">0</span></span>
<span id="cb19-13"><a href="#cb19-13"></a><span class="op">&gt;</span><span class="st"> </span>c    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">0</span></span>
<span id="cb19-14"><a href="#cb19-14"></a><span class="op">&gt;</span><span class="st"> </span>d    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">1</span></span>
<span id="cb19-15"><a href="#cb19-15"></a><span class="op">&gt;</span><span class="st"> </span>e    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">1</span>    <span class="dv">1</span></span>
<span id="cb19-16"><a href="#cb19-16"></a><span class="op">&gt;</span><span class="st"> </span>f    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">1</span></span>
<span id="cb19-17"><a href="#cb19-17"></a><span class="op">&gt;</span><span class="st"> </span>g    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">0</span></span>
<span id="cb19-18"><a href="#cb19-18"></a><span class="op">&gt;</span><span class="st"> </span>h    <span class="dv">0</span>    <span class="dv">1</span>    <span class="dv">0</span>    <span class="dv">0</span></span></code></pre></div>
</div>
<div id="get-statistics-for-all-pairwise-combinations-of-a-list-of-sets" class="section level3">
<h3 class="hasAnchor">
<a href="#get-statistics-for-all-pairwise-combinations-of-a-list-of-sets" class="anchor"></a>Get statistics for all pairwise combinations of a list of sets</h3>
<p>Compare every set to every other set and return statistics about their intersections.</p>
<p>Statistics are returned in a list object. The returned list contains a named vector of the statistic computed. The names of the vectore indicate the pairwise comparison, i.e. “Set A : Set B” for all combinations of sets. Use <code>?pairwise_intersection_stats()</code> for more information about the statistics computed.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>sets &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">"set1"</span> =<span class="st"> </span>letters[<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>],</span>
<span id="cb20-2"><a href="#cb20-2"></a>             <span class="st">"set2"</span> =<span class="st"> </span>letters[<span class="dv">3</span><span class="op">:</span><span class="dv">8</span>],</span>
<span id="cb20-3"><a href="#cb20-3"></a>             <span class="st">"set3"</span> =<span class="st"> </span>letters[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>],</span>
<span id="cb20-4"><a href="#cb20-4"></a>             <span class="st">"set4"</span> =<span class="st"> </span>letters[<span class="dv">4</span><span class="op">:</span><span class="dv">6</span>])</span>
<span id="cb20-5"><a href="#cb20-5"></a></span>
<span id="cb20-6"><a href="#cb20-6"></a><span class="co"># get the intersection stats -- returns a list of statistics</span></span>
<span id="cb20-7"><a href="#cb20-7"></a>stats &lt;-<span class="st"> </span><span class="kw">pairwise_intersection_stats</span>(sets)</span>
<span id="cb20-8"><a href="#cb20-8"></a></span>
<span id="cb20-9"><a href="#cb20-9"></a><span class="co"># individual statistics can be accessed with subsetting the list</span></span>
<span id="cb20-10"><a href="#cb20-10"></a><span class="co"># to get the intersection sizes of every combination of sets:</span></span>
<span id="cb20-11"><a href="#cb20-11"></a>stats<span class="op">$</span>intersection_size</span>
<span id="cb20-12"><a href="#cb20-12"></a></span>
<span id="cb20-13"><a href="#cb20-13"></a><span class="op">&gt;</span><span class="st"> </span>set1 <span class="op">:</span><span class="st"> </span>set2 set1 <span class="op">:</span><span class="st"> </span>set3 set1 <span class="op">:</span><span class="st"> </span>set4 set2 <span class="op">:</span><span class="st"> </span>set3 set2 <span class="op">:</span><span class="st"> </span>set4 set3 <span class="op">:</span><span class="st"> </span>set4 </span>
<span id="cb20-14"><a href="#cb20-14"></a><span class="op">&gt;</span><span class="st">           </span><span class="dv">2</span>           <span class="dv">4</span>           <span class="dv">1</span>           <span class="dv">3</span>           <span class="dv">3</span>           <span class="dv">2</span></span></code></pre></div>
</div>
<div id="perform-gene-ontology-analysis-with-panther" class="section level3">
<h3 class="hasAnchor">
<a href="#perform-gene-ontology-analysis-with-panther" class="anchor"></a>Perform Gene Ontology Analysis with PANTHER</h3>
<p>This function sends a request to the PANTHER REST API for GO over-representation analysis on the user specified vector of genes. The API should accept Ensembl gene identifiers, Ensembl protein identifiers, Ensembl transcript identifiers, Entrez gene ids, gene symbols, NCBI GIs, HGNC Ids, International protein index ids, NCBI UniGene ids, UniProt accessions and UniProt ids. The function will return a tibble of the GO results for the user specified organism, annotation dataset, test type and test correction method.</p>
<p>NOTE: the <code>organism</code> parameter is given by the Entrez Taxon ID. Typical values for this parameter are “9606” for HUMAN, “10090” for MOUSE, “10116” for RAT. Any other Taxon ID could also be used. see <code><a href="reference/panther_go.html">?coriell::panther_go</a></code> for more information.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a>genes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"CTNNB1"</span>, <span class="st">"ADAM17"</span>, <span class="st">"AXIN1"</span>, <span class="st">"AXIN2"</span>, <span class="st">"CCND2"</span>, <span class="st">"CSNK1E"</span>, <span class="st">"CTNNB1"</span>, </span>
<span id="cb21-2"><a href="#cb21-2"></a>           <span class="st">"CUL1"</span>, <span class="st">"DKK1"</span>, <span class="st">"DKK4"</span>, <span class="st">"DLL1"</span>, <span class="st">"DVL2"</span>, <span class="st">"FRAT1"</span>, <span class="st">"FZD1"</span>, <span class="st">"FZD8"</span>, </span>
<span id="cb21-3"><a href="#cb21-3"></a>           <span class="st">"GNAI1"</span>, <span class="st">"HDAC11"</span>, <span class="st">"HDAC2"</span>, <span class="st">"HDAC5"</span>, <span class="st">"HEY1"</span>, <span class="st">"HEY2"</span>, <span class="st">"JAG1"</span>, </span>
<span id="cb21-4"><a href="#cb21-4"></a>           <span class="st">"JAG2"</span>, <span class="st">"KAT2A"</span>, <span class="st">"LEF1"</span>, <span class="st">"MAML1"</span>, <span class="st">"MYC"</span>, <span class="st">"NCOR2"</span>, <span class="st">"NCSTN"</span>, </span>
<span id="cb21-5"><a href="#cb21-5"></a>           <span class="st">"NKD1"</span>, <span class="st">"NOTCH1"</span>, <span class="st">"NOTCH4"</span>, <span class="st">"NUMB"</span>, <span class="st">"PPARD"</span>, <span class="st">"PSEN2"</span>, <span class="st">"PTCH1"</span>, </span>
<span id="cb21-6"><a href="#cb21-6"></a>           <span class="st">"RBPJ"</span>, <span class="st">"SKP2"</span>, <span class="st">"TCF7"</span>, <span class="st">"TP53"</span>, <span class="st">"WNT1"</span>, <span class="st">"WNT5B"</span>, <span class="st">"WNT6"</span>)</span>
<span id="cb21-7"><a href="#cb21-7"></a></span>
<span id="cb21-8"><a href="#cb21-8"></a>go_results &lt;-<span class="st"> </span>coriell<span class="op">::</span><span class="kw">panther_go</span>(</span>
<span id="cb21-9"><a href="#cb21-9"></a>                                genes, </span>
<span id="cb21-10"><a href="#cb21-10"></a>                                <span class="dt">organism =</span> <span class="st">"9606"</span>, </span>
<span id="cb21-11"><a href="#cb21-11"></a>                                <span class="dt">annot_dataset =</span> <span class="st">"biological_process"</span>)</span>
<span id="cb21-12"><a href="#cb21-12"></a></span>
<span id="cb21-13"><a href="#cb21-13"></a><span class="kw">head</span>(go_results, <span class="dt">n =</span> <span class="dv">10</span>)</span>
<span id="cb21-14"><a href="#cb21-14"></a><span class="op">&gt;</span><span class="st"> </span>result_number number_in_list fold_enrichment      fdr expected number_in_reference   pValue plus_minus GO_term    description                            </span>
<span id="cb21-15"><a href="#cb21-15"></a><span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">                  </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st">           </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">    </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">    </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">               </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st">    </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">      </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">      </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">                                  </span></span>
<span id="cb21-16"><a href="#cb21-16"></a><span class="er">&gt;</span><span class="st"> </span><span class="dv">1</span>                         <span class="dv">32</span>            <span class="fl">6.29</span> <span class="fl">3.63e-17</span>    <span class="fl">5.09</span>                 <span class="dv">2525</span> <span class="fl">2.28e-21</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0007166</span> cell surface receptor signaling pathway</span>
<span id="cb21-17"><a href="#cb21-17"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>                         <span class="dv">24</span>           <span class="fl">10.6</span>  <span class="fl">5.46e-16</span>    <span class="fl">2.27</span>                 <span class="dv">1129</span> <span class="fl">6.87e-20</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0060429</span> epithelium development                 </span>
<span id="cb21-18"><a href="#cb21-18"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">3</span>                         <span class="dv">17</span>           <span class="fl">24.3</span>  <span class="fl">8.23e-16</span>    <span class="fl">0.701</span>                 <span class="dv">348</span> <span class="fl">1.55e-19</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0198738</span> cell<span class="op">-</span>cell signaling by wnt             </span>
<span id="cb21-19"><a href="#cb21-19"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">4</span>                         <span class="dv">17</span>           <span class="fl">24.3</span>  <span class="fl">6.18e-16</span>    <span class="fl">0.701</span>                 <span class="dv">348</span> <span class="fl">1.55e-19</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0016055</span> Wnt signaling pathway                  </span>
<span id="cb21-20"><a href="#cb21-20"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span>                         <span class="dv">34</span>            <span class="fl">4.72</span> <span class="fl">9.14e-16</span>    <span class="fl">7.20</span>                 <span class="dv">3576</span> <span class="fl">2.87e-19</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0010646</span> regulation of cell communication       </span>
<span id="cb21-21"><a href="#cb21-21"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">6</span>                         <span class="dv">33</span>            <span class="fl">5.00</span> <span class="fl">9.16e-16</span>    <span class="fl">6.60</span>                 <span class="dv">3277</span> <span class="fl">3.46e-19</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0048513</span> animal organ development               </span>
<span id="cb21-22"><a href="#cb21-22"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">7</span>                         <span class="dv">19</span>           <span class="fl">17.0</span>  <span class="fl">9.19e-16</span>    <span class="fl">1.12</span>                  <span class="dv">555</span> <span class="fl">4.04e-19</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0048729</span> tissue morphogenesis                   </span>
<span id="cb21-23"><a href="#cb21-23"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">8</span>                         <span class="dv">34</span>            <span class="fl">4.67</span> <span class="fl">8.11e-16</span>    <span class="fl">7.28</span>                 <span class="dv">3615</span> <span class="fl">4.08e-19</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0023051</span> regulation of signaling                </span>
<span id="cb21-24"><a href="#cb21-24"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">9</span>                         <span class="dv">12</span>           <span class="fl">70.9</span>  <span class="fl">7.58e-16</span>    <span class="fl">0.169</span>                  <span class="dv">84</span> <span class="fl">4.29e-19</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0060070</span> canonical Wnt signaling pathway        </span>
<span id="cb21-25"><a href="#cb21-25"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span>                        <span class="dv">13</span>           <span class="fl">52.5</span>  <span class="fl">7.24e-16</span>    <span class="fl">0.248</span>                 <span class="dv">123</span> <span class="fl">4.55e-19</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0007219</span> Notch signaling pathway</span>
<span id="cb21-26"><a href="#cb21-26"></a></span>
<span id="cb21-27"><a href="#cb21-27"></a></span>
<span id="cb21-28"><a href="#cb21-28"></a><span class="co"># using ensembl ids ------------------------------------------------------------</span></span>
<span id="cb21-29"><a href="#cb21-29"></a><span class="co"># all unique ensembl IDs for the above gene set</span></span>
<span id="cb21-30"><a href="#cb21-30"></a>ensembl_ids &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"ENSG00000162736"</span>, <span class="st">"ENSG00000143801"</span>, <span class="st">"ENSG00000177283"</span>, </span>
<span id="cb21-31"><a href="#cb21-31"></a>                 <span class="st">"ENSG00000107984"</span>, <span class="st">"ENSG00000165879"</span>, <span class="st">"ENSG00000111186"</span>, </span>
<span id="cb21-32"><a href="#cb21-32"></a>                 <span class="st">"ENSG00000118971"</span>, <span class="st">"ENSG00000125084"</span>, <span class="st">"ENSG00000196498"</span>, </span>
<span id="cb21-33"><a href="#cb21-33"></a>                 <span class="st">"ENSG00000133961"</span>, <span class="st">"ENSG00000103126"</span>, <span class="st">"ENSG00000004975"</span>, </span>
<span id="cb21-34"><a href="#cb21-34"></a>                 <span class="st">"ENSG00000141510"</span>, <span class="st">"ENSG00000108840"</span>, <span class="st">"ENSG00000168646"</span>, </span>
<span id="cb21-35"><a href="#cb21-35"></a>                 <span class="st">"ENSG00000151694"</span>, <span class="st">"ENSG00000115596"</span>, <span class="st">"ENSG00000101384"</span>, </span>
<span id="cb21-36"><a href="#cb21-36"></a>                 <span class="st">"ENSG00000213923"</span>, <span class="st">"ENSG00000163517"</span>, <span class="st">"ENSG00000168036"</span>, </span>
<span id="cb21-37"><a href="#cb21-37"></a>                 <span class="st">"ENSG00000138795"</span>, <span class="st">"ENSG00000145604"</span>, <span class="st">"ENSG00000081059"</span>, </span>
<span id="cb21-38"><a href="#cb21-38"></a>                 <span class="st">"ENSG00000161021"</span>, <span class="st">"ENSG00000204301"</span>, <span class="st">"ENSG00000112033"</span>, </span>
<span id="cb21-39"><a href="#cb21-39"></a>                 <span class="st">"ENSG00000196591"</span>, <span class="st">"ENSG00000135547"</span>, <span class="st">"ENSG00000198719"</span>, </span>
<span id="cb21-40"><a href="#cb21-40"></a>                 <span class="st">"ENSG00000127955"</span>, <span class="st">"ENSG00000157240"</span>, <span class="st">"ENSG00000055130"</span>, </span>
<span id="cb21-41"><a href="#cb21-41"></a>                 <span class="st">"ENSG00000104371"</span>, <span class="st">"ENSG00000164683"</span>, <span class="st">"ENSG00000136997"</span>, </span>
<span id="cb21-42"><a href="#cb21-42"></a>                 <span class="st">"ENSG00000185920"</span>, <span class="st">"ENSG00000148400"</span>, <span class="st">"ENSG00000283780"</span>, </span>
<span id="cb21-43"><a href="#cb21-43"></a>                 <span class="st">"ENSG00000275555"</span>, <span class="st">"ENSG00000238196"</span>, <span class="st">"ENSG00000235396"</span>, </span>
<span id="cb21-44"><a href="#cb21-44"></a>                 <span class="st">"ENSG00000232339"</span>, <span class="st">"ENSG00000223355"</span>, <span class="st">"ENSG00000206312"</span>, </span>
<span id="cb21-45"><a href="#cb21-45"></a>                 <span class="st">"ENSG00000234876"</span>)</span>
<span id="cb21-46"><a href="#cb21-46"></a></span>
<span id="cb21-47"><a href="#cb21-47"></a>ensembl_results &lt;-<span class="st"> </span><span class="kw">panther_go</span>(ensembl_ids, <span class="st">"9606"</span>, <span class="st">"biological_process"</span>)</span>
<span id="cb21-48"><a href="#cb21-48"></a></span>
<span id="cb21-49"><a href="#cb21-49"></a><span class="kw">head</span>(ensembl_results, <span class="dt">n =</span> <span class="dv">10</span>)</span>
<span id="cb21-50"><a href="#cb21-50"></a><span class="op">&gt;</span><span class="st"> </span>result_number number_in_list fold_enrichment      fdr expected number_in_reference   pValue plus_minus GO_term   description                                               </span>
<span id="cb21-51"><a href="#cb21-51"></a><span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">                  </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st">           </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">    </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">    </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">               </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st">    </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">      </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">     </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">                                                     </span></span>
<span id="cb21-52"><a href="#cb21-52"></a><span class="er">&gt;</span><span class="st"> </span><span class="dv">1</span>                         <span class="dv">12</span>           <span class="fl">78.4</span>  <span class="fl">1.70e-15</span>    <span class="fl">0.153</span>                  <span class="dv">84</span> <span class="fl">1.07e-19</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span>00600… canonical Wnt signaling pathway                           </span>
<span id="cb21-53"><a href="#cb21-53"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>                         <span class="dv">29</span>            <span class="fl">6.30</span> <span class="fl">1.25e-15</span>    <span class="fl">4.60</span>                 <span class="dv">2525</span> <span class="fl">1.57e-19</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span>00071… cell surface receptor signaling pathway                   </span>
<span id="cb21-54"><a href="#cb21-54"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">3</span>                         <span class="dv">16</span>           <span class="fl">25.2</span>  <span class="fl">4.33e-15</span>    <span class="fl">0.634</span>                 <span class="dv">348</span> <span class="fl">8.17e-19</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span>01987… cell<span class="op">-</span>cell signaling by wnt                                </span>
<span id="cb21-55"><a href="#cb21-55"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">4</span>                         <span class="dv">16</span>           <span class="fl">25.2</span>  <span class="fl">3.25e-15</span>    <span class="fl">0.634</span>                 <span class="dv">348</span> <span class="fl">8.17e-19</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span>00160… Wnt signaling pathway                                     </span>
<span id="cb21-56"><a href="#cb21-56"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span>                         <span class="dv">16</span>           <span class="fl">20.9</span>  <span class="fl">4.76e-14</span>    <span class="fl">0.767</span>                 <span class="dv">421</span> <span class="fl">1.50e-17</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span>19051… cell surface receptor signaling pathway involved <span class="cf">in</span> cell<span class="op">-</span>…</span>
<span id="cb21-57"><a href="#cb21-57"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">6</span>                         <span class="dv">21</span>           <span class="fl">10.2</span>  <span class="fl">9.38e-14</span>    <span class="fl">2.06</span>                 <span class="dv">1129</span> <span class="fl">3.54e-17</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span>00604… epithelium development                                    </span>
<span id="cb21-58"><a href="#cb21-58"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">7</span>                         <span class="dv">29</span>            <span class="fl">5.01</span> <span class="fl">1.92e-13</span>    <span class="fl">5.78</span>                 <span class="dv">3174</span> <span class="fl">8.46e-17</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span>00099… regulation of signal transduction                         </span>
<span id="cb21-59"><a href="#cb21-59"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">8</span>                         <span class="dv">30</span>            <span class="fl">4.60</span> <span class="fl">2.68e-13</span>    <span class="fl">6.52</span>                 <span class="dv">3576</span> <span class="fl">1.35e-16</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span>00106… regulation of cell communication                          </span>
<span id="cb21-60"><a href="#cb21-60"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">9</span>                         <span class="dv">30</span>            <span class="fl">4.55</span> <span class="fl">3.23e-13</span>    <span class="fl">6.59</span>                 <span class="dv">3615</span> <span class="fl">1.83e-16</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span>00230… regulation of signaling                                   </span>
<span id="cb21-61"><a href="#cb21-61"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span>                        <span class="dv">29</span>            <span class="fl">4.86</span> <span class="fl">3.22e-13</span>    <span class="fl">5.97</span>                 <span class="dv">3277</span> <span class="fl">2.02e-16</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span>00485… animal organ development </span>
<span id="cb21-62"><a href="#cb21-62"></a></span>
<span id="cb21-63"><a href="#cb21-63"></a><span class="co"># using mouse genes ------------------------------------------------------------</span></span>
<span id="cb21-64"><a href="#cb21-64"></a>mouse_genes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"Adam17"</span>, <span class="st">"Axin1"</span>, <span class="st">"Axin2"</span>, <span class="st">"Ccnd2"</span>, <span class="st">"Csnk1e"</span>, <span class="st">"Ctnnb1"</span>, </span>
<span id="cb21-65"><a href="#cb21-65"></a>                 <span class="st">"Cul1"</span>, <span class="st">"Dkk1"</span>, <span class="st">"Dkk4"</span>, <span class="st">"Dll1"</span>, <span class="st">"Dvl2"</span>, <span class="st">"Frat1"</span>, <span class="st">"Fzd1"</span>, </span>
<span id="cb21-66"><a href="#cb21-66"></a>                 <span class="st">"Fzd8"</span>, <span class="st">"Gnai1"</span>, <span class="st">"Hdac11"</span>, <span class="st">"Hdac2"</span>, <span class="st">"Hdac5"</span>, <span class="st">"Hey1"</span>, <span class="st">"Hey2"</span>, </span>
<span id="cb21-67"><a href="#cb21-67"></a>                 <span class="st">"Jag1"</span>, <span class="st">"Lef1"</span>, <span class="st">"Maml1"</span>, <span class="st">"Myc"</span>, <span class="st">"Ncor2"</span>, <span class="st">"Ncstn"</span>, <span class="st">"Notch1"</span>, </span>
<span id="cb21-68"><a href="#cb21-68"></a>                 <span class="st">"Notch4"</span>, <span class="st">"Numb"</span>, <span class="st">"Ppard"</span>, <span class="st">"Psen2"</span>, <span class="st">"Ptch1"</span>, <span class="st">"Skp2"</span>, <span class="st">"Tcf7"</span>, </span>
<span id="cb21-69"><a href="#cb21-69"></a>                 <span class="st">"Wnt1"</span>, <span class="st">"Wnt5b"</span>, <span class="st">"Wnt6"</span>)</span>
<span id="cb21-70"><a href="#cb21-70"></a></span>
<span id="cb21-71"><a href="#cb21-71"></a>mouse_results &lt;-<span class="st"> </span><span class="kw">panther_go</span>(</span>
<span id="cb21-72"><a href="#cb21-72"></a>                            mouse_genes, </span>
<span id="cb21-73"><a href="#cb21-73"></a>                            <span class="dt">organism =</span> <span class="st">"10090"</span>, </span>
<span id="cb21-74"><a href="#cb21-74"></a>                            <span class="dt">annot_dataset =</span> <span class="st">"biological_process"</span>)</span>
<span id="cb21-75"><a href="#cb21-75"></a></span>
<span id="cb21-76"><a href="#cb21-76"></a><span class="kw">head</span>(mouse_results, <span class="dt">n =</span> <span class="dv">10</span>)</span>
<span id="cb21-77"><a href="#cb21-77"></a><span class="op">&gt;</span><span class="st"> </span>result_number number_in_list fold_enrichment      fdr expected number_in_reference   pValue plus_minus GO_term    description                                                     </span>
<span id="cb21-78"><a href="#cb21-78"></a><span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">                  </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st">           </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">    </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">    </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st">               </span><span class="er">&lt;</span>int<span class="op">&gt;</span><span class="st">    </span><span class="er">&lt;</span>dbl<span class="op">&gt;</span><span class="st"> </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">      </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">      </span><span class="er">&lt;</span>chr<span class="op">&gt;</span><span class="st">                                                           </span></span>
<span id="cb21-79"><a href="#cb21-79"></a><span class="er">&gt;</span><span class="st"> </span><span class="dv">1</span>                         <span class="dv">13</span>           <span class="fl">97.8</span>  <span class="fl">2.50e-18</span>    <span class="fl">0.133</span>                  <span class="dv">80</span> <span class="fl">1.58e-22</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0060070</span> canonical Wnt signaling pathway                                 </span>
<span id="cb21-80"><a href="#cb21-80"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>                         <span class="dv">16</span>           <span class="fl">40.5</span>  <span class="fl">4.10e-18</span>    <span class="fl">0.396</span>                 <span class="dv">238</span> <span class="fl">5.18e-22</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0198738</span> cell<span class="op">-</span>cell signaling by wnt                                      </span>
<span id="cb21-81"><a href="#cb21-81"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">3</span>                         <span class="dv">16</span>           <span class="fl">40.5</span>  <span class="fl">2.73e-18</span>    <span class="fl">0.396</span>                 <span class="dv">238</span> <span class="fl">5.18e-22</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0016055</span> Wnt signaling pathway                                           </span>
<span id="cb21-82"><a href="#cb21-82"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">4</span>                         <span class="dv">23</span>           <span class="fl">13.3</span>  <span class="fl">3.77e-18</span>    <span class="fl">1.72</span>                 <span class="dv">1037</span> <span class="fl">9.53e-22</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0060429</span> epithelium development                                          </span>
<span id="cb21-83"><a href="#cb21-83"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span>                         <span class="dv">27</span>            <span class="fl">8.52</span> <span class="fl">8.41e-18</span>    <span class="fl">3.17</span>                 <span class="dv">1907</span> <span class="fl">2.66e-21</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0007166</span> cell surface receptor signaling pathway                         </span>
<span id="cb21-84"><a href="#cb21-84"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">6</span>                         <span class="dv">16</span>           <span class="fl">33.0</span>  <span class="fl">3.10e-17</span>    <span class="fl">0.485</span>                 <span class="dv">292</span> <span class="fl">1.18e-20</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">1905114</span> cell surface receptor signaling pathway involved <span class="cf">in</span> cell<span class="op">-</span>cell s…</span>
<span id="cb21-85"><a href="#cb21-85"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">7</span>                         <span class="dv">25</span>            <span class="fl">9.04</span> <span class="fl">1.39e-16</span>    <span class="fl">2.77</span>                 <span class="dv">1664</span> <span class="fl">6.17e-20</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0009888</span> tissue development                                              </span>
<span id="cb21-86"><a href="#cb21-86"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">8</span>                         <span class="dv">27</span>            <span class="fl">7.21</span> <span class="fl">3.95e-16</span>    <span class="fl">3.75</span>                 <span class="dv">2254</span> <span class="fl">2.00e-19</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0009653</span> anatomical structure morphogenesis                              </span>
<span id="cb21-87"><a href="#cb21-87"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">9</span>                         <span class="dv">29</span>            <span class="fl">6.06</span> <span class="fl">4.50e-16</span>    <span class="fl">4.78</span>                 <span class="dv">2878</span> <span class="fl">2.56e-19</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0009966</span> regulation of signal transduction                               </span>
<span id="cb21-88"><a href="#cb21-88"></a><span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span>                        <span class="dv">21</span>           <span class="fl">12.3</span>  <span class="fl">1.04e-15</span>    <span class="fl">1.71</span>                 <span class="dv">1026</span> <span class="fl">6.57e-19</span> <span class="op">+</span><span class="st">          </span>GO<span class="op">:</span><span class="dv">0009887</span> animal organ morphogenesis </span></code></pre></div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Gennaro Calendo <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Gennaro Calendo.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
