<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Practical analysis with SummarizedExperiments • coriell</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Practical analysis with SummarizedExperiments">
<meta property="og:description" content="coriell">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coriell</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.14.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Practical-analysis-with-SummarizedExperiments.html">Practical analysis with SummarizedExperiments</a>
    </li>
    <li>
      <a href="../articles/REdiscoverTE-workflow.html">REdiscoverTE Workflow</a>
    </li>
    <li>
      <a href="../articles/RNA-seq-workflow.html">RNA-seq-workflow</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/coriell-research/coriell/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Practical analysis with
SummarizedExperiments</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/coriell-research/coriell/blob/HEAD/vignettes/articles/Practical-analysis-with-SummarizedExperiments.Rmd" class="external-link"><code>vignettes/articles/Practical-analysis-with-SummarizedExperiments.Rmd</code></a></small>
      <div class="hidden name"><code>Practical-analysis-with-SummarizedExperiments.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="what-are-summarizedexperiments">What are <code>SummarizedExperiments</code><a class="anchor" aria-label="anchor" href="#what-are-summarizedexperiments"></a>
</h2>
<p><code>SummarizedExperiments</code> are R objects meant for organizing
and manipulating rectangular matrices that are typically produced by
arrays or high-throughput sequencing. If you are doing any kind of
analysis that requires associating feature-level data (RNA-seq gene
counts, methylation array loci, ATAC-seq regions, etc.) with the genomic
coordinates of those features and the sample-level metadata with which
those features were measured, then you should be using a
<code>SummarizedExperiment</code> to organize, manipulate, and store
your results.</p>
<p>Please take a moment to read through the first 2 sections (at least)
of the <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html" class="external-link">SummarizedExperiment
vignette</a> in order to familiarize yourself with what
<code>SummarizedExperiments</code> are and their structure. I will
demonstrate <em>how</em> you can use <code>SummarizedExperiments</code>
below.</p>
</div>
<div class="section level2">
<h2 id="subsetting-in-r">Subsetting in R<a class="anchor" aria-label="anchor" href="#subsetting-in-r"></a>
</h2>
<p><code>SummarizedExperiments</code> allow you to quickly and
effectively subset your data in a synchronized fashion that keeps all
sample-level metadata, feature-level matrix data, and genomic
range-level data consistent. The principles of
<code>SummarizedExperiments</code> are derived from base R subsetting
operations. Therefore, in order to become comfortable with
<code>SummarizedExperiments</code> you should be comfortable with some
base R functions.</p>
<p>You may have only ever encountered R from the perspective of the <a href="https://www.tidyverse.org/" class="external-link">tidyverse</a>. <code>tidyverse</code>
functions provide useful abstractions for munging <a href="https://vita.had.co.nz/papers/tidy-data.html" class="external-link">tidy data</a>
however, most genomics data is often best represented and operated on as
matrices. Keeping your data in matrix format can provide many benefits
as far as speed and code clarity, which in turn helps to ensure
correctness. You can think of matrices as just fancy 2D versions of
vectors. So what are vectors?</p>
<p>Vectors are the main building blocks of most R analyses. Whenever you
use the <code><a href="https://rdrr.io/r/base/c.html" class="external-link">c()</a></code> function, like:
<code>x &lt;- c('a', 'b', 'c')</code> you’re creating a vector. You can
do all kinds of cool things with vectors which will prove useful when
working with <code>SummarizedExperiments</code>.</p>
<p>NOTE: the following is heavily inspired by Norm Matloff’s excellent
<a href="https://github.com/matloff/fasteR" class="external-link">fasteR</a> tutorial. Take a
look there to get a brief and concise overview base R.</p>
<div class="section level3">
<h3 id="subsetting-vectors">Subsetting vectors<a class="anchor" aria-label="anchor" href="#subsetting-vectors"></a>
</h3>
<p>Below, we’ll use the built-in R constant called <code>LETTERS</code>.
The <code>LETTERS</code> vector is simply a ‘list’ of all uppercase
letters in the Roman alphabet.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LETTERS</span></span>
<span><span class="co">#&gt;  [1] "A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K" "L" "M" "N" "O" "P" "Q" "R" "S"</span></span>
<span><span class="co">#&gt; [20] "T" "U" "V" "W" "X" "Y" "Z"</span></span></code></pre></div>
<p>We can subset the vector by position. For example, to get the 3rd
letter</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LETTERS</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "C"</span></span></code></pre></div>
<p>Or a range of positions</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LETTERS</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "C" "D" "E" "F" "G"</span></span></code></pre></div>
<p>We don’t have to select sequential elements either. We can extract
elements by using another vector of positions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LETTERS</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">7</span>, <span class="fl">5</span>, <span class="fl">14</span>, <span class="fl">14</span>, <span class="fl">1</span>, <span class="fl">18</span>, <span class="fl">15</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "G" "E" "N" "N" "A" "R" "O"</span></span></code></pre></div>
<p>Vectors become really powerful when we start combining them with
logical operations.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_favorite_letters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># See that this produces a logical vector of (TRUE/FALSE) values </span></span>
<span><span class="co"># TRUE when LETTERS is one of my_favorite_letters and FALSE otherwise</span></span>
<span><span class="va">LETTERS</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">my_favorite_letters</span></span>
<span><span class="co">#&gt;  [1]  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span></span>
<span><span class="co">#&gt; [13] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE</span></span>
<span><span class="co">#&gt; [25] FALSE FALSE</span></span>
<span></span>
<span><span class="co"># We can use that same expression to filter the vector</span></span>
<span><span class="va">LETTERS</span><span class="op">[</span><span class="va">LETTERS</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">my_favorite_letters</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "A" "B" "C"</span></span></code></pre></div>
<p>This same kind of subsetting works on vectors that contain numeric
data as well. For example, we can filter the measurements of annual flow
of water through the Nile river like so:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Any values strictly greater than 1200</span></span>
<span><span class="va">Nile</span><span class="op">[</span><span class="va">Nile</span> <span class="op">&gt;</span> <span class="fl">1200</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] 1210 1230 1370 1210 1250 1260 1220</span></span>
<span></span>
<span><span class="co"># Any even number</span></span>
<span><span class="va">Nile</span><span class="op">[</span><span class="va">Nile</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html" class="external-link">%%</a></span> <span class="fl">2</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span></span>
<span><span class="co">#&gt;  [1] 1120 1160 1210 1160 1160 1230 1370 1140 1110  994 1020  960 1180  958 1140</span></span>
<span><span class="co">#&gt; [16] 1100 1210 1150 1250 1260 1220 1030 1100  774  840  874  694  940  916  692</span></span>
<span><span class="co">#&gt; [31] 1020 1050  726  456  824  702 1120 1100  832  764  768  864  862  698  744</span></span>
<span><span class="co">#&gt; [46]  796 1040  944  984  822 1010  676  846  812  742 1040  860  874  848  890</span></span>
<span><span class="co">#&gt; [61]  744  838 1050  918  986 1020  906 1170  912  746  718  714  740</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="subsetting-data-frames">Subsetting data.frames<a class="anchor" aria-label="anchor" href="#subsetting-data-frames"></a>
</h3>
<p>But these are just one dimensional vectors. In R we usually deal with
data.frames (tibbles for you tidyers) and matrices. Luckily for us, the
subsetting operations we learned for vectors work the same way for
data.frames and matrices.</p>
<p>Let’s take a look at the built-in <code>ToothGrowth</code> dataset.
The data consists of the length of odontoblasts in 60 guinea pigs
receiving one of three levels of vitamin C by one of two delivery
methods.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ToothGrowth</span><span class="op">)</span></span>
<span><span class="co">#&gt;    len supp dose</span></span>
<span><span class="co">#&gt; 1  4.2   VC  0.5</span></span>
<span><span class="co">#&gt; 2 11.5   VC  0.5</span></span>
<span><span class="co">#&gt; 3  7.3   VC  0.5</span></span>
<span><span class="co">#&gt; 4  5.8   VC  0.5</span></span>
<span><span class="co">#&gt; 5  6.4   VC  0.5</span></span>
<span><span class="co">#&gt; 6 10.0   VC  0.5</span></span></code></pre></div>
<p>The dollar sign <code>$</code> is used to extract an individual
column which is just a vector.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ToothGrowth</span><span class="op">$</span><span class="va">len</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  4.2 11.5  7.3  5.8  6.4 10.0</span></span></code></pre></div>
<p>When subsetting a data.frame in base R the general scheme is</p>
<pre><code>df[the rows you want, the columns you want]</code></pre>
<p>So in order to get the 5th row of the first column we could do:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ToothGrowth</span><span class="op">[</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] 6.4</span></span></code></pre></div>
<p>Again, we can combine this kind of thinking to extract rows and
columns matching logical conditions. For example, if we want to get all
of the animals administered orange juice (‘OJ’)</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ToothGrowth</span><span class="op">[</span><span class="va">ToothGrowth</span><span class="op">$</span><span class="va">supp</span> <span class="op">==</span> <span class="st">"OJ"</span>, <span class="op">]</span></span>
<span><span class="co">#&gt;     len supp dose</span></span>
<span><span class="co">#&gt; 31 15.2   OJ  0.5</span></span>
<span><span class="co">#&gt; 32 21.5   OJ  0.5</span></span>
<span><span class="co">#&gt; 33 17.6   OJ  0.5</span></span>
<span><span class="co">#&gt; 34  9.7   OJ  0.5</span></span>
<span><span class="co">#&gt; 35 14.5   OJ  0.5</span></span>
<span><span class="co">#&gt; 36 10.0   OJ  0.5</span></span>
<span><span class="co">#&gt; 37  8.2   OJ  0.5</span></span>
<span><span class="co">#&gt; 38  9.4   OJ  0.5</span></span>
<span><span class="co">#&gt; 39 16.5   OJ  0.5</span></span>
<span><span class="co">#&gt; 40  9.7   OJ  0.5</span></span>
<span><span class="co">#&gt; 41 19.7   OJ  1.0</span></span>
<span><span class="co">#&gt; 42 23.3   OJ  1.0</span></span>
<span><span class="co">#&gt; 43 23.6   OJ  1.0</span></span>
<span><span class="co">#&gt; 44 26.4   OJ  1.0</span></span>
<span><span class="co">#&gt; 45 20.0   OJ  1.0</span></span>
<span><span class="co">#&gt; 46 25.2   OJ  1.0</span></span>
<span><span class="co">#&gt; 47 25.8   OJ  1.0</span></span>
<span><span class="co">#&gt; 48 21.2   OJ  1.0</span></span>
<span><span class="co">#&gt; 49 14.5   OJ  1.0</span></span>
<span><span class="co">#&gt; 50 27.3   OJ  1.0</span></span>
<span><span class="co">#&gt; 51 25.5   OJ  2.0</span></span>
<span><span class="co">#&gt; 52 26.4   OJ  2.0</span></span>
<span><span class="co">#&gt; 53 22.4   OJ  2.0</span></span>
<span><span class="co">#&gt; 54 24.5   OJ  2.0</span></span>
<span><span class="co">#&gt; 55 24.8   OJ  2.0</span></span>
<span><span class="co">#&gt; 56 30.9   OJ  2.0</span></span>
<span><span class="co">#&gt; 57 26.4   OJ  2.0</span></span>
<span><span class="co">#&gt; 58 27.3   OJ  2.0</span></span>
<span><span class="co">#&gt; 59 29.4   OJ  2.0</span></span>
<span><span class="co">#&gt; 60 23.0   OJ  2.0</span></span></code></pre></div>
<p>We can also combine logical statements. For example, to get all of
the rows for animals administered orange juice and with odontoblast
length (‘len’) less than 10.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ToothGrowth</span><span class="op">[</span><span class="va">ToothGrowth</span><span class="op">$</span><span class="va">supp</span> <span class="op">==</span> <span class="st">"OJ"</span> <span class="op">&amp;</span> <span class="va">ToothGrowth</span><span class="op">$</span><span class="va">len</span> <span class="op">&lt;</span> <span class="fl">10</span>, <span class="op">]</span></span>
<span><span class="co">#&gt;    len supp dose</span></span>
<span><span class="co">#&gt; 34 9.7   OJ  0.5</span></span>
<span><span class="co">#&gt; 37 8.2   OJ  0.5</span></span>
<span><span class="co">#&gt; 38 9.4   OJ  0.5</span></span>
<span><span class="co">#&gt; 40 9.7   OJ  0.5</span></span>
<span></span>
<span><span class="co"># We can also use the bracket notation to select rows and columns at the same time</span></span>
<span><span class="va">ToothGrowth</span><span class="op">[</span><span class="va">ToothGrowth</span><span class="op">$</span><span class="va">supp</span> <span class="op">==</span> <span class="st">"OJ"</span> <span class="op">&amp;</span> <span class="va">ToothGrowth</span><span class="op">$</span><span class="va">len</span> <span class="op">&lt;</span> <span class="fl">10</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"len"</span>, <span class="st">"supp"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#&gt;    len supp</span></span>
<span><span class="co">#&gt; 34 9.7   OJ</span></span>
<span><span class="co">#&gt; 37 8.2   OJ</span></span>
<span><span class="co">#&gt; 38 9.4   OJ</span></span>
<span><span class="co">#&gt; 40 9.7   OJ</span></span></code></pre></div>
<p>It gets annoying typing <code>ToothGrowth</code> every time we want
to subset the data.frame. Base R has a very useful function called
<code><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset()</a></code> that can help us type less. <code><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset()</a></code>
essentially ‘looks inside’ the data.frame that you give it for the given
columns and evaluates the expression without having to explicitly tell R
where to find the columns. Think of it like
<code><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">dplyr::filter()</a></code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">ToothGrowth</span>, <span class="va">supp</span> <span class="op">==</span> <span class="st">"OJ"</span> <span class="op">&amp;</span> <span class="va">len</span> <span class="op">&lt;</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="co">#&gt;    len supp dose</span></span>
<span><span class="co">#&gt; 34 9.7   OJ  0.5</span></span>
<span><span class="co">#&gt; 37 8.2   OJ  0.5</span></span>
<span><span class="co">#&gt; 38 9.4   OJ  0.5</span></span>
<span><span class="co">#&gt; 40 9.7   OJ  0.5</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="subsetting-matrices">Subsetting matrices<a class="anchor" aria-label="anchor" href="#subsetting-matrices"></a>
</h3>
<p>Matrices behave much like data.frames but unlike data.frames matrices
can only contain one type of data. This might sound like a limitation at
first but you’ll soon come to realize that matrices are very powerful
(and fast) to work with in R.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create some random data that looks like</span></span>
<span><span class="op">(</span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">6</span> <span class="op">*</span> <span class="fl">10</span><span class="op">)</span>, </span>
<span>  ncol <span class="op">=</span> <span class="fl">6</span>, </span>
<span>  dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"CpG."</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span>, </span>
<span>                  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;          Sample1    Sample2   Sample3   Sample4    Sample5    Sample6</span></span>
<span><span class="co">#&gt; CpG.1  0.8650113 0.73314697 0.2562499 0.3421035 0.51109453 0.03229310</span></span>
<span><span class="co">#&gt; CpG.2  0.5618446 0.01594823 0.7197144 0.3320535 0.68127468 0.02306232</span></span>
<span><span class="co">#&gt; CpG.3  0.2156928 0.59236443 0.9317838 0.2858852 0.20217931 0.13576589</span></span>
<span><span class="co">#&gt; CpG.4  0.4311463 0.28576197 0.8999364 0.2036986 0.74198920 0.58805204</span></span>
<span><span class="co">#&gt; CpG.5  0.3236574 0.53974444 0.5637210 0.4934779 0.62581354 0.66690693</span></span>
<span><span class="co">#&gt; CpG.6  0.2947591 0.34520790 0.7354448 0.6344497 0.12656096 0.89445699</span></span>
<span><span class="co">#&gt; CpG.7  0.6470855 0.16859332 0.1406321 0.4053713 0.91727803 0.59130453</span></span>
<span><span class="co">#&gt; CpG.8  0.7173902 0.41308704 0.7583214 0.9985040 0.15849981 0.65449142</span></span>
<span><span class="co">#&gt; CpG.9  0.4467033 0.78632410 0.7823403 0.1384543 0.84780102 0.16660458</span></span>
<span><span class="co">#&gt; CpG.10 0.5864781 0.38885772 0.8167770 0.1677617 0.02846175 0.63311029</span></span></code></pre></div>
<p>If we want to extract the value for CpG.3 for Sample3</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span><span class="op">[</span><span class="fl">3</span>, <span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] 0.9317838</span></span></code></pre></div>
<p>Or all values of CpG.3 for each sample</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span><span class="op">[</span><span class="fl">3</span>, <span class="op">]</span></span>
<span><span class="co">#&gt;   Sample1   Sample2   Sample3   Sample4   Sample5   Sample6 </span></span>
<span><span class="co">#&gt; 0.2156928 0.5923644 0.9317838 0.2858852 0.2021793 0.1357659</span></span></code></pre></div>
<p>We can also apply a mask to the entire matrix at once. For example,
the following will mark any value that is greater than 0.5 with
<code>TRUE</code></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">&gt;</span> <span class="fl">0.5</span></span>
<span><span class="co">#&gt;        Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span><span class="co">#&gt; CpG.1     TRUE    TRUE   FALSE   FALSE    TRUE   FALSE</span></span>
<span><span class="co">#&gt; CpG.2     TRUE   FALSE    TRUE   FALSE    TRUE   FALSE</span></span>
<span><span class="co">#&gt; CpG.3    FALSE    TRUE    TRUE   FALSE   FALSE   FALSE</span></span>
<span><span class="co">#&gt; CpG.4    FALSE   FALSE    TRUE   FALSE    TRUE    TRUE</span></span>
<span><span class="co">#&gt; CpG.5    FALSE    TRUE    TRUE   FALSE    TRUE    TRUE</span></span>
<span><span class="co">#&gt; CpG.6    FALSE   FALSE    TRUE    TRUE   FALSE    TRUE</span></span>
<span><span class="co">#&gt; CpG.7     TRUE   FALSE   FALSE   FALSE    TRUE    TRUE</span></span>
<span><span class="co">#&gt; CpG.8     TRUE   FALSE    TRUE    TRUE   FALSE    TRUE</span></span>
<span><span class="co">#&gt; CpG.9    FALSE    TRUE    TRUE   FALSE    TRUE   FALSE</span></span>
<span><span class="co">#&gt; CpG.10    TRUE   FALSE    TRUE   FALSE   FALSE    TRUE</span></span></code></pre></div>
<p>We can use this kind of masking to filter rows of the matrix using
some very helpful base R functions that operate on matrices. For
example, to get only those CpGs where 3 or more samples have a value
&gt; 0.5 we can use the <code><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums()</a></code> like so.</p>
<p>This pattern is very common when dealing with genetics data. Base R
functions like <code><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums()</a></code> and <code><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colMeans()</a></code> are
specialized to operate over matrices and are the most efficient way to
summarize matrix data. The R package <a href="https://github.com/HenrikBengtsson/matrixStats" class="external-link">matrixStats</a>
also contains highly optimized functions for operating on matrices.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">m</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">3</span>, <span class="op">]</span></span>
<span><span class="co">#&gt;         Sample1   Sample2   Sample3   Sample4   Sample5   Sample6</span></span>
<span><span class="co">#&gt; CpG.5 0.3236574 0.5397444 0.5637210 0.4934779 0.6258135 0.6669069</span></span>
<span><span class="co">#&gt; CpG.8 0.7173902 0.4130870 0.7583214 0.9985040 0.1584998 0.6544914</span></span></code></pre></div>
<p>Compare the above to the <code>tidy</code> solution given the same
matrix and converting back to a matrix.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">m</span>, rownames <span class="op">=</span> <span class="st">"CpG"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="op">!</span><span class="va">CpG</span>, names_to <span class="op">=</span> <span class="st">"SampleName"</span>, values_to <span class="op">=</span> <span class="st">"beta"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">CpG</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>n_gt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">beta</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">n_gt</span> <span class="op">&gt;</span> <span class="fl">3</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider</a></span><span class="op">(</span>id_cols <span class="op">=</span> <span class="va">CpG</span>, names_from <span class="op">=</span> <span class="st">"SampleName"</span>, values_from <span class="op">=</span> <span class="st">"beta"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"CpG"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/data.matrix.html" class="external-link">data.matrix</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;         Sample1   Sample2   Sample3   Sample4   Sample5   Sample6</span></span>
<span><span class="co">#&gt; CpG.5 0.3236574 0.5397444 0.5637210 0.4934779 0.6258135 0.6669069</span></span>
<span><span class="co">#&gt; CpG.8 0.7173902 0.4130870 0.7583214 0.9985040 0.1584998 0.6544914</span></span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Gennaro Calendo.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
