<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>REdiscoverTE Workflow â€¢ coriell</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="REdiscoverTE Workflow">
<meta property="og:description" content="coriell">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coriell</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.14.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Practical-analysis-with-SummarizedExperiments.html">Practical-analysis-with-SummarizedExperiments</a>
    </li>
    <li>
      <a href="../articles/REdiscoverTE-workflow.html">REdiscoverTE Workflow</a>
    </li>
    <li>
      <a href="../articles/RNA-seq-workflow.html">RNA-seq-workflow</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/coriell-research/coriell/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>REdiscoverTE Workflow</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/coriell-research/coriell/blob/HEAD/vignettes/articles/REdiscoverTE-workflow.Rmd" class="external-link"><code>vignettes/articles/REdiscoverTE-workflow.Rmd</code></a></small>
      <div class="hidden name"><code>REdiscoverTE-workflow.Rmd</code></div>

    </div>

    
    
<p>This document outlines my general steps for extracting reads in fastq
format from BAM files, aligning them to the REdiscoverTE transcriptome,
and quantifying their counts with <code>tximport</code></p>
<div class="section level2">
<h2 id="get-data-from-tcga">1. Get data from TCGA<a class="anchor" aria-label="anchor" href="#get-data-from-tcga"></a>
</h2>
<p>Typically I have to do this workflow for TCGA samples so include a
step here to document downloading data from TCGA.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="va">MANIFEST</span><span class="op">=</span>/path/to/gdc_manifest.txt</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="va">TOKEN</span><span class="op">=</span>/path/to/tcga-token-file.txt</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="va">BAM_DIR</span><span class="op">=</span>/path/to/save/bamfiles</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="va">PROC</span><span class="op">=</span>12</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co"># Download all files from the manifest into BAM_DIR using 12 processes</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="ex">gdc-client</span> download <span class="dt">\</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>  <span class="at">--manifest</span> <span class="va">$MANIFEST</span> <span class="dt">\</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>  <span class="at">--token-file</span> <span class="va">$TOKEN</span> <span class="dt">\</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>  <span class="at">--dir</span> <span class="va">$BAM_DIR</span> <span class="dt">\</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>  <span class="at">--n-processes</span> <span class="va">$PROC</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="extract-fastq-files-from-bams">2. Extract fastq files from BAMs<a class="anchor" aria-label="anchor" href="#extract-fastq-files-from-bams"></a>
</h2>
<p>Use <code>parallel</code> to extract fastq reads from all bamfiles
found by <code>find</code> in the <code>$BAM_DIR</code> defined above.
<code>samtools collate</code> groups the two reads in a read pair and
outputs an uncompressed BAM stream. <code>samtools fastq</code> consumes
this stream and outputs the reads pairs to separate files. Unpaired
reads are written to /dev/null.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="va">FQ_DIR</span><span class="op">=</span>/path/to/save/fastq_files</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="co"># Extract Reads</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="ex">parallel</span> <span class="at">--jobs</span> <span class="va">$PROC</span> <span class="st">"samtools collate -u -O {} /tmp/{/.}_tmp | samtools fastq - -1 </span><span class="va">$FQ_DIR</span><span class="st">/{/.}.1.fq.gz -2 </span><span class="va">$FQ_DIR</span><span class="st">/{/.}.2.fq.gz -0 /dev/null -n"</span> ::: <span class="va">$(</span><span class="fu">find</span> <span class="va">$BAM_DIR</span> <span class="at">-iname</span> <span class="st">"*.bam"</span><span class="va">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="align-to-rediscoverte-transcriptome">3. Align to REdiscoverTE transcriptome<a class="anchor" aria-label="anchor" href="#align-to-rediscoverte-transcriptome"></a>
</h2>
<p>This assumes you have a text file with basenames for the sample fastq
files like:</p>
<p><strong>sample-names.txt</strong></p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">sample1</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="ex">sample2</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="ex">sample3</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="ex">sample4</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="va">IDX</span><span class="op">=</span>/mnt/data/gdata/human/REdiscoverTE_hg38/REdiscoverTE_hg38_GFP/REdiscoverTE_GFP_salmon_idx/</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="va">OUT_DIR</span><span class="op">=</span>/path/to/qaunts</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="va">FQ_DIR</span><span class="op">=</span>/path/to/fastq_files</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="va">SAMPLES</span><span class="op">=</span>/path/to/sample-names.txt</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="va">THREADS</span><span class="op">=</span>12</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="cf">for</span> SAMPLE <span class="kw">in</span> <span class="va">$(</span><span class="fu">cat</span> <span class="va">$SAMPLES)</span><span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>  <span class="ex">salmon</span> quant <span class="dt">\</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>    <span class="at">-i</span> <span class="va">$IDX</span> <span class="dt">\</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>    <span class="at">-l</span> <span class="st">'A'</span> <span class="dt">\</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a>    <span class="at">-p</span> <span class="va">$THREADS</span> <span class="dt">\</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>    <span class="at">--gcBias</span> <span class="dt">\</span></span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a>    <span class="at">--seqBias</span> <span class="dt">\</span></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a>    <span class="at">---validateMappings</span> <span class="dt">\</span></span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a>    <span class="at">-1</span> <span class="va">$FQ_DIR</span>/<span class="va">${SAMPLE}</span>.1.fq.gz <span class="dt">\</span></span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a>    <span class="at">-2</span> <span class="va">$FQ_DIR</span>/<span class="va">${SAMPLE}</span>.fq.gz <span class="dt">\</span></span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a>    <span class="at">-o</span> <span class="va">$OUT_DIR</span>/<span class="va">${SAMPLE}</span>_quants<span class="kw">;</span></span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a><span class="cf">done</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="import-counts-into-r">4. Import counts into <code>R</code><a class="anchor" aria-label="anchor" href="#import-counts-into-r"></a>
</h2>
<p>The next step imports the quant files from <code>salmon</code> into R
for all mapped samples and creates a matrix of gene-level counts that
can be used in downstream analyses.</p>
<p>It is necessary to use <code>tximport</code> for this step instead of
raw counts from the quant.sf files to account for differences in
aggregation due to transcript differences between samples.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/tximport" class="external-link">tximport</a></span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">quant_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span></span>
<span>  path <span class="op">=</span> <span class="st">"/path/to/quants"</span>,</span>
<span>  pattern <span class="op">=</span> <span class="st">"quant.sf"</span>,</span>
<span>  recursive <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  full.names <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Name each file with the basename of the sample -- this should be changed</span></span>
<span><span class="co"># names() will be used as the column names in the final count matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">quant_files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">desired_vector_of_sample_names</span></span>
<span></span>
<span><span class="co"># tximport maps gene ids (md5sums) to genes and RE sub-families</span></span>
<span><span class="va">tx2gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"/mnt/data/gdata/human/REdiscoverTE_hg38/tx2gene_REdiscoverTE.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Import counts for all files with tximport -- set scaledTPM to account for tx usage</span></span>
<span><span class="va">txi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tximport/man/tximport.html" class="external-link">tximport</a></span><span class="op">(</span></span>
<span>  <span class="va">quant_files</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"salmon"</span>,</span>
<span>  countsFromAbundance <span class="op">=</span> <span class="st">"scaledTPM"</span>,</span>
<span>  tx2gene <span class="op">=</span> <span class="va">tx2gene</span>,</span>
<span>  importer <span class="op">=</span> <span class="fu">data.table</span><span class="fu">::</span><span class="va"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract counts from the txi object</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">txi</span><span class="op">$</span><span class="va">counts</span></span>
<span></span>
<span><span class="co"># Remove any all 0 rows to save some space</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="op">]</span></span></code></pre></div>
<p>We download and import the GENCODE v26 gtf file used in the creation
of the REdiscoverTE index in order to define protein coding genes. Then
we split the <code>counts</code> matrix into genes, exonic, intronic,
and intergenic repeat element sub-families from the five â€˜mainâ€™ TE
classes (LINE, SINE, LTR, SVA, DNA).</p>
<p>The final <code>counts</code> matrix used in downstream analysis
contains the gene expression and the expression of REs from intergenic
and intronic regions.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the reference GTF file and extract a vector of protein coding genes</span></span>
<span><span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_26/gencode.v26.annotation.gtf.gz"</span></span>
<span><span class="va">gtf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="va">url</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">coding</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">gtf</span><span class="op">[</span><span class="va">gtf</span><span class="op">$</span><span class="va">type</span> <span class="op">==</span> <span class="st">"gene"</span> <span class="op">&amp;</span> <span class="va">gtf</span><span class="op">$</span><span class="va">gene_type</span> <span class="op">==</span> <span class="st">"protein_coding"</span>, <span class="st">"gene_name"</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Filter matrix for protein coding genes</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">%chin%</span> <span class="va">coding</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Filter matrix for repeats from five 'main' classes/families</span></span>
<span><span class="va">repeats</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^LINE\\.|^SINE\\.|^LTR\\.|.*\\.SVA\\..*|^DNA\\."</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Split into repeat types by location</span></span>
<span><span class="va">intergenic</span> <span class="op">&lt;-</span> <span class="va">repeats</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"__intergenic"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">repeats</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">exonic</span> <span class="op">&lt;-</span> <span class="va">repeats</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"__exon"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">repeats</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">intronic</span> <span class="op">&lt;-</span> <span class="va">repeats</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"__intron"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">repeats</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Only use repeats that are not located in exons</span></span>
<span><span class="va">repeats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">intronic</span>, <span class="va">intergenic</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Strip the __intron __intergenic labels and sum counts for each element</span></span>
<span><span class="va">elem</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"__intergenic|__intron"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">repeats</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">repeats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rowsum.html" class="external-link">rowsum</a></span><span class="op">(</span><span class="va">repeats</span>, group <span class="op">=</span> <span class="va">elem</span>, reorder <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Bind all back into a single matrix, ready for downstream analysis</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">genes</span>, <span class="va">repeats</span><span class="op">)</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Gennaro Calendo.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
