<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RNA-seq-workflow • coriell</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="RNA-seq-workflow">
<meta property="og:description" content="coriell">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coriell</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.14.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Practical-analysis-with-SummarizedExperiments.html">Practical-analysis-with-SummarizedExperiments</a>
    </li>
    <li>
      <a href="../articles/REdiscoverTE-workflow.html">REdiscoverTE Workflow</a>
    </li>
    <li>
      <a href="../articles/RNA-seq-workflow.html">RNA-seq-workflow</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/coriell-research/coriell/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>RNA-seq-workflow</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/coriell-research/coriell/blob/HEAD/vignettes/articles/RNA-seq-workflow.Rmd" class="external-link"><code>vignettes/articles/RNA-seq-workflow.Rmd</code></a></small>
      <div class="hidden name"><code>RNA-seq-workflow.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>Below, I will illustrate some of the functions present in the
<code>coriell</code> package by walking through a typical RNA-seq
workflow, starting with checking data for quality and then progressing
through quantifying with <a href="https://salmon.readthedocs.io/en/latest/salmon.html" class="external-link">Salmon</a>
and performing differential expression analysis with
<code>edgeR</code>.</p>
<p><strong>update (2024-02)</strong> Since writing this vignette I have
updated my preferred TE workflow away from REdiscoverTE and towards
using a bootstrapped based quantification method instead. This vignette
still contains many examples of the <code>coriell</code> package and
typical RNA-seq analysis steps.</p>
</div>
<div class="section level2">
<h2 id="checking-sequences-for-quality">Checking sequences for quality<a class="anchor" aria-label="anchor" href="#checking-sequences-for-quality"></a>
</h2>
<p>The first step in any high-throughput analysis should be to check
your reads for quality. <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
is the <em>de facto</em> tool for this job.</p>
<p>To run <code>fastqc</code> enter the directory where your fastq files
are saved and run:</p>
<pre><code><span><span class="va">fastqc</span> <span class="op">*</span><span class="va">.fastq.gz</span></span></code></pre>
<p>Additional arguments can be passed to the program to increase the
number of parallel processes as well as specify the output directory.
e.g.  <code>fastqc *.fq.gz --threads 12 --outdir $OUT</code></p>
<p>An especially useful step following <code>fastqc</code> is to run the
program <a href="https://multiqc.info/" class="external-link">MultiQC</a> which automatically
scans a directory and summarizes all QC files it finds. If you have many
samples <code>multiqc</code> is an invaluable tools for quickly
examining the overall sequencing quality</p>
</div>
<div class="section level2">
<h2 id="aligning-with-salmon">Aligning with Salmon<a class="anchor" aria-label="anchor" href="#aligning-with-salmon"></a>
</h2>
<p>Assuming you have already built or downloaded a Salmon index then
performing quantification with Salmon is run with a simple shell
script:</p>
<pre><code>#!/usr/bin/env bash

SAMPLE_IDS=sample-ids.txt
FQ_DIR=path/to/fastq/files
OUT_DIR=quants
SALMON_IDX=path/to/salmon_idx
THREADS=48

mkdir -p $OUT_DIR

for SAMPLE_ID in $(cat $SAMPLE_IDS); do
  salmon quant \
     -i $SALMON_IDX \
     -l A \
     -1 $FQ_DIR/${SAMPLE_ID}_R1.fq.gz \
     -2 $FQ_DIR/${SAMPLE_ID}_R2.fq.gz \
     --validateMappings \
     --gcBias \
     --seqBias \
     --threads $THREADS \
     -o $OUT_DIR/${SAMPLE_ID}_quants;
done</code></pre>
<p>Where sample-ids.txt contains the basenames of the fastq files
like:</p>
<pre><code><span><span class="va">control1</span></span>
<span><span class="va">control2</span></span>
<span><span class="va">control3</span></span>
<span><span class="va">treatment1</span></span>
<span><span class="va">treatment2</span></span>
<span><span class="va">treatment3</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="load-libraries">Load libraries<a class="anchor" aria-label="anchor" href="#load-libraries"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/coriell-research/coriell" class="external-link">coriell</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tximportData</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/tximport" class="external-link">tximport</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kevinblighe/PCAtools" class="external-link">PCAtools</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">clusterProfiler</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ctlab/fgsea/" class="external-link">fgsea</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://davislaboratory.github.io/msigdb" class="external-link">msigdb</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="differential-gene-expression-analysis">Differential Gene Expression Analysis<a class="anchor" aria-label="anchor" href="#differential-gene-expression-analysis"></a>
</h2>
<p>For this sample analysis we will use the included data in the
<code>tximportData</code> package.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Select the directory where the quants files are stored</span></span>
<span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"tximportData"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Read in the sample metadata</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"samples.txt"</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">$</span><span class="va">run</span></span>
<span></span>
<span><span class="co"># List all quant files in dir</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"salmon"</span>, <span class="va">samples</span><span class="op">$</span><span class="va">run</span>, <span class="st">"quant.sf.gz"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Name the files by their run ID</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/regmatches.html" class="external-link">regmatches</a></span><span class="op">(</span><span class="va">files</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">regexpr</a></span><span class="op">(</span><span class="st">"ERR[0-9]+"</span>, <span class="va">files</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Read in the tx2gene file that maps transcripts to genes</span></span>
<span><span class="va">tx2gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"tx2gene.gencode.v27.csv"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>In your code, quant files will be saved in separate folders per
sample if you followed the example script from above. The command to
find these files might look more like:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span></span>
<span>  path <span class="op">=</span> <span class="st">"/path/to/OUT_DIR"</span>,</span>
<span>  pattern <span class="op">=</span> <span class="st">"quant.sf"</span>,</span>
<span>  recursive <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  full.names <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="import-counts-with-tximport">Import counts with <code>tximport</code><a class="anchor" aria-label="anchor" href="#import-counts-with-tximport"></a>
</h3>
<p>We often want to perform differential expression not on the
transcript level but the gene-level. This requires summarizing the
counts and applying an offset to adjust for length biases. The easiest
way to import counts with an offset is to use the <code>tximport</code>
function with the “scaledTPM” option.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Import the counts using tximport</span></span>
<span><span class="va">txi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tximport/man/tximport.html" class="external-link">tximport</a></span><span class="op">(</span></span>
<span>  <span class="va">files</span>, </span>
<span>  type <span class="op">=</span> <span class="st">"salmon"</span>, </span>
<span>  tx2gene <span class="op">=</span> <span class="va">tx2gene</span>, </span>
<span>  countsFromAbundance <span class="op">=</span> <span class="st">"scaledTPM"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; reading in files with read_tsv</span></span>
<span><span class="co">#&gt; 1 2 3 4 5 6 </span></span>
<span><span class="co">#&gt; summarizing abundance</span></span>
<span><span class="co">#&gt; summarizing counts</span></span>
<span><span class="co">#&gt; summarizing length</span></span>
<span></span>
<span><span class="co"># Extract the counts matrix from the txi object</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">txi</span><span class="op">$</span><span class="va">counts</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filter-the-count-matrix">Filter the count matrix<a class="anchor" aria-label="anchor" href="#filter-the-count-matrix"></a>
</h3>
<p>It is often only of interest to study the protein coding genes. To
get the protein coding genes and the gene symbols all in one go it is
usually easiest to work with GTF files directly.</p>
<p>The example data was generated using GENCODE v27 annotations on hg19.
We can download this annotation file and use it to select the protein
coding genes as well as extract other useful information about the
genes.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Download and import the annotations</span></span>
<span><span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_27/GRCh37_mapping/gencode.v27lift37.basic.annotation.gtf.gz"</span></span>
<span><span class="va">gtf_file</span> <span class="op">&lt;-</span> <span class="st">"~/Downloads/gencode.v27lift37.basic.annotation.gtf.gz"</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">gtf_file</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">url</span>, <span class="va">gtf_file</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">gtf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="va">gtf_file</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract all of the protein coding genes from the GTF</span></span>
<span><span class="va">protein_coding</span> <span class="op">&lt;-</span> <span class="va">gtf</span><span class="op">[</span><span class="va">gtf</span><span class="op">$</span><span class="va">type</span> <span class="op">==</span> <span class="st">"gene"</span> <span class="op">&amp;</span> <span class="va">gtf</span><span class="op">$</span><span class="va">gene_type</span> <span class="op">==</span> <span class="st">"protein_coding"</span>, <span class="st">"gene_id"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Removing trailing ids if present</span></span>
<span><span class="va">protein_coding</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_[0-9]+$"</span>, <span class="st">""</span>, <span class="va">protein_coding</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Filter the count matrix for protein coding genes</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">counts</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">protein_coding</span>, <span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="checking-the-library-distributions">Checking the library distributions<a class="anchor" aria-label="anchor" href="#checking-the-library-distributions"></a>
</h3>
<p>It is always good to check to see if the sequenced libraries are
behaving as expected. The <code>coriell</code> package has a few
diagnostic plots to help with this.</p>
<p>The first diagnostic plot is the <code><a href="../reference/plot_boxplot.html">plot_boxplot()</a></code> function
which will plot boxplots for each sample in a matrix. Importantly, the
<code><a href="../reference/plot_boxplot.html">plot_boxplot()</a></code> function also allows the user to examine the
relative log expression values for each sample in order to assess
technical factors that may be influencing each library. RLE plots should
show sample distributions centered on zero with similar distributions of
outlier values.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate the log2 counts per million for each sample</span></span>
<span><span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="fu">edgeR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html" class="external-link">cpm</a></span><span class="op">(</span><span class="va">counts</span>, log <span class="op">=</span> <span class="cn">TRUE</span>, prior.count <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Show the distribution of the logcounts for each sample</span></span>
<span><span class="fu"><a href="../reference/plot_boxplot.html">plot_boxplot</a></span><span class="op">(</span><span class="va">logcounts</span>, metadata <span class="op">=</span> <span class="va">samples</span>, fillBy <span class="op">=</span> <span class="st">"pop"</span>, rle <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>             outlier.shape <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Relative Log Expression"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"Sample"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"RLE"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.5</span>, <span class="fl">1.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 5292 rows containing non-finite values</span></span>
<span><span class="co">#&gt; (`stat_boxplot()`).</span></span></code></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>Another diagnostic plot is the <code>plot_distribution()</code>
function. This function creates a density plot for each sample allowing
the user to view the density of the counts for all samples.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_density.html">plot_density</a></span><span class="op">(</span><span class="va">logcounts</span>, metadata <span class="op">=</span> <span class="va">samples</span>, colBy <span class="op">=</span> <span class="st">"pop"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Density plot"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"log2CPM"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>Another useful diagnostic plot is to view the euclidean distances for
the unscaled counts between samples. This plot gives you an idea of how
similar samples are to each other on an absolute level. This can be
accomplished with the <code><a href="../reference/plot_dist.html">plot_dist()</a></code> function.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_dist.html">plot_dist</a></span><span class="op">(</span><span class="va">counts</span>, metadata <span class="op">=</span> <span class="va">samples</span><span class="op">[</span>, <span class="st">"pop"</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="analysis-with-edger">Analysis with <code>edgeR</code><a class="anchor" aria-label="anchor" href="#analysis-with-edger"></a>
</h3>
<p>Downstream analysis of count data can be performed in
<code>edgeR</code>. Since this example data does not contain any actual
groups i will create artificial groups only for the purpose of
demonstrating the steps involved in a simple differential expression
analysis.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add an arbitrary group factor to the samples</span></span>
<span><span class="va">samples</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/gl.html" class="external-link">gl</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Treatment"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Specify group based design</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">group</span>, data <span class="op">=</span> <span class="va">samples</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"group"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Import counts into edger</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">DGEList</a></span><span class="op">(</span><span class="va">counts</span>, samples <span class="op">=</span> <span class="va">samples</span>, group <span class="op">=</span> <span class="va">samples</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Filter for expression</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="va">keep</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Calculate normalization factors</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">y</span>, method <span class="op">=</span> <span class="st">"TMM"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Estimate genewise dispersion</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">y</span>, design <span class="op">=</span> <span class="va">design</span>, robust <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Fit the model</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html" class="external-link">glmQLFit</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span>, robust <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Specify contrasts to test</span></span>
<span><span class="va">con</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html" class="external-link">makeContrasts</a></span><span class="op">(</span>TvC <span class="op">=</span> <span class="va">Treatment</span> <span class="op">-</span> <span class="va">Control</span>, levels <span class="op">=</span> <span class="va">design</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Test for differential expression</span></span>
<span><span class="va">qlf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html" class="external-link">glmQLFTest</a></span><span class="op">(</span><span class="va">fit</span>, contrast <span class="op">=</span> <span class="va">con</span><span class="op">[</span>, <span class="st">"TvC"</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the test results as a data.frame</span></span>
<span><span class="va">qlf_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/edger_to_df.html">edger_to_df</a></span><span class="op">(</span><span class="va">qlf</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h2>
<p>We can perform Principal Components Analysis on the normalized
logcounts using functions from <code>edgeR</code> and
<code>PCAtools</code>. See the <a href="https://bioconductor.org/packages/devel/bioc/vignettes/PCAtools/inst/doc/PCAtools.html" class="external-link">PCAtools
vignette</a> for more useful functions.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate the normalized log counts per million</span></span>
<span><span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html" class="external-link">cpm</a></span><span class="op">(</span><span class="va">y</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform PCA on the 20% most variable logcounts</span></span>
<span><span class="va">pca.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PCAtools/man/pca.html" class="external-link">pca</a></span><span class="op">(</span><span class="va">logcounts</span>, metadata <span class="op">=</span> <span class="va">samples</span>, removeVar <span class="op">=</span> <span class="fl">0.8</span>, scale <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Plot the pca biplot and color by the arbitrary group variable.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/PCAtools/man/biplot.html" class="external-link">biplot</a></span><span class="op">(</span></span>
<span>  <span class="va">pca.res</span>,</span>
<span>  colby <span class="op">=</span> <span class="st">"group"</span>,</span>
<span>  hline <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  vline <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  hlineType <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  vlineType <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  legendPosition <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"PCA on logcounts"</span>,</span>
<span>  caption <span class="op">=</span> <span class="st">"20% Most Variable Features"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="heatmap">Heatmap<a class="anchor" aria-label="anchor" href="#heatmap"></a>
</h2>
<p>The <code>coriell</code> package has a wrapper around
<code>pheatmap()</code> that sets sensible defaults for sequencing data.
Any argument that can be passed to <code>pheatmap()</code> can be passed
to the <code><a href="../reference/quickmap.html">quickmap()</a></code> function to override the defaults. The
<code><a href="../reference/quickmap.html">quickmap()</a></code> function also allows the user to remove low
variance features and fix the scales of the the color so that extreme
values do not wash out the heatmap.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a dataframe of annotations to use</span></span>
<span><span class="va">col_df</span> <span class="op">&lt;-</span> <span class="va">samples</span><span class="op">[</span>, <span class="st">"group"</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Create color scheme for treatment conditions</span></span>
<span><span class="va">ann_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span> <span class="op">=</span> <span class="st">"steelblue"</span>, <span class="st">"Treatment"</span> <span class="op">=</span> <span class="st">"firebrick"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot the heatmap, passing additional args to pheatmap</span></span>
<span><span class="fu"><a href="../reference/quickmap.html">quickmap</a></span><span class="op">(</span></span>
<span>  <span class="va">logcounts</span>,</span>
<span>  removeVar <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  fix_extreme <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  annotation_col <span class="op">=</span> <span class="va">col_df</span>,</span>
<span>  annotation_colors <span class="op">=</span> <span class="va">ann_colors</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"20% Most Variable Genes"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Removing 80% lowest variance features...</span></span></code></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="volcano-plots">Volcano Plots<a class="anchor" aria-label="anchor" href="#volcano-plots"></a>
</h2>
<p>The <code>coriell</code> contains functions for plotting volcano
plots and displaying the counts of up and down regulated genes on the
plots. The aesthetics of the plots produced by
<code><a href="../reference/plot_volcano.html">plot_volcano()</a></code> and <code><a href="../reference/plot_md.html">plot_md()</a></code> below can be
modified. Please check the function documentation for details.</p>
<p>Of course, for this analysis there are no differentially expressed
genes because the groups are arbitrary. However, we can still
demonstrate the functions from the <code>coriell</code> package.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a volcano plot with the default settings</span></span>
<span><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">qlf_df</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Volcano Plot of Differential Expression Results"</span><span class="op">)</span></span></code></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>For a real dataset with differences the plot looks more like:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># GSE161650_de is a dataset built into the coriell package for demonstrations</span></span>
<span><span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">GSE161650_de</span>, fdr <span class="op">=</span> <span class="fl">0.05</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="ma-plots">MA Plots<a class="anchor" aria-label="anchor" href="#ma-plots"></a>
</h2>
<p>Likewise, MA plots can be drawn with <code><a href="../reference/plot_md.html">plot_md()</a></code></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">qlf_df</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"MA Plot of Differential Expression Results"</span><span class="op">)</span></span></code></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>and again for real data:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">GSE161650_de</span>, fdr <span class="op">=</span> <span class="fl">0.05</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="gene-ontology-analysis">Gene ontology analysis<a class="anchor" aria-label="anchor" href="#gene-ontology-analysis"></a>
</h2>
<p>Once you have a vector of genes that is interesting you can see if
those genes are over-represented in a particular pathway. The easiest
way to do this is with the <code>clusterProfiler</code> package.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract a vector of the up-regulated genes</span></span>
<span><span class="va">up_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span></span>
<span>  <span class="va">GSE161650_de</span>, </span>
<span>  <span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span>, </span>
<span>  select <span class="op">=</span> <span class="va">feature_id</span>, </span>
<span>  drop <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract a vector of the down-regulated genes</span></span>
<span><span class="va">down_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span></span>
<span>  <span class="va">GSE161650_de</span>, </span>
<span>  <span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">logFC</span> <span class="op">&lt;</span> <span class="fl">0</span>, </span>
<span>  select <span class="op">=</span> <span class="va">feature_id</span>, </span>
<span>  drop <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Test each set for over-representation of all ontologies</span></span>
<span><span class="va">ego_up</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html" class="external-link">enrichGO</a></span><span class="op">(</span></span>
<span>  <span class="va">up_genes</span>,</span>
<span>  OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>  keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>  ont <span class="op">=</span> <span class="st">"ALL"</span>,</span>
<span>  pool <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  readable <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">ego_down</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html" class="external-link">enrichGO</a></span><span class="op">(</span></span>
<span>  <span class="va">down_genes</span>,</span>
<span>  OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>  keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>  ont <span class="op">=</span> <span class="st">"ALL"</span>,</span>
<span>  pool <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  readable <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We can visualize the over-representation results with dotplots</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html" class="external-link">dotplot</a></span><span class="op">(</span><span class="va">ego_up</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Up-regulated genes"</span><span class="op">)</span></span></code></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html" class="external-link">dotplot</a></span><span class="op">(</span><span class="va">ego_down</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Down-regulated genes"</span><span class="op">)</span></span></code></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-19-2.png" width="700"></p>
<p>To view the results of the over-representation test as a data.frame
simply coerce:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Convert the ego results to a data.frame</span></span>
<span><span class="va">ego_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">ego_up</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ego_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;            ONTOLOGY         ID                        Description GeneRatio</span></span>
<span><span class="co">#&gt; GO:0022626       CC GO:0022626                 cytosolic ribosome   81/2253</span></span>
<span><span class="co">#&gt; GO:0002181       BP GO:0002181            cytoplasmic translation  103/2253</span></span>
<span><span class="co">#&gt; GO:0003735       MF GO:0003735 structural constituent of ribosome   93/2253</span></span>
<span><span class="co">#&gt; GO:0044391       CC GO:0044391                  ribosomal subunit   92/2253</span></span>
<span><span class="co">#&gt; GO:0005840       CC GO:0005840                           ribosome  100/2253</span></span>
<span><span class="co">#&gt; GO:0022625       CC GO:0022625  cytosolic large ribosomal subunit   47/2253</span></span>
<span><span class="co">#&gt;              BgRatio       pvalue     p.adjust       qvalue</span></span>
<span><span class="co">#&gt; GO:0022626 100/20700 4.273439e-60 2.734946e-56 2.099485e-56</span></span>
<span><span class="co">#&gt; GO:0002181 159/20700 6.862242e-60 2.734946e-56 2.099485e-56</span></span>
<span><span class="co">#&gt; GO:0003735 177/20700 4.325713e-43 1.149342e-39 8.822936e-40</span></span>
<span><span class="co">#&gt; GO:0044391 185/20700 4.799591e-40 9.564386e-37 7.342112e-37</span></span>
<span><span class="co">#&gt; GO:0005840 229/20700 3.478621e-37 5.545618e-34 4.257100e-34</span></span>
<span><span class="co">#&gt; GO:0022625  58/20700 2.353191e-35 3.126214e-32 2.399841e-32</span></span>
<span><span class="co">#&gt;                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           geneID</span></span>
<span><span class="co">#&gt; GO:0022626                                                                                                                                       RPL41/RPL18/RPS2/RPL18A/RPLP0/RPL12/RPS17/RPLP1/RPL10/RPS18/RPL39/RPL35/RPL19/RPS11/RPLP2/RPS15/RPS26/RPS9/RPS28/RPS5/RPL13/RPL28/RPS16/RPL30/RPL5/RPS14/RPL24/RPL15/RPL27/RPS13/RPL36A/RPL38/UBA52/RPL32/RPS25/RPS3/RPS10/RPSA/RPS27/RPL21/RPL3/RPL10A/RPS7/RPL26/FAU/RPL7/RPL35A/RPS20/RPL23A/RPS8/RPS15A/RPL23/RPS21/RPL22/RPS19/RACK1/RPL9/RPS24/RPS3A/RPL31/RPL14/RPS29/RPS27A/RPL17/RPL13A/RPL29/RPL34/RPS12/RPS4Y1/RPL36/RPL11/RPS4X/RPS23/RPL4/RPL37/RPL8/RPS6/RPL37A/RPL27A/RPL7A/RPL6</span></span>
<span><span class="co">#&gt; GO:0002181 RPL41/RPL18/RPS2/RPL18A/RPLP0/RPL12/RPS17/RPLP1/RPL10/RPS18/RPL39/RPL35/RPL19/RPS11/RPL22L1/RPLP2/RPS15/RPS26/RPS9/RPS28/RPS5/RPL13/EIF3B/RPL28/RPS16/RPL30/YBX1/RPL5/RPS14/RPL24/RPL15/RPL27/RPS13/RPL36A/RPL38/UBA52/RPL32/RPS25/RPS3/RPS10/RPSA/RWDD1/RPS27/RPL21/RPL3/RPL10A/RPS7/RPL26/FAU/RPL7/EIF3G/RPL35A/RPS20/RPL23A/RPS8/RPS15A/RPL23/RPS21/RPL22/EIF3H/EIF4A1/RPS19/RACK1/RPL9/RPS24/RPS3A/RPL31/RPL14/RPS29/RPS27A/RPL17/RPL13A/RPL29/RPL34/RPS12/RPL36/RPL11/RPS4X/EIF4A2/EIF3F/RPS23/RPL4/PAIP1/RPL37/RPL8/YBX3/EIF3I/RPS6/SH3BGRL/RPL37A/PKM/RBM4/EIF4G1/PABPC1/RPL27A/ZC3H15/RPL7A/RPL6/EIF3L/EIF3E/DHX9/EIF2D/CSDE1</span></span>
<span><span class="co">#&gt; GO:0003735                                                   RPL41/RPL18/RPS2/RPL18A/MRPL41/RPLP0/RPL12/RPS17/RPLP1/RPL10/RPS18/RPL39/RPL35/RPL19/MRPL54/RPS11/RPL22L1/RPLP2/RPS15/RPS26/RPS9/RPS28/RPS5/RPL13/RPL28/RPS16/RPL30/RPL5/RPS14/RPL24/RPL15/RPL27/RPS13/RPL36A/RPL38/UBA52/RPL32/RPS25/RPS3/RPS10/RPSA/RPS27/RPL21/RPL3/RPL10A/RPS7/RPL26/FAU/RPL7/MRPL33/RPL35A/RPS20/RPL23A/RPS8/RPS15A/NDUFA7/RPL23/MRPL34/RPS21/RPL22/RPS19/RPL9/RPS24/RPS3A/RPL31/RPL14/RPS29/RPS27A/RPL17/RPL13A/MRPL44/RPL29/RPL34/RPS12/RPS4Y1/RPL36/RPL11/RPS4X/RSL24D1/RPS23/RPL4/RPL37/RPL8/IMP3/RPS6/RPL37A/RPL27A/MRPL11/RPL7A/RPL6/MRPS2/MRPS24/MRPL23</span></span>
<span><span class="co">#&gt; GO:0044391                                                             RPL41/RPL18/RPS2/RPL18A/MRPL41/RPLP0/RPL12/RPS17/RPLP1/RPL10/RPS18/RPL39/RPL35/RPL19/MRPL54/RPS11/RPLP2/RPS15/RPS26/RPS9/RPS28/RPS5/RPL13/RPL28/RPS16/RPL30/RPL5/RPS14/RPL24/RPL15/RPL27/RPS13/RPL36A/RPL38/UBA52/RPL32/RPS25/RPS3/RPS10/RPSA/RPS27/RPL21/RPL3/RPL10A/RPS7/RPL26/FAU/RPL7/MRPL33/RPL35A/RPS20/RPL23A/RPS8/RPS15A/RPL23/MRPL34/RPS21/RPL22/RPS19/RACK1/RPL9/RPS24/RPS3A/RPL31/RPL14/RPS29/RPS27A/RPL17/RPL13A/MRPL44/RPL29/RPL34/RPS12/RPS4Y1/RPL36/RPL11/RPS4X/RPS23/RPL4/RPL37/RPL8/IMP3/RPS6/RPL37A/RPL27A/MRPL11/RPL7A/RPL6/MRPS2/MRPS24/MRPL23/MRPL39</span></span>
<span><span class="co">#&gt; GO:0005840            RPL41/RPL18/RPS2/RPL18A/MRPL41/RPLP0/RPL12/RPS17/RPLP1/RPL10/RPS18/RPL39/RPL35/RPL19/MRPL54/RPS11/RPL22L1/RPLP2/RPS15/RPS26/RPS9/RPS28/RPS5/RPL13/RPL28/RPS16/RPL30/RPL5/RPS14/RPL24/RPL15/RPL27/RPS13/RPL36A/RPL38/UBA52/RPL32/RPS25/RPS3/RPS10/RPSA/RPS27/RPL21/RPL3/RPL10A/BTF3/RPS7/RPL26/FAU/RPL7/MRPL33/RPL35A/RPS20/RPL23A/RPS8/RPS15A/NDUFA7/RPL23/MRPL34/RPS21/RPL22/EIF3H/RPS19/RACK1/RPL9/RPS24/RPS3A/RPL31/RPL14/RPS29/RPS27A/RPL17/RPL13A/MRPL44/RPL29/RPL34/RPS12/RPS4Y1/RPL36/RPL11/RPS4X/RSL24D1/RPS23/RPL4/RPL37/RPL8/IMP3/RPS6/SERP1/RPL37A/RPL27A/MRPL11/RPL7A/RPL6/MRPS2/MRPS24/SF1/DHX9/MRPL23/MRPL39</span></span>
<span><span class="co">#&gt; GO:0022625                                                                                                                                                                                                                                                                                                                                            RPL41/RPL18/RPL18A/RPLP0/RPL12/RPLP1/RPL10/RPL39/RPL35/RPL19/RPLP2/RPL13/RPL28/RPL30/RPL5/RPL24/RPL15/RPL27/RPL36A/RPL38/UBA52/RPL32/RPL21/RPL3/RPL10A/RPL26/RPL7/RPL35A/RPL23A/RPL23/RPL22/RPL9/RPL31/RPL14/RPL17/RPL13A/RPL29/RPL34/RPL36/RPL11/RPL4/RPL37/RPL8/RPL37A/RPL27A/RPL7A/RPL6</span></span>
<span><span class="co">#&gt;            Count</span></span>
<span><span class="co">#&gt; GO:0022626    81</span></span>
<span><span class="co">#&gt; GO:0002181   103</span></span>
<span><span class="co">#&gt; GO:0003735    93</span></span>
<span><span class="co">#&gt; GO:0044391    92</span></span>
<span><span class="co">#&gt; GO:0005840   100</span></span>
<span><span class="co">#&gt; GO:0022625    47</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="gene-set-enrichment-analysis">Gene set enrichment analysis<a class="anchor" aria-label="anchor" href="#gene-set-enrichment-analysis"></a>
</h2>
<p>We can perform gene set enrichment on a list of genes and their ranks
from a differential expression analysis using the <code>fgsea</code>
package and gene sets from the Molecular Signatures Database.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the msigDB data</span></span>
<span><span class="va">msigdb.hs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/msigdb/man/getMsigdb.html" class="external-link">getMsigdb</a></span><span class="op">(</span>org <span class="op">=</span> <span class="st">"hs"</span>, id <span class="op">=</span> <span class="st">"SYM"</span>, version <span class="op">=</span> <span class="st">"7.4"</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?msigdb and browseVignettes('msigdb') for documentation</span></span>
<span><span class="co">#&gt; downloading 1 resources</span></span>
<span><span class="co">#&gt; retrieving 1 resource</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; require("GSEABase")</span></span>
<span></span>
<span><span class="co"># Subset for the hallmark gene set</span></span>
<span><span class="va">hallmark</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/msigdb/man/subsetCollection.html" class="external-link">subsetCollection</a></span><span class="op">(</span><span class="va">msigdb.hs</span>, <span class="st">"h"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert to a simple list for fgsea</span></span>
<span><span class="va">hallmark</span> <span class="op">&lt;-</span> <span class="fu">geneIds</span><span class="op">(</span><span class="va">hallmark</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a ranking stat vector -- use coriell example data</span></span>
<span><span class="va">stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">GSE161650_de</span>, <span class="va">logFC</span> <span class="op">*</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">PValue</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">stats</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">GSE161650_de</span><span class="op">$</span><span class="va">feature_id</span></span>
<span></span>
<span><span class="co"># Perform GSEA on all of the hallmark sets with fgsea</span></span>
<span><span class="va">fgsea.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/fgsea/man/fgsea.html" class="external-link">fgsea</a></span><span class="op">(</span></span>
<span>  <span class="va">hallmark</span>,</span>
<span>  <span class="va">stats</span>,</span>
<span>  nPermSimple <span class="op">=</span> <span class="fl">1e4</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">0.0</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The results of the GSEA test can be viewed as a table:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">fgsea.results</span><span class="op">)</span></span>
<span><span class="co">#&gt;                         pathway        pval        padj    log2err        ES</span></span>
<span><span class="co">#&gt;                          &lt;char&gt;       &lt;num&gt;       &lt;num&gt;      &lt;num&gt;     &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:        HALLMARK_ADIPOGENESIS 0.001944172 0.008100719 0.45505987 0.4727442</span></span>
<span><span class="co">#&gt; 2: HALLMARK_ALLOGRAFT_REJECTION 0.454743137 0.583004021 0.03219222 0.3315436</span></span>
<span><span class="co">#&gt; 3:   HALLMARK_ANDROGEN_RESPONSE 0.255434783 0.400613328 0.04360878 0.3909382</span></span>
<span><span class="co">#&gt; 4:        HALLMARK_ANGIOGENESIS 0.030353332 0.084709008 0.12694980 0.7526828</span></span>
<span><span class="co">#&gt; 5:     HALLMARK_APICAL_JUNCTION 0.092831640 0.193399249 0.07719384 0.4003922</span></span>
<span><span class="co">#&gt; 6:      HALLMARK_APICAL_SURFACE 0.843459614 0.878603765 0.01949138 0.3317011</span></span>
<span><span class="co">#&gt;          NES  size  leadingEdge</span></span>
<span><span class="co">#&gt;        &lt;num&gt; &lt;int&gt;       &lt;list&gt;</span></span>
<span><span class="co">#&gt; 1: 1.4638970   158 GADD45A,....</span></span>
<span><span class="co">#&gt; 2: 0.9907230   118 IRF7, TN....</span></span>
<span><span class="co">#&gt; 3: 1.0986264    77 SAT1, SG....</span></span>
<span><span class="co">#&gt; 4: 1.5207820    15 MSX1, CX....</span></span>
<span><span class="co">#&gt; 5: 1.2007151   122 ICAM5, F....</span></span>
<span><span class="co">#&gt; 6: 0.7654131    26 HSPB1, L....</span></span></code></pre></div>
<p>Or selected pathways can be plotted with:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/fgsea/man/plotEnrichment.html" class="external-link">plotEnrichment</a></span><span class="op">(</span><span class="va">hallmark</span><span class="op">[[</span><span class="st">"HALLMARK_APOPTOSIS"</span><span class="op">]</span><span class="op">]</span>, <span class="va">stats</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"HALLMARK APOPTOSIS"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Gennaro Calendo.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
