<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>RNA-seq-workflow • coriell</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="RNA-seq-workflow">
<meta property="og:description" content="coriell">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coriell</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/RNA-seq-workflow.html">RNA-seq-workflow</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="RNA-seq-workflow_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>RNA-seq-workflow</h1>
            
      
      
      <div class="hidden name"><code>RNA-seq-workflow.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>Below, I will illustrate some of the functions present in the <code>coriell</code> package by walking through a typical RNA-seq workflow, starting with count files generated using the STAR aligner and progressing through differential expression analysis.</p>
</div>
<div id="load-in-the-data-generated-by-star" class="section level2">
<h2 class="hasAnchor">
<a href="#load-in-the-data-generated-by-star" class="anchor"></a>Load in the data generated by STAR</h2>
<p>The example data were generated by aligning reads from the <a href="https://pubmed.ncbi.nlm.nih.gov/30454645/">Targeting CDK9 Reactivates Epigenetically Silenced Genes in Cancer</a> paper by Dr. Issa’s lab to the hg38 genome using STAR. For brevity, only the day 4 DMSO and HH1 10uM samples have been included. The complete raw data can be found on <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE104837">GEO</a>.</p>
<p>The output from STAR is a tab delimited file with columns containing the gene name, unstranded count, forward stranded count, and reverse stranded count.</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"coriell"</span><span class="op">)</span>
<span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="va">dir</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"*.ReadsPerGene.out.tab"</span><span class="op">)</span>

<span class="va">files</span>
<span class="co">#&gt; [1] "/tmp/RtmpYC9mD7/temp_libpath1b5ea7f7a3ecf/coriell/extdata/DMSO_4day_1_S1.ReadsPerGene.out.tab"    </span>
<span class="co">#&gt; [2] "/tmp/RtmpYC9mD7/temp_libpath1b5ea7f7a3ecf/coriell/extdata/DMSO_4day_2_S2.ReadsPerGene.out.tab"    </span>
<span class="co">#&gt; [3] "/tmp/RtmpYC9mD7/temp_libpath1b5ea7f7a3ecf/coriell/extdata/DMSO_4day_3_S3.ReadsPerGene.out.tab"    </span>
<span class="co">#&gt; [4] "/tmp/RtmpYC9mD7/temp_libpath1b5ea7f7a3ecf/coriell/extdata/HH1_10uM_4day_1_S4.ReadsPerGene.out.tab"</span>
<span class="co">#&gt; [5] "/tmp/RtmpYC9mD7/temp_libpath1b5ea7f7a3ecf/coriell/extdata/HH1_10uM_4day_2_S5.ReadsPerGene.out.tab"</span>
<span class="co">#&gt; [6] "/tmp/RtmpYC9mD7/temp_libpath1b5ea7f7a3ecf/coriell/extdata/HH1_10uM_4day_3_S1.ReadsPerGene.out.tab"</span></pre></div>
<p>Read the files into a single dataframe with <code>vroom</code>. Since the sequencing prep was an Illumina Truseq kit we want to select the unstranded counts for our analysis.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="va">count_df</span> <span class="op">&lt;-</span> <span class="fu">vroom</span><span class="fu">::</span><span class="fu"><a href="https://vroom.r-lib.org/reference/vroom.html">vroom</a></span><span class="op">(</span><span class="va">files</span>,
  id <span class="op">=</span> <span class="st">"fpath"</span>,
  delim <span class="op">=</span> <span class="st">"\t"</span>,
  col_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="st">"unstranded"</span>, <span class="st">"forward"</span>, <span class="st">"reverse"</span><span class="op">)</span>
<span class="op">)</span>
<span class="co">#&gt; Rows: 364,122</span>
<span class="co">#&gt; Columns: 5</span>
<span class="co">#&gt; Delimiter: "\t"</span>
<span class="co">#&gt; chr [1]: gene</span>
<span class="co">#&gt; dbl [3]: unstranded, forward, reverse</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Use `spec()` to retrieve the guessed column specification</span>
<span class="co">#&gt; Pass a specification to the `col_types` argument to quiet this message</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">count_df</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 x 5</span>
<span class="co">#&gt;   fpath                                    gene       unstranded forward reverse</span>
<span class="co">#&gt;   &lt;chr&gt;                                    &lt;chr&gt;           &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt; 1 /tmp/RtmpYC9mD7/temp_libpath1b5ea7f7a3e… N_unmapped     418254  4.18e5  418254</span>
<span class="co">#&gt; 2 /tmp/RtmpYC9mD7/temp_libpath1b5ea7f7a3e… N_multima…    5553623  5.55e6 5553623</span>
<span class="co">#&gt; 3 /tmp/RtmpYC9mD7/temp_libpath1b5ea7f7a3e… N_noFeatu…    1296232  1.42e7 1360388</span>
<span class="co">#&gt; 4 /tmp/RtmpYC9mD7/temp_libpath1b5ea7f7a3e… N_ambiguo…    1828686  1.39e4 1312715</span>
<span class="co">#&gt; 5 /tmp/RtmpYC9mD7/temp_libpath1b5ea7f7a3e… ENSG00000…          0  0.           0</span>
<span class="co">#&gt; 6 /tmp/RtmpYC9mD7/temp_libpath1b5ea7f7a3e… ENSG00000…          0  0.           0</span></pre></div>
<p>We can now clean up the dataframe by extracting the sample names from the fpath column, removing the gene rows that begin with “N_”, and selecting the unstranded counts. To make the resulting dataset a little smaller we can also filter out rows with zero counts.</p>
<p><em>Note the use of the <code><a href="https://rdrr.io/pkg/magrittr/man/compound.html">%&lt;&gt;%</a></code> pipe below which reassigns the results of filtering to the <code>count_df</code> variable</em></p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">magrittr</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tools</span><span class="op">)</span>


<span class="va">count_df</span> <span class="op">%&lt;&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>
    sample_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/tools/fileutils.html">file_path_sans_ext</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">fpath</span><span class="op">)</span><span class="op">)</span>,
    sample_name <span class="op">=</span> <span class="fu">str_remove</span><span class="op">(</span><span class="va">sample_name</span>, <span class="st">"_S[0-9].ReadsPerGene.out"</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu">str_detect</span><span class="op">(</span><span class="va">gene</span>, <span class="st">"^N_"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/methylKit/man/select-methods.html">select</a></span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample_name</span>, <span class="va">unstranded</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">unstranded</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">count_df</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 x 3</span>
<span class="co">#&gt;   gene            sample_name unstranded</span>
<span class="co">#&gt;   &lt;chr&gt;           &lt;chr&gt;            &lt;dbl&gt;</span>
<span class="co">#&gt; 1 ENSG00000268903 DMSO_4day_1          3</span>
<span class="co">#&gt; 2 ENSG00000241860 DMSO_4day_1          1</span>
<span class="co">#&gt; 3 ENSG00000279457 DMSO_4day_1         21</span>
<span class="co">#&gt; 4 ENSG00000237094 DMSO_4day_1          4</span>
<span class="co">#&gt; 5 ENSG00000230021 DMSO_4day_1          1</span>
<span class="co">#&gt; 6 ENSG00000225630 DMSO_4day_1         13</span></pre></div>
</div>
<div id="coerce-the-data-frame-into-a-count-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#coerce-the-data-frame-into-a-count-matrix" class="anchor"></a>Coerce the data frame into a count matrix</h2>
<p>All software dealing with RNA-seq data expects the data to be in the shape where columns represent samples and rows represent genes. To acheive this we will pivot the count_df wider into a count matrix. Additionally, we will turn the gene column into rownames.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="va">count_mat</span> <span class="op">&lt;-</span> <span class="va">count_df</span> <span class="op">%&gt;%</span>
  <span class="fu">pivot_wider</span><span class="op">(</span>
    names_from <span class="op">=</span> <span class="va">sample_name</span>,
    values_from <span class="op">=</span> <span class="va">unstranded</span>,
    values_fill <span class="op">=</span> <span class="fl">0</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">column_to_rownames</span><span class="op">(</span>var <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">count_mat</span><span class="op">)</span>
<span class="co">#&gt;                 DMSO_4day_1 DMSO_4day_2 DMSO_4day_3 HH1_10uM_4day_1</span>
<span class="co">#&gt; ENSG00000268903           3           5           8               6</span>
<span class="co">#&gt; ENSG00000241860           1           3           5               8</span>
<span class="co">#&gt; ENSG00000279457          21          32          46              24</span>
<span class="co">#&gt; ENSG00000237094           4           7           9               6</span>
<span class="co">#&gt; ENSG00000230021           1           1           1               2</span>
<span class="co">#&gt; ENSG00000225630          13           9          21              25</span>
<span class="co">#&gt;                 HH1_10uM_4day_2 HH1_10uM_4day_3</span>
<span class="co">#&gt; ENSG00000268903               2               5</span>
<span class="co">#&gt; ENSG00000241860               9              15</span>
<span class="co">#&gt; ENSG00000279457              50              47</span>
<span class="co">#&gt; ENSG00000237094              11              17</span>
<span class="co">#&gt; ENSG00000230021               3               1</span>
<span class="co">#&gt; ENSG00000225630              45              47</span></pre></div>
</div>
<div id="create-sample-annotation-information" class="section level2">
<h2 class="hasAnchor">
<a href="#create-sample-annotation-information" class="anchor"></a>Create sample annotation information</h2>
<p>We can either read in metadata about our samples from a file that we created externally or we can create sample metadata from the sample names themselves. Below, we will create sample metadata from the colnames of the count_mat.</p>
<p>Since there is only one time point present we will ignore extracting the time from the sample name and simply compare the two treatment groups.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="va">sample_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">count_mat</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">enframe</span><span class="op">(</span>name <span class="op">=</span> <span class="cn">NULL</span>, value <span class="op">=</span> <span class="st">"sample_name"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">mutate</span><span class="op">(</span>
    group <span class="op">=</span> <span class="fu">if_else</span><span class="op">(</span><span class="fu">str_detect</span><span class="op">(</span><span class="va">sample_name</span>, <span class="st">"DMSO"</span><span class="op">)</span>, <span class="st">"DMSO"</span>, <span class="st">"HH1_10uM"</span><span class="op">)</span>,
    group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">group</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"DMSO"</span>, <span class="st">"HH1_10uM"</span><span class="op">)</span><span class="op">)</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">column_to_rownames</span><span class="op">(</span>var <span class="op">=</span> <span class="st">"sample_name"</span><span class="op">)</span>

<span class="co"># it is important that the colnames of the count_mat match the rownames of samples_df</span>
<span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">count_mat</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sample_df</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">sample_df</span><span class="op">)</span>
<span class="co">#&gt;                    group</span>
<span class="co">#&gt; DMSO_4day_1         DMSO</span>
<span class="co">#&gt; DMSO_4day_2         DMSO</span>
<span class="co">#&gt; DMSO_4day_3         DMSO</span>
<span class="co">#&gt; HH1_10uM_4day_1 HH1_10uM</span>
<span class="co">#&gt; HH1_10uM_4day_2 HH1_10uM</span>
<span class="co">#&gt; HH1_10uM_4day_3 HH1_10uM</span></pre></div>
</div>
<div id="read-in-information-about-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#read-in-information-about-genes" class="anchor"></a>Read in information about genes</h2>
<p>This step is not strictly necessary but it can be useful to read in information about the genes. For example, since the genes in the count matrix have Ensembl IDs we might want to know their gene symbol or we might want to calculate RPKMs in which case we would need information about gene lengths.</p>
<p>These samples were aligned to the Ensembl hg38.100 build using the matching GTF file. Therefore, we can use this GTF file to annotate our genes. We will use <code>rtracklayer</code> to import the GTF file we download from the Ensembl FTP server. <code>rtracklayer</code> imports the file as a <code>GRanges</code> object so we will coerce it to a dataframe so we can filter.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="co"># get the GTF file from Ensembl</span>
<span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"ftp://ftp.ensembl.org/pub/release-100/gtf/homo_sapiens/Homo_sapiens.GRCh38.100.gtf.gz"</span>
<span class="va">gtf_file</span> <span class="op">&lt;-</span> <span class="st">"~/Downloads/Homo_sapiens.GRCh38.100.gtf.gz"</span>

<span class="co"># if the file isn't in the Downloads folder, download it</span>
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">gtf_file</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="va">url</span>, <span class="va">gtf_file</span><span class="op">)</span>

<span class="co"># load the gtf file as a dataframe</span>
<span class="va">gtf_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/export.html">import</a></span><span class="op">(</span><span class="va">gtf_file</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">gtf_df</span><span class="op">)</span>
<span class="co">#&gt;   seqnames start   end width strand source       type score phase</span>
<span class="co">#&gt; 1        1 11869 14409  2541      + havana       gene    NA    NA</span>
<span class="co">#&gt; 2        1 11869 14409  2541      + havana transcript    NA    NA</span>
<span class="co">#&gt; 3        1 11869 12227   359      + havana       exon    NA    NA</span>
<span class="co">#&gt; 4        1 12613 12721   109      + havana       exon    NA    NA</span>
<span class="co">#&gt; 5        1 13221 14409  1189      + havana       exon    NA    NA</span>
<span class="co">#&gt; 6        1 12010 13670  1661      + havana transcript    NA    NA</span>
<span class="co">#&gt;           gene_id gene_version gene_name gene_source</span>
<span class="co">#&gt; 1 ENSG00000223972            5   DDX11L1      havana</span>
<span class="co">#&gt; 2 ENSG00000223972            5   DDX11L1      havana</span>
<span class="co">#&gt; 3 ENSG00000223972            5   DDX11L1      havana</span>
<span class="co">#&gt; 4 ENSG00000223972            5   DDX11L1      havana</span>
<span class="co">#&gt; 5 ENSG00000223972            5   DDX11L1      havana</span>
<span class="co">#&gt; 6 ENSG00000223972            5   DDX11L1      havana</span>
<span class="co">#&gt;                         gene_biotype   transcript_id transcript_version</span>
<span class="co">#&gt; 1 transcribed_unprocessed_pseudogene            &lt;NA&gt;               &lt;NA&gt;</span>
<span class="co">#&gt; 2 transcribed_unprocessed_pseudogene ENST00000456328                  2</span>
<span class="co">#&gt; 3 transcribed_unprocessed_pseudogene ENST00000456328                  2</span>
<span class="co">#&gt; 4 transcribed_unprocessed_pseudogene ENST00000456328                  2</span>
<span class="co">#&gt; 5 transcribed_unprocessed_pseudogene ENST00000456328                  2</span>
<span class="co">#&gt; 6 transcribed_unprocessed_pseudogene ENST00000450305                  2</span>
<span class="co">#&gt;   transcript_name transcript_source                 transcript_biotype   tag</span>
<span class="co">#&gt; 1            &lt;NA&gt;              &lt;NA&gt;                               &lt;NA&gt;  &lt;NA&gt;</span>
<span class="co">#&gt; 2     DDX11L1-202            havana               processed_transcript basic</span>
<span class="co">#&gt; 3     DDX11L1-202            havana               processed_transcript basic</span>
<span class="co">#&gt; 4     DDX11L1-202            havana               processed_transcript basic</span>
<span class="co">#&gt; 5     DDX11L1-202            havana               processed_transcript basic</span>
<span class="co">#&gt; 6     DDX11L1-201            havana transcribed_unprocessed_pseudogene basic</span>
<span class="co">#&gt;   transcript_support_level exon_number         exon_id exon_version protein_id</span>
<span class="co">#&gt; 1                     &lt;NA&gt;        &lt;NA&gt;            &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt; 2                        1        &lt;NA&gt;            &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt; 3                        1           1 ENSE00002234944            1       &lt;NA&gt;</span>
<span class="co">#&gt; 4                        1           2 ENSE00003582793            1       &lt;NA&gt;</span>
<span class="co">#&gt; 5                        1           3 ENSE00002312635            1       &lt;NA&gt;</span>
<span class="co">#&gt; 6                       NA        &lt;NA&gt;            &lt;NA&gt;         &lt;NA&gt;       &lt;NA&gt;</span>
<span class="co">#&gt;   protein_version ccds_id</span>
<span class="co">#&gt; 1            &lt;NA&gt;    &lt;NA&gt;</span>
<span class="co">#&gt; 2            &lt;NA&gt;    &lt;NA&gt;</span>
<span class="co">#&gt; 3            &lt;NA&gt;    &lt;NA&gt;</span>
<span class="co">#&gt; 4            &lt;NA&gt;    &lt;NA&gt;</span>
<span class="co">#&gt; 5            &lt;NA&gt;    &lt;NA&gt;</span>
<span class="co">#&gt; 6            &lt;NA&gt;    &lt;NA&gt;</span></pre></div>
<p>Next, we will filter the gtf_df to include only the Ensembl IDs present in our count_mat. We can left_join the GTF information from genes onto our gene_ids and then select only relevant columns.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="va">gene_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">count_mat</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">enframe</span><span class="op">(</span>name <span class="op">=</span> <span class="cn">NULL</span>, value <span class="op">=</span> <span class="st">"gene_id"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">left_join</span><span class="op">(</span>y <span class="op">=</span> <span class="op">{</span>
    <span class="va">gtf_df</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">type</span> <span class="op">==</span> <span class="st">"gene"</span><span class="op">)</span>
  <span class="op">}</span>, by <span class="op">=</span> <span class="st">"gene_id"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/methylKit/man/select-methods.html">select</a></span><span class="op">(</span><span class="va">gene_id</span>,
    chr <span class="op">=</span> <span class="va">seqnames</span>,
    <span class="va">start</span>,
    <span class="va">end</span>,
    length <span class="op">=</span> <span class="va">width</span>,
    <span class="va">strand</span>,
    <span class="va">gene_name</span>,
    <span class="va">gene_biotype</span>
  <span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">column_to_rownames</span><span class="op">(</span>var <span class="op">=</span> <span class="st">"gene_id"</span><span class="op">)</span>

<span class="co"># check to see if rownames of gene_df match those of count_mat</span>
<span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">count_mat</span><span class="op">)</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gene_df</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">gene_df</span><span class="op">)</span>
<span class="co">#&gt;                 chr  start    end length strand  gene_name</span>
<span class="co">#&gt; ENSG00000268903   1 135141 135895    755      - AL627309.6</span>
<span class="co">#&gt; ENSG00000241860   1 141474 173862  32389      - AL627309.5</span>
<span class="co">#&gt; ENSG00000279457   1 185217 195411  10195      -     WASH9P</span>
<span class="co">#&gt; ENSG00000237094   1 365389 522928 157540      - AL732372.2</span>
<span class="co">#&gt; ENSG00000230021   1 586071 827796 241726      - AL669831.3</span>
<span class="co">#&gt; ENSG00000225630   1 629640 630683   1044      +   MTND2P28</span>
<span class="co">#&gt;                                       gene_biotype</span>
<span class="co">#&gt; ENSG00000268903               processed_pseudogene</span>
<span class="co">#&gt; ENSG00000241860                             lncRNA</span>
<span class="co">#&gt; ENSG00000279457             unprocessed_pseudogene</span>
<span class="co">#&gt; ENSG00000237094 transcribed_unprocessed_pseudogene</span>
<span class="co">#&gt; ENSG00000230021   transcribed_processed_pseudogene</span>
<span class="co">#&gt; ENSG00000225630             unprocessed_pseudogene</span></pre></div>
</div>
<div id="load-data-into-edger" class="section level2">
<h2 class="hasAnchor">
<a href="#load-data-into-edger" class="anchor"></a>Load data into edgeR</h2>
<p>We can now load our data into <code>edgeR</code> and start our analysis. <strong>NOTE</strong> the naming of some of the variable names above was deliberate. <code>edgeR</code> expects groups to be named ‘group’ as a column in your sample annotation, otherwise you must specify group names in <code>edgeR</code> functions. Likewise, in the gene annotation <code>edgeR</code> looks for a column named ‘length’ to calculate RPKMs. You don’t have to name your columns with these conventions however it makes working with <code>edgeR</code> a little easier.</p>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioinf.wehi.edu.au/edgeR">edgeR</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: limma</span>

<span class="co"># load data into an edger object</span>
<span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html">DGEList</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">count_mat</span>,
             samples <span class="op">=</span> <span class="va">sample_df</span>,
             genes <span class="op">=</span> <span class="va">gene_df</span><span class="op">)</span></pre></div>
</div>
<div id="specify-study-design" class="section level2">
<h2 class="hasAnchor">
<a href="#specify-study-design" class="anchor"></a>Specify study design</h2>
<p>Since we’re only comparing two groups the design is straightforward. We can specify the design matrix by either setting an intercept to be the first factor level <code>design &lt;- model.matrix(~sample_df$group)</code> or by setting each group as a column in the resulting design matrix and using contrasts to make comparisons. I am a fan of setting up contrasts because they are generally easier to understand and are easier to expand upon if the experimental design gets more complicated.</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fl">0</span> <span class="op">+</span> <span class="va">sample_df</span><span class="op">$</span><span class="va">group</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sample_df</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">sample_df</span><span class="op">$</span><span class="va">group</span><span class="op">)</span>

<span class="va">design</span>
<span class="co">#&gt;                 DMSO HH1_10uM</span>
<span class="co">#&gt; DMSO_4day_1        1        0</span>
<span class="co">#&gt; DMSO_4day_2        1        0</span>
<span class="co">#&gt; DMSO_4day_3        1        0</span>
<span class="co">#&gt; HH1_10uM_4day_1    0        1</span>
<span class="co">#&gt; HH1_10uM_4day_2    0        1</span>
<span class="co">#&gt; HH1_10uM_4day_3    0        1</span>
<span class="co">#&gt; attr(,"assign")</span>
<span class="co">#&gt; [1] 1 1</span>
<span class="co">#&gt; attr(,"contrasts")</span>
<span class="co">#&gt; attr(,"contrasts")$`sample_df$group`</span>
<span class="co">#&gt; [1] "contr.treatment"</span></pre></div>
</div>
<div id="filter-genes-by-their-expression" class="section level2">
<h2 class="hasAnchor">
<a href="#filter-genes-by-their-expression" class="anchor"></a>Filter genes by their expression</h2>
<p>Since a gene has to be transcribed a certain number of times to really be expressed we want to filter out genes that are lowly expressed. <code>edgeR</code> accomplishes this with the <code>filterByExpr</code> function.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html">filterByExpr</a></span><span class="op">(</span><span class="va">y</span>, design <span class="op">=</span> <span class="va">design</span><span class="op">)</span>
<span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="va">keep</span>, <span class="op">]</span>

<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">keep</span><span class="op">)</span>
<span class="co">#&gt; keep</span>
<span class="co">#&gt; FALSE  TRUE </span>
<span class="co">#&gt; 21679 16841</span></pre></div>
</div>
<div id="calculate-normalization-factors" class="section level2">
<h2 class="hasAnchor">
<a href="#calculate-normalization-factors" class="anchor"></a>Calculate normalization factors</h2>
<p><code>edgeR</code> uses TMM to calculate normalization factors. These norm factors are then used in downstream analysis to normalize the gene counts within and between samples.</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html">calcNormFactors</a></span><span class="op">(</span><span class="va">y</span>, method <span class="op">=</span> <span class="st">"TMM"</span><span class="op">)</span></pre></div>
<p>We can check the normalization by creating an MD plot for each sample vs the expression levels in all other samples. <code>edgeR</code> has a function <code>plotMD</code> for this purpose. We can wrap this function in another caller and map it across all of our samples to easily check the entire dataset.</p>
<p>The bulk of the data should be zero-centered and generally horizontal across. Since this is only a subset of the entire dataset it is not surprising that there is some skew in the normalization.</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="va">plot_meanDifference</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">col</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu">plotMD</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html">cpm</a></span><span class="op">(</span><span class="va">y</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, column <span class="op">=</span> <span class="va">col</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fl">0</span>, col <span class="op">=</span> <span class="st">"red"</span>, lty <span class="op">=</span> <span class="fl">2</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="op">}</span>
<span class="co"># apply plot func to all columns</span>
<span class="fu">walk</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span>, <span class="va">plot_meanDifference</span><span class="op">)</span></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-13-1.png" width="700"><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-13-2.png" width="700"><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-13-3.png" width="700"><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-13-4.png" width="700"><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-13-5.png" width="700"><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-13-6.png" width="700"></p>
</div>
<div id="estimate-dispersion" class="section level2">
<h2 class="hasAnchor">
<a href="#estimate-dispersion" class="anchor"></a>Estimate dispersion</h2>
<p>Next we estimate the dispersion. The dispersion values are critical for estimating DE. You need at least two replicates in a single group to calculate dispersions. Of course, however, more groups and more replicates (at least triplicates) will improve the analysis.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html">estimateDisp</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span>, robust <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></pre></div>
<div id="plot-bcv" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-bcv" class="anchor"></a>Plot BCV</h3>
<p>We can plot the dispersion estimates below. <code>edgeR</code> calculates dispersion in three ways which may or may not be used depending on the DE test you choose. The tagwise dispersion is the dispersion for any individual gene. The trended dispersion is a smoothed dispersion estimate for genes of a given expression level and the common dispersion is essentially an estimate of the mean dispersion of all genes.</p>
<p>It is typical to see dispersions higher and more variable for lower expressed genes and then asymptotically get closer to zero as expression increases.</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/plotBCV.html">plotBCV</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
</div>
</div>
<div id="fit-ql-model" class="section level2">
<h2 class="hasAnchor">
<a href="#fit-ql-model" class="anchor"></a>Fit QL model</h2>
<p>Depending on the test to be used you will typically need to fit a model to the data. <code>edgeR</code> does this with the <code>glmQLFit</code> framework. Running this function fits a model to the expression data that can then be used downstream in DE testing.</p>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html">glmQLFit</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span>, robust <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></pre></div>
<div id="plot-ql-dispersions" class="section level3">
<h3 class="hasAnchor">
<a href="#plot-ql-dispersions" class="anchor"></a>Plot QL dispersions</h3>
<p>The <code>glmQLFit</code> function is similar to <code>glmFit</code> but uses <code>squeezeVar</code> from <code>limma</code> to conduct empirical Bayes moderation of the genewise QL dispersions. These squeezed estimates help to provide better estimates in the final DE test. We can view the fitted estimates with the <code>plotQLDisp</code> function.</p>
<div class="sourceCode" id="cb16"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/plotQLDisp.html">plotQLDisp</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
</div>
</div>
<div id="specify-contrasts-for-de-testing" class="section level2">
<h2 class="hasAnchor">
<a href="#specify-contrasts-for-de-testing" class="anchor"></a>Specify contrasts for DE testing</h2>
<p>We can now use the design matrix to select the columns we want to compare. Below we are looking for the differential gene expression in HH1_10uM vs DMSO. Positive logFCs in the results indicate that a gene is up-regulated in HH1_10uM whereas negative values for the logFC indicate the gene was down-regulated in HH1_10uM.</p>
<div class="sourceCode" id="cb17"><pre class="downlit">
<span class="va">con</span> <span class="op">&lt;-</span> <span class="fu">makeContrasts</span><span class="op">(</span>hh1_vs_dmso <span class="op">=</span> <span class="va">HH1_10uM</span> <span class="op">-</span> <span class="va">DMSO</span>, levels <span class="op">=</span> <span class="va">design</span><span class="op">)</span></pre></div>
</div>
<div id="perform-de-testing" class="section level2">
<h2 class="hasAnchor">
<a href="#perform-de-testing" class="anchor"></a>Perform DE testing</h2>
<p>Below, we will use the <code>glmTreat</code> function to test for DE relative to a logFC cutoff. Moderate values for the <code>lfc</code> parameter in the <code>glmTreat</code> function typically reflect testing for a roughly 2 fold-change. This method is better than performing a LRT or <code>exactTest</code> and then filtering for FDR and logFCs.</p>
<p>This is also the moment where the <code>coriell</code> functions become helpful. We will now attach the <code>coriell</code> package and start using some of the functions.</p>
<div class="sourceCode" id="cb18"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">coriell</span><span class="op">)</span>


<span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmTreat.html">glmTreat</a></span><span class="op">(</span><span class="va">fit</span>, contrast <span class="op">=</span> <span class="va">con</span><span class="op">[</span>, <span class="st">"hh1_vs_dmso"</span><span class="op">]</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">1.2</span><span class="op">)</span><span class="op">)</span>

<span class="co"># convert the edger result to a dataframe</span>
<span class="va">res_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/edger_to_df.html">edger_to_df</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">res_df</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 x 13</span>
<span class="co">#&gt;   feature_id chr    start    end length strand gene_name gene_biotype logFC</span>
<span class="co">#&gt;   &lt;chr&gt;      &lt;fct&gt;  &lt;int&gt;  &lt;int&gt;  &lt;int&gt; &lt;fct&gt;  &lt;chr&gt;     &lt;chr&gt;        &lt;dbl&gt;</span>
<span class="co">#&gt; 1 ENSG00000… 1     1.46e8 1.46e8   4145 -      TXNIP     protein_cod…  4.91</span>
<span class="co">#&gt; 2 ENSG00000… 3     4.86e7 4.86e7  31195 -      COL7A1    protein_cod…  3.22</span>
<span class="co">#&gt; 3 ENSG00000… 12    5.37e7 5.37e7  19229 -      CALCOCO1  protein_cod…  2.43</span>
<span class="co">#&gt; 4 ENSG00000… 9     1.15e8 1.15e8  98683 -      TNC       protein_cod…  3.23</span>
<span class="co">#&gt; 5 ENSG00000… 17    1.97e7 1.97e7  10960 -      ALDH3A1   protein_cod…  2.36</span>
<span class="co">#&gt; 6 ENSG00000… 11    6.39e6 6.39e6   4559 +      SMPD1     protein_cod…  2.80</span>
<span class="co">#&gt; # … with 4 more variables: unshrunk.logFC &lt;dbl&gt;, logCPM &lt;dbl&gt;, PValue &lt;dbl&gt;,</span>
<span class="co">#&gt; #   FDR &lt;dbl&gt;</span></pre></div>
</div>
<div id="create-plots-from-de-testing" class="section level2">
<h2 class="hasAnchor">
<a href="#create-plots-from-de-testing" class="anchor"></a>Create plots from DE testing</h2>
<p>We can now use some of the <code>coriell</code> functions to create plots. For example, we can use the <code>plot_volcano</code> function to create a volcano plot of the results. The <code>plot_volcano</code> function automatically summarizes and annotates the plot with the counts of the up and down regulated genes. The <code>plot_volcano</code> and the <code>plot_md</code> functions both return <code>ggplot</code> objects so you can always modify them, for example to add a title.</p>
<p>The plotting functions also have options to adjust the positioning of the labels or remove them entirely. The plot functions can also be used the label the genes using <code>ggrepel</code>. Use <code><a href="../reference/plot_volcano.html">?plot_volcano</a></code> and <code><a href="../reference/plot_md.html">?plot_md</a></code> to see all of the functionality.</p>
<div class="sourceCode" id="cb19"><pre class="downlit">
<span class="co"># summarize the results into a nice table</span>
<span class="fu"><a href="../reference/summarize_dge.html">summarize_dge</a></span><span class="op">(</span><span class="va">res_df</span>, fdr <span class="op">=</span> <span class="fl">0.1</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">1.2</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 3 x 3</span>
<span class="co">#&gt;   dge         n  perc</span>
<span class="co">#&gt;   &lt;fct&gt;   &lt;int&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1 up       4857  28.8</span>
<span class="co">#&gt; 2 down     3845  22.8</span>
<span class="co">#&gt; 3 non-dge  8139  48.3</span>

<span class="co"># create a volcano plot</span>
<span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">res_df</span>, fdr <span class="op">=</span> <span class="fl">0.1</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">1.2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"HH1 vs DMSO at 4 Days"</span><span class="op">)</span></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<div class="sourceCode" id="cb20"><pre class="downlit">

<span class="co"># create an md plot</span>
<span class="fu"><a href="../reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">res_df</span>, fdr <span class="op">=</span> <span class="fl">0.1</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">1.2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"HH1 vs DMSO at 4 Days"</span><span class="op">)</span></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-20-2.png" width="700"></p>
</div>
<div id="create-a-heatmap-from-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#create-a-heatmap-from-the-data" class="anchor"></a>Create a heatmap from the data</h2>
<p>Another useful way of looking at expression data is to compare gene expression on a heatmap. The <code><a href="../reference/quickmap.html">coriell::quickmap</a></code> function provides a wrapper around <code>pheatmap</code> with some defaults that are useful for RNA-seq data. Any parameter that <code>pheatmap</code> accepts can be passed into the <code>quickmap</code> function as well.</p>
<div class="sourceCode" id="cb21"><pre class="downlit">
<span class="co"># calculate the logCPMs with edger</span>
<span class="va">lcpms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html">cpm</a></span><span class="op">(</span><span class="va">y</span>, log <span class="op">=</span> <span class="cn">TRUE</span>, prior.count <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>

<span class="co"># plot the heatmap with quickmap</span>
<span class="fu"><a href="../reference/quickmap.html">quickmap</a></span><span class="op">(</span><span class="va">lcpms</span>, main <span class="op">=</span> <span class="st">"All Genes"</span><span class="op">)</span></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<div class="sourceCode" id="cb22"><pre class="downlit">

<span class="co"># plotting only significant genes</span>
<span class="va">sig_genes</span> <span class="op">&lt;-</span> <span class="va">res_df</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">pull</span><span class="op">(</span><span class="va">feature_id</span><span class="op">)</span>

<span class="co"># subset the lcpms matrix to include only genes found to be DE</span>
<span class="va">sig_lcpms</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">lcpms</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">lcpms</span><span class="op">)</span> <span class="op">%in%</span> <span class="va">sig_genes</span><span class="op">)</span>

<span class="fu"><a href="../reference/quickmap.html">quickmap</a></span><span class="op">(</span><span class="va">sig_lcpms</span>, main <span class="op">=</span> <span class="st">"Significant Genes"</span><span class="op">)</span></pre></div>
<p><img src="RNA-seq-workflow_files/figure-html/unnamed-chunk-21-2.png" width="700"></p>
</div>
<div id="perform-gene-ontology-search-on-de-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#perform-gene-ontology-search-on-de-genes" class="anchor"></a>Perform Gene Ontology search on DE genes</h2>
<p>Another common task is to perform gene ontology on the differentially expressed genes. <code><a href="../reference/panther_go.html">coriell::panther_go</a></code> provides a function that takes in a list of gene ids and submits a request to the PANTHER db for GO analysis. It then returns the parsed json results as a tibble.</p>
<p>The <code>panther_go</code> function accepts any gene id that PANTHER accepts. The only arguments that must be provided are the list of genes, the taxon id, and the GO terms you want to search for. For a complete listing of the options see <code><a href="../reference/panther_go.html">?coriell::panther_go</a></code></p>
<div class="sourceCode" id="cb23"><pre class="downlit">
<span class="co"># extract the top 100 up-regulated ensembl ids</span>
<span class="va">up_ids</span> <span class="op">&lt;-</span> <span class="va">res_df</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.1</span> <span class="op">&amp;</span> <span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">slice_max</span><span class="op">(</span>n <span class="op">=</span> <span class="fl">100</span>, order_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">pull</span><span class="op">(</span><span class="va">feature_id</span><span class="op">)</span>

<span class="co"># extract the top 100 down-regulated ensembl gene ids</span>
<span class="va">down_ids</span> <span class="op">&lt;-</span> <span class="va">res_df</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.1</span> <span class="op">&amp;</span> <span class="va">logFC</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">slice_max</span><span class="op">(</span>n <span class="op">=</span> <span class="fl">100</span>, order_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">pull</span><span class="op">(</span><span class="va">feature_id</span><span class="op">)</span>

<span class="co"># perform GO analysis with PANTHER on up-genes</span>
<span class="va">panther_up</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/panther_go.html">panther_go</a></span><span class="op">(</span><span class="va">up_ids</span>, <span class="st">"9606"</span>, <span class="st">"biological_process"</span><span class="op">)</span>

<span class="co"># perform GO analysis with PANTHER on down-genes</span>
<span class="va">panther_down</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/panther_go.html">panther_go</a></span><span class="op">(</span><span class="va">down_ids</span>, <span class="st">"9606"</span>, <span class="st">"biological_process"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">panther_up</span><span class="op">$</span><span class="va">table</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 x 10</span>
<span class="co">#&gt;   result_number number_in_list fold_enrichment     fdr expected number_in_refer…</span>
<span class="co">#&gt;   &lt;chr&gt;                  &lt;int&gt;           &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;            &lt;int&gt;</span>
<span class="co">#&gt; 1 1                         10            8.55 0.00471  1.17                 364</span>
<span class="co">#&gt; 2 2                          9            7.72 0.0221   1.17                 363</span>
<span class="co">#&gt; 3 3                          2          207.   0.531    0.00964                3</span>
<span class="co">#&gt; 4 4                          2          207.   0.398    0.00964                3</span>
<span class="co">#&gt; 5 5                         43            1.57 0.482   27.4                 8527</span>
<span class="co">#&gt; 6 6                          8            5.21 0.416    1.54                 478</span>
<span class="co">#&gt; # … with 4 more variables: pValue &lt;dbl&gt;, plus_minus &lt;chr&gt;, GO_term &lt;chr&gt;,</span>
<span class="co">#&gt; #   description &lt;chr&gt;</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">panther_down</span><span class="op">$</span><span class="va">table</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 x 10</span>
<span class="co">#&gt;   result_number number_in_list fold_enrichment     fdr expected number_in_refer…</span>
<span class="co">#&gt;   &lt;chr&gt;                  &lt;int&gt;           &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;            &lt;int&gt;</span>
<span class="co">#&gt; 1 1                         12            16.7 2.09e-7    0.718              192</span>
<span class="co">#&gt; 2 2                         10            23.0 3.08e-7    0.434              116</span>
<span class="co">#&gt; 3 3                         13            11.2 1.01e-6    1.16               309</span>
<span class="co">#&gt; 4 4                         10            18.2 1.36e-6    0.550              147</span>
<span class="co">#&gt; 5 5                         11            14.5 1.34e-6    0.759              203</span>
<span class="co">#&gt; 6 6                         10            17.0 1.66e-6    0.587              157</span>
<span class="co">#&gt; # … with 4 more variables: pValue &lt;dbl&gt;, plus_minus &lt;chr&gt;, GO_term &lt;chr&gt;,</span>
<span class="co">#&gt; #   description &lt;chr&gt;</span></pre></div>
<p>The function can also work with the gene names.</p>
<div class="sourceCode" id="cb24"><pre class="downlit">
<span class="va">up_genes</span> <span class="op">&lt;-</span> <span class="va">res_df</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.1</span> <span class="op">&amp;</span> <span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">slice_max</span><span class="op">(</span>n <span class="op">=</span> <span class="fl">100</span>, order_by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">distinct</span><span class="op">(</span><span class="va">gene_name</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">pull</span><span class="op">(</span><span class="va">gene_name</span><span class="op">)</span>
  
<span class="va">panther_up2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/panther_go.html">panther_go</a></span><span class="op">(</span><span class="va">up_genes</span>, <span class="st">"9606"</span>, <span class="st">"biological_process"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">panther_up2</span><span class="op">$</span><span class="va">table</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 6 x 10</span>
<span class="co">#&gt;   result_number number_in_list fold_enrichment    fdr expected number_in_refer…</span>
<span class="co">#&gt;   &lt;chr&gt;                  &lt;int&gt;           &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt;            &lt;int&gt;</span>
<span class="co">#&gt; 1 1                         10            7.25 0.0225   1.38                364</span>
<span class="co">#&gt; 2 2                          9            6.54 0.0877   1.38                363</span>
<span class="co">#&gt; 3 3                         50            1.55 0.407   32.3                8527</span>
<span class="co">#&gt; 4 4                          2          176.   0.554    0.0114                3</span>
<span class="co">#&gt; 5 5                          2          176.   0.443    0.0114                3</span>
<span class="co">#&gt; 6 6                         35            1.74 1       20.1                5301</span>
<span class="co">#&gt; # … with 4 more variables: pValue &lt;dbl&gt;, plus_minus &lt;chr&gt;, GO_term &lt;chr&gt;,</span>
<span class="co">#&gt; #   description &lt;chr&gt;</span></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Gennaro Calendo.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
