<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>RNAseq Workflow • coriell</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="RNAseq Workflow">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">coriell</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.15.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/Practical-analysis-with-SummarizedExperiments.html">Practical analysis with SummarizedExperiments</a></li>
    <li><a class="dropdown-item" href="../articles/rnaseq-workflow.html">RNAseq Workflow</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/coriell-research/coriell/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>RNAseq Workflow</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/coriell-research/coriell/blob/HEAD/vignettes/articles/rnaseq-workflow.Rmd" class="external-link"><code>vignettes/articles/rnaseq-workflow.Rmd</code></a></small>
      <div class="d-none name"><code>rnaseq-workflow.Rmd</code></div>
    </div>

    
    
<p>This vignette outlines some common steps for RNA-seq analysis
highlighting functions present in the <code>coriell</code> package. In
order to illustrate some of the analysis steps I will borrow examples
and data from the <a href="https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html" class="external-link">rnaseqGene</a>
and <a href="https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html" class="external-link">RNAseq123</a>
Bioconductor workflows. Please check out the above workflows for more
details regarding RNA-seq analysis.</p>
<p>This vignette contains my opinionated notes on performing RNA-seq
analyses. I try to closely follow best practices from package authors
but if any information is out of date or incorrect, please let me
know.</p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>Differential gene expression analysis using RNA-seq typically
consists of several steps:</p>
<ol style="list-style-type: decimal">
<li>Quality control of the fastq files with a tool like <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/" class="external-link">FastQC</a>
or <a href="https://github.com/OpenGene/fastp" class="external-link">fastp</a>
</li>
<li>Alignment of fastq files to a reference genome using a splice-aware
aligner like <a href="https://github.com/alexdobin/STAR" class="external-link">STAR</a>
<em>or</em> transcript quantification using a pseudoaligner like <a href="https://combine-lab.github.io/salmon/" class="external-link">Salmon</a>.</li>
<li>If using a genome aligner, read counting with <a href="https://bioconductor.org/packages/release/bioc/html/Rsubread.html" class="external-link">Rsubread::featureCounts</a>
or <a href="https://htseq.readthedocs.io/en/release_0.11.1/count.html" class="external-link">HTSeq
count</a> to generate gene counts. If using a transcript aligner,
importing gene-level counts using the appropriate offsets with <a href="https://bioconductor.org/packages/release/bioc/html/tximport.html" class="external-link">tximport</a>
or <a href="https://bioconductor.org/packages/release/bioc/html/tximeta.html" class="external-link">tximeta</a>
</li>
<li>Quality control plots of the count level data including PCA,
heatmaps, relative-log expression boxplots, density plots of count
distributions, and parallel coordinate plots of libraries. Additionally,
check the assumptions of global scaling normalization.</li>
<li>Differential expression testing on the raw counts using <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html" class="external-link">edgeR</a>,
<a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html" class="external-link">DESeq2</a>,
<a href="https://bioconductor.org/packages/release/bioc/html/baySeq.html" class="external-link">baySeq</a>,
or <a href="https://bioconductor.org/packages/release/bioc/html/limma.html" class="external-link">limma::voom</a>
</li>
<li>Creation of results plots such as volcano or MA plots.</li>
<li>Gene ontology analysis of interesting genes.</li>
<li>Gene set enrichment analysis.</li>
</ol>
</div>
<div class="section level2">
<h2 id="quality-control">Quality Control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<p><a href="https://github.com/OpenGene/fastp" class="external-link">fastp</a> has quickly
become my favorite tool for QC’ing fastq files primarily because it is
fast and produces nice looking output files that are also amenable to
summarization with <a href="https://multiqc.info/" class="external-link">MultiQC</a>. fastp
will also perform adapter trimming by default on fastq files. I tend to
fall in the camp that believes read quality trimming is <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7671312/" class="external-link">not
necessary</a> for RNA-seq alignment. However, I have never experienced
<em>worse</em> results after using the defaults with fastp so I leave it
be and just inspect the output carefully.</p>
<p>A simple bash script for running fastp over a set of fastq files
might look something like this:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co"># Run fastp on the raw fastq files</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co"># ----------------------------------------------------------------------------</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="bu">set</span> <span class="at">-Eeou</span> pipefail</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="va">FQ</span><span class="op">=</span>/path/to/00_fastq       <span class="co"># Directory containing raw fastq files</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="va">SAMPLES</span><span class="op">=</span>sample-names.txt   <span class="co"># A text file listing basenames of fastq files</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="va">OUT</span><span class="op">=</span>/path/to/put/01_fastp  <span class="co"># Where to save the fastp results</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="va">THREADS</span><span class="op">=</span>8</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$OUT</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="cf">for</span> SAMPLE <span class="kw">in</span> <span class="va">$(</span><span class="fu">cat</span> <span class="va">$SAMPLES)</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a>    <span class="ex">fastp</span> <span class="at">-i</span> <span class="va">$FQ</span>/<span class="va">${SAMPLE}</span>_R1.fq.gz <span class="dt">\</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a>          <span class="at">-I</span> <span class="va">$FQ</span>/<span class="va">${SAMPLE}</span>_R2.fq.gz <span class="dt">\</span></span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a>          <span class="at">-o</span> <span class="va">$OUT</span>/<span class="va">${SAMPLE}</span>.trimmed.1.fq.gz <span class="dt">\</span></span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a>          <span class="at">-O</span> <span class="va">$OUT</span>/<span class="va">${SAMPLE}</span>.trimmed.2.fq.gz <span class="dt">\</span></span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a>          <span class="at">-h</span> <span class="va">$OUT</span>/<span class="va">${SAMPLE}</span>.fastp.html <span class="dt">\</span></span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a>          <span class="at">-j</span> <span class="va">$OUT</span>/<span class="va">${SAMPLE}</span>.fastp.json <span class="dt">\</span></span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a>          <span class="at">-w</span> <span class="va">$THREADS</span></span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a><span class="cf">done</span></span></code></pre></div>
<p>Where <code>sample-names.txt</code> is a simple text file with each
basename like so:</p>
<pre><code><span><span class="va">Control1</span></span>
<span><span class="va">Control2</span></span>
<span><span class="va">Control3</span></span>
<span><span class="va">Treatment1</span></span>
<span><span class="va">Treatment2</span></span>
<span><span class="va">Treatment3</span></span></code></pre>
<p>It is important to name the results files with
<code>*.fastp.{json|html}</code> so that <code>multiqc</code> can
recognize the extensions and combine the results automatically.</p>
</div>
<div class="section level2">
<h2 id="alignment-and-quantification">Alignment and Quantification<a class="anchor" aria-label="anchor" href="#alignment-and-quantification"></a>
</h2>
<div class="section level3">
<h3 id="salmon">Salmon<a class="anchor" aria-label="anchor" href="#salmon"></a>
</h3>
<p>I tend to perform quantification with Salmon in order to obtain
transcript-level counts for each sample. A simple bash script for
performing quantification with Salmon looks like:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="co"># Perform transcript quantification with Salmon</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co"># ----------------------------------------------------------------------------</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="bu">set</span> <span class="at">-Eeou</span> pipefail</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="va">SAMPLES</span><span class="op">=</span>sample-names.txt  <span class="co"># Same sample-names.txt file as above  </span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="va">IDX</span><span class="op">=</span>/path/to/salmon-idx   <span class="co"># Index used by Salmon</span></span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="va">FQ</span><span class="op">=</span>/path/to/01_fastp      <span class="co"># Directory containing the fastp output</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a><span class="va">OUT</span><span class="op">=</span>/path/to/02_quants    <span class="co"># Where to save the Salmon results</span></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a><span class="va">THREADS</span><span class="op">=</span>12</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$OUT</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a><span class="cf">for</span> SAMPLE <span class="kw">in</span> <span class="va">$(</span><span class="fu">cat</span> <span class="va">$SAMPLES)</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a>    <span class="ex">salmon</span> quant <span class="dt">\</span></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a>           <span class="at">-i</span> <span class="va">$IDX</span> <span class="dt">\</span></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a>           <span class="at">-l</span> A <span class="dt">\</span></span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a>           <span class="at">-1</span> <span class="va">$FQ</span>/<span class="va">${SAMPLE}</span>.trimmed.1.fq.gz <span class="dt">\</span></span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a>           <span class="at">-2</span> <span class="va">$FQ</span>/<span class="va">${SAMPLE}</span>.trimmed.2.fq.gz <span class="dt">\</span></span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a>           <span class="at">--validateMappings</span> <span class="dt">\</span></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a>           <span class="at">--gcBias</span> <span class="dt">\</span></span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a>           <span class="at">--seqBias</span> <span class="dt">\</span></span>
<span id="cb3-26"><a href="#cb3-26" tabindex="-1"></a>           <span class="at">--threads</span> <span class="va">$THREADS</span> <span class="dt">\</span></span>
<span id="cb3-27"><a href="#cb3-27" tabindex="-1"></a>           <span class="at">-o</span> <span class="va">$OUT</span>/<span class="va">${SAMPLE}</span>_quants</span>
<span id="cb3-28"><a href="#cb3-28" tabindex="-1"></a><span class="cf">done</span></span></code></pre></div>
<p>I tend to always use the <code>--gcBias</code> and
<code>--seqBias</code> flags as they don’t impair accuracy in the
absence of biases (quantification just takes a little longer).</p>
</div>
<div class="section level3">
<h3 id="star">STAR<a class="anchor" aria-label="anchor" href="#star"></a>
</h3>
<p>Sometimes I also need to produce genomic coordinates for alignments.
For this purpose I tend to use STAR to generate BAM files as well as
produce gene-level counts with it’s inbuilt HTSeq-count functionality. A
simple bash script for running STAR might look like:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co">#!/usr/bin/env bash</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co"># Align reads with STAR</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co"># ----------------------------------------------------------------------------</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="bu">set</span> <span class="at">-Eeou</span> pipefail</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="va">SAMPLES</span><span class="op">=</span>sample-names.txt   <span class="co"># Same sample-names.txt file as above</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="va">FQ</span><span class="op">=</span>/path/to/01_fastp       <span class="co"># Directory containing the fastp output</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="va">OUT</span><span class="op">=</span>/path/to/03_STAR_outs  <span class="co"># Where to save the STAR results</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="va">IDX</span><span class="op">=</span>/path/to/STAR-idx      <span class="co"># Index used by STAR for alignment</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="va">THREADS</span><span class="op">=</span>24</span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$OUT</span></span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a></span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a><span class="cf">for</span> SAMPLE <span class="kw">in</span> <span class="va">$(</span><span class="fu">cat</span> <span class="va">$SAMPLES)</span></span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a><span class="cf">do</span></span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a>  <span class="ex">STAR</span> <span class="at">--runThreadN</span> <span class="va">$THREADS</span> <span class="dt">\</span></span>
<span id="cb4-19"><a href="#cb4-19" tabindex="-1"></a>       <span class="at">--genomeDir</span> <span class="va">$IDX</span> <span class="dt">\</span></span>
<span id="cb4-20"><a href="#cb4-20" tabindex="-1"></a>       <span class="at">--readFilesIn</span> <span class="va">${FQ}</span>/<span class="va">${SAMPLE}</span>.trimmed.1.fq.gz <span class="va">${FQ}</span>/<span class="va">${SAMPLE}</span>.trimmed.2.fq.gz <span class="dt">\</span></span>
<span id="cb4-21"><a href="#cb4-21" tabindex="-1"></a>       <span class="at">--readFilesCommand</span> zcat <span class="dt">\</span></span>
<span id="cb4-22"><a href="#cb4-22" tabindex="-1"></a>       <span class="at">--outFilterType</span> BySJout <span class="dt">\</span></span>
<span id="cb4-23"><a href="#cb4-23" tabindex="-1"></a>       <span class="at">--outFileNamePrefix</span> <span class="va">${OUT}</span>/<span class="va">${SAMPLE}</span>_ <span class="dt">\</span></span>
<span id="cb4-24"><a href="#cb4-24" tabindex="-1"></a>       <span class="at">--alignSJoverhangMin</span> 8 <span class="dt">\</span></span>
<span id="cb4-25"><a href="#cb4-25" tabindex="-1"></a>       <span class="at">--alignSJDBoverhangMin</span> 1 <span class="dt">\</span></span>
<span id="cb4-26"><a href="#cb4-26" tabindex="-1"></a>       <span class="at">--outFilterMismatchNmax</span> 999 <span class="dt">\</span></span>
<span id="cb4-27"><a href="#cb4-27" tabindex="-1"></a>       <span class="at">--outFilterMismatchNoverReadLmax</span> 0.04 <span class="dt">\</span></span>
<span id="cb4-28"><a href="#cb4-28" tabindex="-1"></a>       <span class="at">--alignIntronMin</span> 20 <span class="dt">\</span></span>
<span id="cb4-29"><a href="#cb4-29" tabindex="-1"></a>       <span class="at">--alignIntronMax</span> 1000000 <span class="dt">\</span></span>
<span id="cb4-30"><a href="#cb4-30" tabindex="-1"></a>       <span class="at">--alignMatesGapMax</span> 1000000 <span class="dt">\</span></span>
<span id="cb4-31"><a href="#cb4-31" tabindex="-1"></a>       <span class="at">--outMultimapperOrder</span> Random <span class="dt">\</span></span>
<span id="cb4-32"><a href="#cb4-32" tabindex="-1"></a>       <span class="at">--outSAMtype</span> BAM SortedByCoordinate <span class="dt">\</span></span>
<span id="cb4-33"><a href="#cb4-33" tabindex="-1"></a>       <span class="at">--quantMode</span> GeneCounts<span class="kw">;</span></span>
<span id="cb4-34"><a href="#cb4-34" tabindex="-1"></a><span class="cf">done</span></span></code></pre></div>
<p>For STAR I tend to use the ENCODE default parameters above for human
samples and also output gene level counts using the
<code>--quantMode GeneCounts</code> flag.</p>
</div>
</div>
<div class="section level2">
<h2 id="generating-a-matrix-of-gene-counts">Generating a matrix of gene counts<a class="anchor" aria-label="anchor" href="#generating-a-matrix-of-gene-counts"></a>
</h2>
<p>The recommended methods for performing differential expression
analysis implemented in <a href="https://bioconductor.org/packages/release/bioc/html/edgeR.html" class="external-link">edgeR</a>,
<a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html" class="external-link">DESeq2</a>,
<a href="https://bioconductor.org/packages/release/bioc/html/baySeq.html" class="external-link">baySeq</a>,
and <a href="https://bioconductor.org/packages/release/bioc/html/limma.html" class="external-link">limma::voom</a>
all require raw count matrices as input data.</p>
<div class="section level3">
<h3 id="importing-transcript-level-counts-from-salmon">Importing transcript level counts from <code>Salmon</code><a class="anchor" aria-label="anchor" href="#importing-transcript-level-counts-from-salmon"></a>
</h3>
<p>We use <code>R</code> to import the quant files into the active
session. <code>tximeta</code> will download the appropriate metadata for
the reference genome used and import the results as a
<code>SummarizedExperiment</code> object. Check out the <a href="https://coriell-research.github.io/coriell/articles/Practical-analysis-with-SummarizedExperiments.html" class="external-link">tutorial</a>
for working with <code>SummarizedExperiment</code> objects if you are
unfamiliar with their structure.</p>
<p>The code below will create a data.frame mapping sample names to file
paths containing quantification results. This data.frame is then used by
<code>tximeta</code> to import <code>Salmon</code> quantification
results at the transcript level (along with transcript annotations).
Then, we use <code>summarizeToGene()</code> to summarize the tx counts
to the gene level. Finally, we transform the
<code>SummarizedExperiment</code> object to a <code>DGEList</code> for
use in downstream analysis with <code>edgeR</code></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tximeta</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">quant_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span></span>
<span>  path <span class="op">=</span> <span class="st">"02_quants"</span>,</span>
<span>  pattern <span class="op">=</span> <span class="st">"quant.sf"</span>,</span>
<span>  full.names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  recursive <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract samples names from filepaths</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">quant_files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"02_quants"</span>, <span class="st">""</span>, <span class="va">quant_files</span>, fixed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">quant_files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_quants/quant.sf"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">quant_files</span><span class="op">)</span>, fixed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create metadata for import</span></span>
<span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">quant_files</span><span class="op">)</span>, </span>
<span>  files <span class="op">=</span> <span class="va">quant_files</span>,</span>
<span>  group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Treatment"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Import transcript counts with tximeta</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">tximeta</span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Summarize tx counts to the gene-level</span></span>
<span><span class="va">gse</span> <span class="op">&lt;-</span> <span class="fu">summarizeToGene</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Import into edgeR for downstream analysis</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html" class="external-link">SE2DGEList</a></span><span class="op">(</span><span class="va">gse</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="importing-gene-counts-from-star">Importing gene counts from STAR<a class="anchor" aria-label="anchor" href="#importing-gene-counts-from-star"></a>
</h3>
<p>If you used STAR to generate counts with HTSeq-count then
<code>edgeR</code> can directly import the results for downstream
analysis like so:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Specify the filepaths to gene counts from STAR</span></span>
<span><span class="va">count_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span></span>
<span>  path <span class="op">=</span> <span class="st">"03_STAR_outs"</span>, </span>
<span>  pattern <span class="op">=</span> <span class="st">"*.ReadsPerGene.out.tab"</span>, </span>
<span>  full.names <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Name the file with their sample names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">count_files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".ReadsPerGene.out.tab"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">count_files</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Import HTSeq counts into a DGEList </span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/readDGE.html" class="external-link">readDGE</a></span><span class="op">(</span></span>
<span>  files <span class="op">=</span> <span class="va">count_files</span>, </span>
<span>  columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,  <span class="co"># Gene name and 'unstranded' count columns</span></span>
<span>  group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Treatment"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">count_files</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="test-data">Test data<a class="anchor" aria-label="anchor" href="#test-data"></a>
</h2>
<p>We will use data from the <a href="https://bioconductor.org/packages/3.19/data/experiment/html/airway.html" class="external-link">airway</a>
package to illustrate differential expression analysis steps. Please see
<a href="https://bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html#preparing-quantification-input-to-deseq2" class="external-link">Section
2</a> of the <em>rnaseqGene</em> workflow for more information.</p>
<p>Below, we load the data from the airway package and use
<code>SE2DGEList</code> to convert the object to a <code>DGElist</code>
for use with <code>edgeR</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Load the SummarizedExperiment object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set the group levels</span></span>
<span><span class="va">airway</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">airway</span><span class="op">$</span><span class="va">dex</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"untrt"</span>, <span class="st">"trt"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert to a DGEList to be consistent with above steps</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/SE2DGEList.html" class="external-link">SE2DGEList</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="library-qc">Library QC<a class="anchor" aria-label="anchor" href="#library-qc"></a>
</h2>
<p>Before we perform differential expression analysis it is important to
explore the samples’ library distributions in order to ensure good
quality before downstream analysis. There are several diagnostic plots
we can use for this purpose implemented in the <code>coriell</code>
package. However, first we must remove any features that have too low of
counts for meaningful differential expression analysis. This can be
achieved using <code><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">edgeR::filterByExpr()</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Determine which genes have enough counts to keep around</span></span>
<span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Remove the unexpressed genes</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="va">keep</span>,,keep.lib.sizes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span></code></pre></div>
<p>At this stage it is often wise to perform library QC on the library
size normalized counts. This will give us an idea about potential global
expression differences and potential outliers <em>before</em>
introducing normalization factors. We can use <code>edgeR</code> to
generate log2 counts-per-million values for the retained genes.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html" class="external-link">cpm</a></span><span class="op">(</span><span class="va">y</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="relative-log-expression-boxplots">Relative log expression boxplots<a class="anchor" aria-label="anchor" href="#relative-log-expression-boxplots"></a>
</h3>
<p>The first diagnostic plot we can look at is a plot of the relative
log expression values. <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0191629" class="external-link">RLE
plots</a> are good diagnostic tools for evaluating unwanted variation in
libraries.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/coriell-research/coriell" class="external-link">coriell</a></span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_boxplot.html">plot_boxplot</a></span><span class="op">(</span><span class="va">logcounts</span>, metadata <span class="op">=</span> <span class="va">y</span><span class="op">$</span><span class="va">samples</span>, fillBy <span class="op">=</span> <span class="st">"group"</span>, rle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Relative Log Expression"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"Sample"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"RLE"</span>,</span>
<span>       color <span class="op">=</span> <span class="st">"Treatment Group"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="rnaseq-workflow_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>We can see from the above RLE plot that the samples are centered
around zero and have mostly similar distributions. It is also clear that
two of the samples, “SRR1039520” and “SRR1039521”, have slightly
different distributions than the others.</p>
</div>
<div class="section level3">
<h3 id="library-density-plots">Library density plots<a class="anchor" aria-label="anchor" href="#library-density-plots"></a>
</h3>
<p>Library density plots show the density of reads corresponding to a
particular magnitude of counts. Shifts of these curves should align with
group differences and generally samples from the same group should have
overlapping density curves</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_density.html">plot_density</a></span><span class="op">(</span><span class="va">logcounts</span>, metadata <span class="op">=</span> <span class="va">y</span><span class="op">$</span><span class="va">samples</span>, colBy <span class="op">=</span> <span class="st">"group"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Library Densities"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"logCPM"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"Density"</span>,</span>
<span>       color <span class="op">=</span> <span class="st">"Treatment Group"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="rnaseq-workflow_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="sample-vs-sample-distances">Sample vs Sample Distances<a class="anchor" aria-label="anchor" href="#sample-vs-sample-distances"></a>
</h3>
<p>We can also calculate the euclidean distance between all pairs of
samples and display this on a heatmap. Again, samples from the same
group should show smaller distances than sample pairs from differing
groups.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_dist.html">plot_dist</a></span><span class="op">(</span><span class="va">logcounts</span>, metadata <span class="op">=</span> <span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">[</span>, <span class="st">"group"</span>, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="rnaseq-workflow_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="parallel-coordinates-plot">Parallel coordinates plot<a class="anchor" aria-label="anchor" href="#parallel-coordinates-plot"></a>
</h3>
<p>Parallel coordinates plots are useful for giving you an idea of how
the most variable genes change between treatment groups. These plots
show the expression of each gene as a line on the y-axis traced between
samples on the x-axis.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_parallel.html">plot_parallel</a></span><span class="op">(</span><span class="va">logcounts</span>, <span class="va">y</span><span class="op">$</span><span class="va">samples</span>, colBy <span class="op">=</span> <span class="st">"group"</span>, </span>
<span>              removeVar <span class="op">=</span> <span class="fl">0.9</span>, alpha <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"10% Most Variable Genes"</span>,</span>
<span>       x <span class="op">=</span> <span class="st">"Sample"</span>,</span>
<span>       y <span class="op">=</span> <span class="st">"logCPM"</span>,</span>
<span>       color <span class="op">=</span> <span class="st">"Treatment Group"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Removing 90% lowest variance features...</span></span></code></pre></div>
<p><img src="rnaseq-workflow_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h3>
<p>Principal components analysis is an unsupervised method for reducing
the dimensionality of a dataset while maintaining its fundamental
structure. PCA biplots can be used to examine sample groupings following
PCA. These biplots can reveal overall patterns of expression as well as
potential problematic samples prior to downstream analysis. For simple
analyses we expect to see the ‘main’ effect primarily along the first
component.</p>
<p>I like to use the <a href="https://bioconductor.org/packages/release/bioc/vignettes/PCAtools/inst/doc/PCAtools.html" class="external-link">PCAtools</a>
package for quickly computing and plotting principal components. For
more complicated experiments I have also found UMAP (see
<code><a href="../reference/UMAP.html">coriell::UMAP()</a></code>) to be useful for dimensionality
reduction.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kevinblighe/PCAtools" class="external-link">PCAtools</a></span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Perform PCA on the 20% most variable genes</span></span>
<span><span class="co"># Center and scale the variable after selecting most variable</span></span>
<span><span class="va">pca_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/PCAtools/man/pca.html" class="external-link">pca</a></span><span class="op">(</span></span>
<span>  <span class="va">logcounts</span>, </span>
<span>  metadata <span class="op">=</span> <span class="va">y</span><span class="op">$</span><span class="va">samples</span>, </span>
<span>  center <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  removeVar <span class="op">=</span> <span class="fl">0.8</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Show the PCA biplot</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/PCAtools/man/biplot.html" class="external-link">biplot</a></span><span class="op">(</span></span>
<span>  <span class="va">pca_result</span>, </span>
<span>  colby <span class="op">=</span> <span class="st">"group"</span>, </span>
<span>  hline <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  vline <span class="op">=</span> <span class="fl">0</span>, </span>
<span>  hlineType <span class="op">=</span> <span class="fl">2</span>, </span>
<span>  vlineType <span class="op">=</span> <span class="fl">2</span>, </span>
<span>  legendPosition <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>  caption <span class="op">=</span> <span class="st">"20% Most Variable Features"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="rnaseq-workflow_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="assessing-global-scaling-normalization-assumptions">Assessing global scaling normalization assumptions<a class="anchor" aria-label="anchor" href="#assessing-global-scaling-normalization-assumptions"></a>
</h3>
<p>Most downstream differential expression testing methods apply a
global scaling normalization factor to each library prior to DE testing.
Applying these normalization factors when there are global expression
differences can lead to spurious results. In typical experiments this is
usually not a problem but when dealing with cancer or epigenetic drug
treatment this can actually lead to many problems if not identified.</p>
<p>In order to identify potential violations of global scaling
normalization I use the <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/quantro/inst/doc/quantro.html" class="external-link">quantro</a>
R package. <code>quantro</code> uses two data driven approaches to
assess the appropriateness of global scaling normalization. The first
involves testing if the medians of the distributions differ between
groups. These differences could indicate technical or real biological
variation. The second test assesses the ratio of between group
variability to within group variability using a permutation test similar
to an ANOVA. If this value is large, it suggests global adjustment
methods might not be appropriate.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">quantro</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Initialize multiple (8) cores for permutation testing</span></span>
<span><span class="fu">doParallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/doParallel/man/registerDoParallel.html" class="external-link">registerDoParallel</a></span><span class="op">(</span>cores <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compute the qstat on the filtered libraries</span></span>
<span><span class="va">qtest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/quantro/man/quantro.html" class="external-link">quantro</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">counts</span>, groupFactor <span class="op">=</span> <span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">group</span>, B <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></code></pre></div>
<p>Now we can assess the results. We can use <code><a href="https://rdrr.io/pkg/quantro/man/anova.html" class="external-link">anova()</a></code> to
test for differences in medians across groups. Here, they do not
significantly differ.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/quantro/man/anova.html" class="external-link">anova</a></span><span class="op">(</span><span class="va">qtest</span><span class="op">)</span></span>
<span><span class="co">#&gt; Analysis of Variance Table</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Response: objectMedians</span></span>
<span><span class="co">#&gt;             Df  Sum Sq Mean Sq F value Pr(&gt;F)</span></span>
<span><span class="co">#&gt; groupFactor  1  1984.5  1984.5  0.3813 0.5596</span></span>
<span><span class="co">#&gt; Residuals    6 31225.5  5204.2</span></span></code></pre></div>
<p>We can also plot the results of the permutation test to see the
between:within group ratios. Again, there are no large differences in
this dataset suggesting that global scaling normalization such as TMM is
appropriate.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/quantro/man/quantroPlot.html" class="external-link">quantroPlot</a></span><span class="op">(</span><span class="va">qtest</span><span class="op">)</span></span></code></pre></div>
<p><img src="rnaseq-workflow_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="differential-expression-testing-with-edger">Differential expression testing with <code>edgeR</code><a class="anchor" aria-label="anchor" href="#differential-expression-testing-with-edger"></a>
</h2>
<p>After removing lowly expressed features and checking the assumptions
of normalization we can perform downstream differential expression
testing with <code>edgeR</code>. The <a href="https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf" class="external-link">edgeR
manual</a> contains a detailed explanation of all steps involved in
differential expression testing.</p>
<p>In short, we need to specify the experimental design, estimate
normalization factors, fit the models, and perform DE testing.</p>
<div class="section level3">
<h3 id="creating-the-experimental-design">Creating the experimental design<a class="anchor" aria-label="anchor" href="#creating-the-experimental-design"></a>
</h3>
<p>Maybe the most important step in DE analysis is properly constructing
a design matrix. The details of design matrices are outside of the scope
of this tutorial but a good overview can be found <a href="https://bioconductor.org/packages/release/workflows/vignettes/RNAseq123/inst/doc/designmatrices.html" class="external-link">here</a>.
Generally, your samples will fall nicely into several well defined
groups, facilitating the use of a design matrix without an intercept
e.g. <code>design ~ model.matrix(~0 + group, ...)</code>. This kind of
design matrix makes it relatively simple to construct contrasts that
describe exactly what pairs of groups you want to compare.</p>
<p>Since this example experiment is simply comparing treatments to
control samples we can model the differences in means by using a model
<em>with</em> an intercept where the intercept is the mean of the
control samples and the 2nd coefficient represents the differences in
the treatment group.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Model with intercept</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">group</span>, data <span class="op">=</span> <span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span></code></pre></div>
<p>We can make an equivalent model and test <em>without</em> an
intercept like so:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># A means model</span></span>
<span><span class="va">design_no_intercept</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fl">0</span> <span class="op">+</span> <span class="va">group</span>, data <span class="op">=</span> <span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Construct contrasts to test the difference in means between the groups</span></span>
<span><span class="va">cm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html" class="external-link">makeContrasts</a></span><span class="op">(</span></span>
<span>  Treatment_vs_Control <span class="op">=</span> <span class="va">grouptrt</span> <span class="op">-</span> <span class="va">groupuntrt</span>,</span>
<span>  levels <span class="op">=</span> <span class="va">design_no_intercept</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The choice of which design is up to you. I typically use whatever is
clearer for the experiment at hand. In this case that is the model with
an intercept.</p>
</div>
<div class="section level3">
<h3 id="estimating-normalization-factors">Estimating normalization factors<a class="anchor" aria-label="anchor" href="#estimating-normalization-factors"></a>
</h3>
<p>We use <code>edgeR</code> to calculate trimmed mean of the M-value
(TMM) normalization factors for each library.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Estimate TMM normalization factors</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">normLibSizes</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre></div>
<p>We can check the normalization by creating MA plots for each
library</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMD.html" class="external-link">plotMD</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html" class="external-link">cpm</a></span><span class="op">(</span><span class="va">y</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, column <span class="op">=</span> <span class="va">i</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fl">0</span>, lty <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"red2"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="rnaseq-workflow_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<div class="section level4">
<h4 id="what-to-do-if-global-scaling-normalization-is-violated">What to do if global scaling normalization is violated?<a class="anchor" aria-label="anchor" href="#what-to-do-if-global-scaling-normalization-is-violated"></a>
</h4>
<p>Above I described testing for violations of global scaling
normalization. So what should we do if these assumptions are violated
and we don’t have a good set of control genes or spike-ins etc.?</p>
<p>If we believe that the differences we are observing are due to true
biological phenomena (<strong>this is a big assumption</strong>) then we
can try to apply a method such as <a href="https://academic.oup.com/biostatistics/article/19/2/185/3949169" class="external-link">smooth
quantile normalization</a> to the data using the <a href="https://www.bioconductor.org/packages/release/bioc/html/qsmooth.html" class="external-link">qsmooth</a>
package.</p>
<p>Below I will show how to apply <code>qsmooth</code> to our filtered
counts and then calculate offsets to be used in downstream DE analysis
with <code>edgeR</code>. Please note <strong>this is not a benchmarked
or ‘official’ workflow</strong> just a method that I have implemented
based on reading forums and github issues.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">qsmooth</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Compute the smooth quantile factors </span></span>
<span><span class="va">qs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/qsmooth/man/qsmooth.html" class="external-link">qsmooth</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">counts</span>, group_factor <span class="op">=</span> <span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">group</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the qsmooth transformed data</span></span>
<span><span class="va">qsd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/qsmooth/man/qsmoothData.html" class="external-link">qsmoothData</a></span><span class="op">(</span><span class="va">qs</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate offsets to be used by edgeR in place of norm.factors</span></span>
<span><span class="co"># Offsets are on the natural log scale. Add a small offset to avoid</span></span>
<span><span class="co"># taking logs of zero </span></span>
<span><span class="va">offset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="va">qsd</span> <span class="op">+</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Scale the offsets for internal usage by the DGEList object</span></span>
<span><span class="co"># Now the object is ready for downstream analysis</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/scaleOffset.html" class="external-link">scaleOffset</a></span><span class="op">(</span><span class="va">y</span>, offset <span class="op">=</span> <span class="va">offset</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># To create logCPM values with the new norm factors use</span></span>
<span><span class="va">lcpm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html" class="external-link">cpm</a></span><span class="op">(</span><span class="va">y</span>, offset <span class="op">=</span> <span class="va">y</span><span class="op">$</span><span class="va">offset</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="fit-the-model">Fit the model<a class="anchor" aria-label="anchor" href="#fit-the-model"></a>
</h3>
<p>New in edgeR 4.0 is the ability to estimate dispersions while
performing the model fitting step. I typically tend to ‘robustify’ the
fit to outliers. Below I will perform dispersion estimation in legacy
mode so that we can use competitive gene set testing later. If we want
to use the new workflow we can use the following:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># edgeR 4.0 workflow</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFit.html" class="external-link">glmQLFit</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span>, legacy <span class="op">=</span> <span class="cn">FALSE</span>, robust <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>We will continue with the legacy workflow.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span>, robust <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFit.html" class="external-link">glmQLFit</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span>, robust <span class="op">=</span> <span class="cn">TRUE</span>, legacy <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>It’s always a good idea at this step to check some of the diagnostic
plots from edgeR</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Show the biological coefficient of variation</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/plotBCV.html" class="external-link">plotBCV</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre></div>
<p><img src="rnaseq-workflow_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Show the dispersion estimates</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/plotQLDisp.html" class="external-link">plotQLDisp</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<p><img src="rnaseq-workflow_files/figure-html/unnamed-chunk-25-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="test-for-differential-expression">Test for differential expression<a class="anchor" aria-label="anchor" href="#test-for-differential-expression"></a>
</h3>
<p>Now that the models have been fit we can test for differential
expression.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Test the treatment fs control condition</span></span>
<span><span class="va">qlf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html" class="external-link">glmQLFTest</a></span><span class="op">(</span><span class="va">fit</span>, coef <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p>Often it is more biologically relevant to give more weight to higher
fold changes. This can be acheived using <code><a href="https://rdrr.io/pkg/edgeR/man/glmTreat.html" class="external-link">glmTreat()</a></code>.
<strong>NOTE</strong> do not use <code><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html" class="external-link">glmQLFTest()</a></code> and then
filter by fold-change - you destroy the FDR correction!</p>
<p>When testing against a fold change we can use relatively modest
values since the fold change must exceed this threshold before being
considered for significance. Values such as <code>log2(1.2)</code> or
<code>log2(1.5)</code> work well in practice.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">trt_vs_control_fc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmTreat.html" class="external-link">glmTreat</a></span><span class="op">(</span><span class="va">fit</span>, coef <span class="op">=</span> <span class="fl">2</span>, lfc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fl">1.2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>In any case, the results of the differential expression test can be
extracted to a data.frame for downstream plotting with
<code><a href="../reference/edger_to_df.html">coriell::edger_to_df()</a></code></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/edger_to_df.html">edger_to_df</a></span><span class="op">(</span><span class="va">qlf</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="plotting-de-results">Plotting DE results<a class="anchor" aria-label="anchor" href="#plotting-de-results"></a>
</h2>
<p>The two most common plots for differential expression analysis
results are the volcano plot and the MA plot. Volcano plots display the
negative log10 of the significance value on the y-axis vs the log2
fold-change on the x-axis. MA plots show the average expression of the
gene on the x-axis vs the log2 fold-change of the gene on the y-axis.
The <code>coriell</code> package includes functions for producing
both.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Create a volcano plot of the results</span></span>
<span><span class="va">v</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">de_result</span>, fdr <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create and MA plot of the results</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_md.html">plot_md</a></span><span class="op">(</span><span class="va">de_result</span>, fdr <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Patch both plots together</span></span>
<span><span class="op">(</span><span class="va">v</span> <span class="op">|</span> <span class="va">m</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Treatment vs. Control"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="../reference/theme_coriell.html">theme_coriell</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="rnaseq-workflow_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="competitive-gene-set-testing-with-camera">Competitive gene set testing with <code>camera()</code><a class="anchor" aria-label="anchor" href="#competitive-gene-set-testing-with-camera"></a>
</h2>
<p>I’ve recently become aware of some of the <a href="https://support.bioconductor.org/p/9158100/#9158105" class="external-link">problems</a>
with gene set enrichment analysis using the <code>fgsea</code> package.
Following Gordon Smyth’s advice, I have switched all of my pipelines to
using competitive gene set testing (when appropriate) in
<code>limma</code> to avoid problems with correlated genes.</p>
<p>Below we use the <code>msigdb</code> R package to retrieve HALLMARK
gene sets and then use <code><a href="https://rdrr.io/pkg/limma/man/camera.html" class="external-link">limma::camera()</a></code> for gene set
testing.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://davislaboratory.github.io/msigdb" class="external-link">msigdb</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GSEABase</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Get the gene set data</span></span>
<span><span class="va">msigdb_hs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/msigdb/man/getMsigdb.html" class="external-link">getMsigdb</a></span><span class="op">(</span>org <span class="op">=</span> <span class="st">"hs"</span>, id <span class="op">=</span> <span class="st">"SYM"</span>, version <span class="op">=</span> <span class="st">"7.5"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset for only the HALLMARK sets</span></span>
<span><span class="va">hallmark</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/msigdb/man/subsetCollection.html" class="external-link">subsetCollection</a></span><span class="op">(</span><span class="va">msigdb_hs</span>, <span class="st">"h"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the gene symbols for each set as a list</span></span>
<span><span class="va">msigdb_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GSEABase/man/GeneSet-class.html" class="external-link">geneIds</a></span><span class="op">(</span><span class="va">hallmark</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Convert the gene sets into lists of indeces for edgeR</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ids2indices.html" class="external-link">ids2indices</a></span><span class="op">(</span>gene.sets <span class="op">=</span> <span class="va">msigdb_ids</span>, identifiers <span class="op">=</span> <span class="va">y</span><span class="op">$</span><span class="va">genes</span><span class="op">$</span><span class="va">gene_name</span><span class="op">)</span></span></code></pre></div>
<p>Perform gene set testing. Note here we can use
<code><a href="https://rdrr.io/pkg/limma/man/camera.html" class="external-link">limma::camera()</a></code> <code><a href="https://rdrr.io/pkg/limma/man/roast.html" class="external-link">limma::mroast()</a></code>, or
<code><a href="https://rdrr.io/pkg/limma/man/romer.html" class="external-link">limma::romer()</a></code> depending on the hypothesis being tested.
The above setup code provides valid input for all of the above
functions.</p>
<p>See this <a href="https://support.bioconductor.org/p/103504/#103505" class="external-link">comment</a>
from Aaron Lun describing the difference between <code>camera</code> and
<code>roast</code>. For GSEA like hypothesis we can use
<code><a href="https://rdrr.io/pkg/limma/man/romer.html" class="external-link">limma::romer()</a></code></p>
<blockquote>
<p><code><a href="https://rdrr.io/pkg/limma/man/roast.html" class="external-link">roast()</a></code> performs a self-contained gene set test, where
it looks for any DE within the set of genes. <code><a href="https://rdrr.io/pkg/limma/man/camera.html" class="external-link">camera()</a></code>
performs a competitive gene set test, where it compares the DE within
the gene set to the DE outside of the gene set.</p>
</blockquote>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use camera to perform competitive gene set testing</span></span>
<span><span class="va">camera_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/camera.html" class="external-link">camera</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">idx</span>, <span class="va">design</span>, contrast <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Use mroast for rotational gene set testing - bump up number of rotations</span></span>
<span><span class="va">mroast_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/roast.html" class="external-link">mroast</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">idx</span>, <span class="va">design</span>, contrast <span class="op">=</span> <span class="fl">2</span>, nrot <span class="op">=</span> <span class="fl">1e4</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Use romer for GSEA like hypothesis testing</span></span>
<span><span class="va">romer_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/romer.html" class="external-link">romer</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">idx</span>, <span class="va">design</span>, contrast <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p>We can also perform a pre-ranked version of the camera test using
<code><a href="https://rdrr.io/pkg/limma/man/camera.html" class="external-link">cameraPR()</a></code>. In order to use the pre-ranked version we need
to create a ranking statistic. The <a href="https://support.bioconductor.org/p/120780/#120785" class="external-link">suggestion</a>
from Gordon Smyth is to derive a z-statistic from the F-scores like
so:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">t_stat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">de_result</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">de_result</span><span class="op">$</span><span class="va">`F`</span><span class="op">)</span></span>
<span><span class="va">z</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/zscoreT.html" class="external-link">zscoreT</a></span><span class="op">(</span><span class="va">t_stat</span>, df <span class="op">=</span> <span class="va">qlf</span><span class="op">$</span><span class="va">df.total</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Name the stat vector with the gene names </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">z</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">de_result</span><span class="op">$</span><span class="va">gene_name</span></span>
<span></span>
<span><span class="co"># Use the z-scores as the ranking stat for cameraPR</span></span>
<span><span class="va">camera_pr_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/camera.html" class="external-link">cameraPR</a></span><span class="op">(</span><span class="va">z</span>, <span class="va">idx</span><span class="op">)</span></span></code></pre></div>
<p>Another useful plot to show following gene set testing is a
barcodeplot. We The barcodeplot displays the enrichment of a given
signature for a ranked list of genes. The
<code><a href="https://rdrr.io/pkg/limma/man/barcodeplot.html" class="external-link">limma::barcodeplot()</a></code> function allows us to easily create
these plots for any of the gene sets of interest using any ranking stat
of our choice.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Show barcodeplot using the z-scores</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/barcodeplot.html" class="external-link">barcodeplot</a></span><span class="op">(</span></span>
<span>  <span class="va">z</span>, </span>
<span>  index <span class="op">=</span> <span class="va">idx</span><span class="op">[[</span><span class="st">"HALLMARK_ANDROGEN_RESPONSE"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>  main <span class="op">=</span> <span class="st">"HALLMARK_ANDROGEN_RESPONSE"</span>,</span>
<span>  xlab <span class="op">=</span> <span class="st">"z-score"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="rnaseq-workflow_files/figure-html/unnamed-chunk-33-1.png" width="700"></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Or you can use the logFC</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/barcodeplot.html" class="external-link">barcodeplot</a></span><span class="op">(</span></span>
<span>  <span class="va">de_result</span><span class="op">$</span><span class="va">logFC</span>, </span>
<span>  index <span class="op">=</span> <span class="va">idx</span><span class="op">[[</span><span class="st">"HALLMARK_ANDROGEN_RESPONSE"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>  main <span class="op">=</span> <span class="st">"HALLMARK_ANDROGEN_RESPONSE"</span>,</span>
<span>  xlab <span class="op">=</span> <span class="st">"logFC"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="rnaseq-workflow_files/figure-html/unnamed-chunk-33-2.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="gene-ontology-go-over-representation-test">Gene ontology (GO) over representation test<a class="anchor" aria-label="anchor" href="#gene-ontology-go-over-representation-test"></a>
</h2>
<p>Over representation analysis can be performed with the
<code>clusterProfiler</code> package. Here, instead of using the entire
gene list as input we select separate sets of up and down-regulated
genes and test to see if these sets are enriched in our differentially
expressed gene list.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/" class="external-link">clusterProfiler</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Split the genes into up and down</span></span>
<span><span class="va">up_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span></span>
<span>  <span class="va">de_result</span>, </span>
<span>  <span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span>, </span>
<span>  <span class="st">"gene_name"</span>, </span>
<span>  drop <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">down_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span></span>
<span>  <span class="va">de_result</span>, </span>
<span>  <span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">logFC</span> <span class="op">&lt;</span> <span class="fl">0</span>, </span>
<span>  <span class="st">"gene_name"</span>, </span>
<span>  drop <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the list of all genes expressed in the experiment</span></span>
<span><span class="co"># to use as a background set</span></span>
<span><span class="va">universe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">genes</span><span class="op">$</span><span class="va">gene_name</span><span class="op">)</span></span></code></pre></div>
<p>Create results objects for each set of genes</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ego_up</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html" class="external-link">enrichGO</a></span><span class="op">(</span></span>
<span>  gene <span class="op">=</span> <span class="va">up_genes</span>,</span>
<span>  universe <span class="op">=</span> <span class="va">universe</span>,</span>
<span>  OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>  keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>  ont <span class="op">=</span> <span class="st">"ALL"</span>,</span>
<span>  pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  pvalueCutoff <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  qvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  readable <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="va">ego_down</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html" class="external-link">enrichGO</a></span><span class="op">(</span></span>
<span>  gene <span class="op">=</span> <span class="va">down_genes</span>,</span>
<span>  universe <span class="op">=</span> <span class="va">universe</span>,</span>
<span>  OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>  keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>  ont <span class="op">=</span> <span class="st">"ALL"</span>,</span>
<span>  pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  pvalueCutoff <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  qvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  readable <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>These results can be converted to data.frames and combined with:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ego_up_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">ego_up</span><span class="op">)</span></span>
<span><span class="va">ego_down_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">ego_down</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ego_df</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/rbindlist.html" class="external-link">rbindlist</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>up <span class="op">=</span> <span class="va">ego_up_df</span>, down <span class="op">=</span> <span class="va">ego_down_df</span><span class="op">)</span>, </span>
<span>  idcol <span class="op">=</span> <span class="st">"Direction"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>Or the results can be plotted as dotplots with:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html" class="external-link">dotplot</a></span><span class="op">(</span><span class="va">ego_up</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Up-regulated genes"</span><span class="op">)</span></span>
<span><span class="va">d2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html" class="external-link">dotplot</a></span><span class="op">(</span><span class="va">ego_down</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Down-regulated genes"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">d1</span> <span class="op">|</span> <span class="va">d2</span></span></code></pre></div>
<p><img src="rnaseq-workflow_files/figure-html/unnamed-chunk-37-1.png" width="768"></p>
<p>You can also create a nice enrichment map showing similarity between
the significant GO terms like so:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">em_up</span> <span class="op">&lt;-</span> <span class="fu">enrichplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/pairwise_termsim.html" class="external-link">pairwise_termsim</a></span><span class="op">(</span><span class="va">ego_up</span><span class="op">)</span></span>
<span><span class="va">em_down</span> <span class="op">&lt;-</span> <span class="fu">enrichplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/pairwise_termsim.html" class="external-link">pairwise_termsim</a></span><span class="op">(</span><span class="va">ego_down</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">enrichplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/emapplot.html" class="external-link">emapplot</a></span><span class="op">(</span><span class="va">em_up</span>, showCategory <span class="op">=</span> <span class="fl">10</span>, edge.params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>min <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Up-regulated genes"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">enrichplot</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/emapplot.html" class="external-link">emapplot</a></span><span class="op">(</span><span class="va">em_down</span>, showCategory <span class="op">=</span> <span class="fl">10</span>, edge.params <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>min <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Down-regulated genes"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">|</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="rnaseq-workflow_files/figure-html/unnamed-chunk-38-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin20</span></span>
<span><span class="co">#&gt; Running under: macOS Sonoma 14.5</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: America/New_York</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] org.Hs.eg.db_3.19.1         clusterProfiler_4.12.0     </span></span>
<span><span class="co">#&gt;  [3] GSEABase_1.66.0             graph_1.82.0               </span></span>
<span><span class="co">#&gt;  [5] annotate_1.82.0             XML_3.99-0.17              </span></span>
<span><span class="co">#&gt;  [7] AnnotationDbi_1.66.0        ExperimentHub_2.12.0       </span></span>
<span><span class="co">#&gt;  [9] AnnotationHub_3.12.0        BiocFileCache_2.12.0       </span></span>
<span><span class="co">#&gt; [11] dbplyr_2.5.0                msigdb_1.12.0              </span></span>
<span><span class="co">#&gt; [13] patchwork_1.2.0             quantro_1.38.0             </span></span>
<span><span class="co">#&gt; [15] PCAtools_2.16.0             ggrepel_0.9.5              </span></span>
<span><span class="co">#&gt; [17] coriell_0.15.0              ggplot2_3.5.1              </span></span>
<span><span class="co">#&gt; [19] edgeR_4.2.0                 limma_3.60.3               </span></span>
<span><span class="co">#&gt; [21] airway_1.24.0               SummarizedExperiment_1.34.0</span></span>
<span><span class="co">#&gt; [23] Biobase_2.64.0              GenomicRanges_1.56.1       </span></span>
<span><span class="co">#&gt; [25] GenomeInfoDb_1.40.1         IRanges_2.38.1             </span></span>
<span><span class="co">#&gt; [27] S4Vectors_0.42.1            BiocGenerics_0.50.0        </span></span>
<span><span class="co">#&gt; [29] MatrixGenerics_1.16.0       matrixStats_1.3.0          </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] fs_1.6.4                  bitops_1.0-7             </span></span>
<span><span class="co">#&gt;   [3] enrichplot_1.24.0         HDO.db_0.99.1            </span></span>
<span><span class="co">#&gt;   [5] httr_1.4.7                RColorBrewer_1.1-3       </span></span>
<span><span class="co">#&gt;   [7] doParallel_1.0.17         tools_4.4.1              </span></span>
<span><span class="co">#&gt;   [9] doRNG_1.8.6               utf8_1.2.4               </span></span>
<span><span class="co">#&gt;  [11] R6_2.5.1                  HDF5Array_1.32.0         </span></span>
<span><span class="co">#&gt;  [13] lazyeval_0.2.2            rhdf5filters_1.16.0      </span></span>
<span><span class="co">#&gt;  [15] withr_3.0.0               gridExtra_2.3            </span></span>
<span><span class="co">#&gt;  [17] base64_2.0.1              preprocessCore_1.66.0    </span></span>
<span><span class="co">#&gt;  [19] cli_3.6.3                 textshaping_0.4.0        </span></span>
<span><span class="co">#&gt;  [21] scatterpie_0.2.3          labeling_0.4.3           </span></span>
<span><span class="co">#&gt;  [23] sass_0.4.9                readr_2.1.5              </span></span>
<span><span class="co">#&gt;  [25] genefilter_1.86.0         askpass_1.2.0            </span></span>
<span><span class="co">#&gt;  [27] pkgdown_2.1.0             Rsamtools_2.20.0         </span></span>
<span><span class="co">#&gt;  [29] systemfonts_1.1.0         yulab.utils_0.1.4        </span></span>
<span><span class="co">#&gt;  [31] gson_0.1.0                siggenes_1.78.0          </span></span>
<span><span class="co">#&gt;  [33] illuminaio_0.46.0         DOSE_3.30.1              </span></span>
<span><span class="co">#&gt;  [35] scrime_1.3.5              rstudioapi_0.16.0        </span></span>
<span><span class="co">#&gt;  [37] RSQLite_2.3.7             gridGraphics_0.5-1       </span></span>
<span><span class="co">#&gt;  [39] generics_0.1.3            BiocIO_1.14.0            </span></span>
<span><span class="co">#&gt;  [41] dplyr_1.1.4               GO.db_3.19.1             </span></span>
<span><span class="co">#&gt;  [43] Matrix_1.7-0              fansi_1.0.6              </span></span>
<span><span class="co">#&gt;  [45] abind_1.4-5               lifecycle_1.0.4          </span></span>
<span><span class="co">#&gt;  [47] yaml_2.3.9                rhdf5_2.48.0             </span></span>
<span><span class="co">#&gt;  [49] qvalue_2.36.0             SparseArray_1.4.8        </span></span>
<span><span class="co">#&gt;  [51] grid_4.4.1                blob_1.2.4               </span></span>
<span><span class="co">#&gt;  [53] dqrng_0.4.1               crayon_1.5.3             </span></span>
<span><span class="co">#&gt;  [55] lattice_0.22-6            beachmat_2.20.0          </span></span>
<span><span class="co">#&gt;  [57] cowplot_1.1.3             GenomicFeatures_1.56.0   </span></span>
<span><span class="co">#&gt;  [59] KEGGREST_1.44.1           pillar_1.9.0             </span></span>
<span><span class="co">#&gt;  [61] knitr_1.48                beanplot_1.3.1           </span></span>
<span><span class="co">#&gt;  [63] fgsea_1.30.0              rjson_0.2.21             </span></span>
<span><span class="co">#&gt;  [65] codetools_0.2-20          fastmatch_1.1-4          </span></span>
<span><span class="co">#&gt;  [67] glue_1.7.0                ggfun_0.1.5              </span></span>
<span><span class="co">#&gt;  [69] data.table_1.15.4         treeio_1.28.0            </span></span>
<span><span class="co">#&gt;  [71] vctrs_0.6.5               png_0.1-8                </span></span>
<span><span class="co">#&gt;  [73] gtable_0.3.5              cachem_1.1.0             </span></span>
<span><span class="co">#&gt;  [75] xfun_0.45                 S4Arrays_1.4.1           </span></span>
<span><span class="co">#&gt;  [77] mime_0.12                 tidygraph_1.3.1          </span></span>
<span><span class="co">#&gt;  [79] survival_3.6-4            pheatmap_1.0.12          </span></span>
<span><span class="co">#&gt;  [81] iterators_1.0.14          statmod_1.5.0            </span></span>
<span><span class="co">#&gt;  [83] nlme_3.1-164              ggtree_3.12.0            </span></span>
<span><span class="co">#&gt;  [85] bit64_4.0.5               filelock_1.0.3           </span></span>
<span><span class="co">#&gt;  [87] bslib_0.7.0               nor1mix_1.3-3            </span></span>
<span><span class="co">#&gt;  [89] irlba_2.3.5.1             colorspace_2.1-0         </span></span>
<span><span class="co">#&gt;  [91] DBI_1.2.3                 tidyselect_1.2.1         </span></span>
<span><span class="co">#&gt;  [93] bit_4.0.5                 compiler_4.4.1           </span></span>
<span><span class="co">#&gt;  [95] curl_5.2.1                xml2_1.3.6               </span></span>
<span><span class="co">#&gt;  [97] desc_1.4.3                DelayedArray_0.30.1      </span></span>
<span><span class="co">#&gt;  [99] shadowtext_0.1.3          rtracklayer_1.64.0       </span></span>
<span><span class="co">#&gt; [101] scales_1.3.0              quadprog_1.5-8           </span></span>
<span><span class="co">#&gt; [103] rappdirs_0.3.3            stringr_1.5.1            </span></span>
<span><span class="co">#&gt; [105] digest_0.6.36             rmarkdown_2.27           </span></span>
<span><span class="co">#&gt; [107] GEOquery_2.72.0           XVector_0.44.0           </span></span>
<span><span class="co">#&gt; [109] htmltools_0.5.8.1         pkgconfig_2.0.3          </span></span>
<span><span class="co">#&gt; [111] sparseMatrixStats_1.16.0  highr_0.11               </span></span>
<span><span class="co">#&gt; [113] fastmap_1.2.0             rlang_1.1.4              </span></span>
<span><span class="co">#&gt; [115] htmlwidgets_1.6.4         UCSC.utils_1.0.0         </span></span>
<span><span class="co">#&gt; [117] DelayedMatrixStats_1.26.0 farver_2.1.2             </span></span>
<span><span class="co">#&gt; [119] jquerylib_0.1.4           jsonlite_1.8.8           </span></span>
<span><span class="co">#&gt; [121] BiocParallel_1.38.0       mclust_6.1.1             </span></span>
<span><span class="co">#&gt; [123] GOSemSim_2.30.0           BiocSingular_1.20.0      </span></span>
<span><span class="co">#&gt; [125] RCurl_1.98-1.14           magrittr_2.0.3           </span></span>
<span><span class="co">#&gt; [127] ggplotify_0.1.2           GenomeInfoDbData_1.2.12  </span></span>
<span><span class="co">#&gt; [129] Rhdf5lib_1.26.0           munsell_0.5.1            </span></span>
<span><span class="co">#&gt; [131] Rcpp_1.0.12               ggnewscale_0.4.10        </span></span>
<span><span class="co">#&gt; [133] ape_5.8                   viridis_0.6.5            </span></span>
<span><span class="co">#&gt; [135] stringi_1.8.4             ggraph_2.2.1             </span></span>
<span><span class="co">#&gt; [137] zlibbioc_1.50.0           MASS_7.3-60.2            </span></span>
<span><span class="co">#&gt; [139] plyr_1.8.9                bumphunter_1.46.0        </span></span>
<span><span class="co">#&gt; [141] minfi_1.50.0              parallel_4.4.1           </span></span>
<span><span class="co">#&gt; [143] Biostrings_2.72.1         graphlayouts_1.1.1       </span></span>
<span><span class="co">#&gt; [145] splines_4.4.1             multtest_2.60.0          </span></span>
<span><span class="co">#&gt; [147] hms_1.1.3                 locfit_1.5-9.10          </span></span>
<span><span class="co">#&gt; [149] igraph_2.0.3              rngtools_1.5.2           </span></span>
<span><span class="co">#&gt; [151] reshape2_1.4.4            ScaledMatrix_1.12.0      </span></span>
<span><span class="co">#&gt; [153] BiocVersion_3.19.1        evaluate_0.24.0          </span></span>
<span><span class="co">#&gt; [155] BiocManager_1.30.23       tzdb_0.4.0               </span></span>
<span><span class="co">#&gt; [157] foreach_1.5.2             tweenr_2.0.3             </span></span>
<span><span class="co">#&gt; [159] tidyr_1.3.1               openssl_2.2.0            </span></span>
<span><span class="co">#&gt; [161] purrr_1.0.2               polyclip_1.10-6          </span></span>
<span><span class="co">#&gt; [163] reshape_0.8.9             ggforce_0.4.2            </span></span>
<span><span class="co">#&gt; [165] rsvd_1.0.5                xtable_1.8-4             </span></span>
<span><span class="co">#&gt; [167] restfulr_0.0.15           tidytree_0.4.6           </span></span>
<span><span class="co">#&gt; [169] viridisLite_0.4.2         ragg_1.3.2               </span></span>
<span><span class="co">#&gt; [171] tibble_3.2.1              aplot_0.2.3              </span></span>
<span><span class="co">#&gt; [173] memoise_2.0.1             GenomicAlignments_1.40.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gennaro Calendo.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
