<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Perform p-value combination on differential expression results — meta_pcombine • coriell</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Perform p-value combination on differential expression results — meta_pcombine" />
<meta property="og:description" content="This function aims to combine p-values and log fold-change values into a single
combined p-value and logFC value for each gene across all studies. Several methods 
for combining p-values are included. The useer can also pass an arbitrary function
to the lfc_fun argument used to summarize the logFC values across studies." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coriell</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/RNA-seq-workflow.html">RNA-seq-workflow</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform p-value combination on differential expression results</h1>
    
    <div class="hidden name"><code>meta_pcombine.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function aims to combine p-values and log fold-change values into a single
combined p-value and logFC value for each gene across all studies. Several methods 
for combining p-values are included. The useer can also pass an arbitrary function
to the <code>lfc_fun</code> argument used to summarize the logFC values across studies.</p>
    </div>

    <pre class="usage"><span class='fu'>meta_pcombine</span><span class='op'>(</span>
  <span class='va'>exp_list</span>,
  lfc_col <span class='op'>=</span> <span class='st'>"logFC"</span>,
  pval_col <span class='op'>=</span> <span class='st'>"FDR"</span>,
  gene_col <span class='op'>=</span> <span class='st'>"feature_id"</span>,
  all_common <span class='op'>=</span> <span class='cn'>FALSE</span>,
  lfc_fun <span class='op'>=</span> <span class='va'>mean</span>,
  method <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"fisher"</span>, <span class='st'>"pearson"</span>, <span class='st'>"stouffer"</span>, <span class='st'>"tippet"</span>, <span class='st'>"wilkinson"</span><span class='op'>)</span>,
  plot <span class='op'>=</span> <span class='cn'>FALSE</span>,
  plot_lfc <span class='op'>=</span> <span class='fl'>0</span>,
  plot_pval <span class='op'>=</span> <span class='fl'>0.05</span>,
  plot_labels <span class='op'>=</span> <span class='cn'>FALSE</span>,
  plot_count <span class='op'>=</span> <span class='cn'>TRUE</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>exp_list</th>
      <td><p>list() of data.frames of differential expression results.</p></td>
    </tr>
    <tr>
      <th>lfc_col</th>
      <td><p>Column name in the expression data.frames that contains log fold-change values. default = "logFC".</p></td>
    </tr>
    <tr>
      <th>pval_col</th>
      <td><p>Column name in expression data.frames that contains the significance values. default = "FDR".</p></td>
    </tr>
    <tr>
      <th>gene_col</th>
      <td><p>Column name in the expression data.frames that contains the gene IDs. default = "feature_id".</p></td>
    </tr>
    <tr>
      <th>all_common</th>
      <td><p>Logical. Use only genes present in all experiments. default = FALSE.</p></td>
    </tr>
    <tr>
      <th>lfc_fun</th>
      <td><p>Function for summarizing logFC values across studies.</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>Method for combining p-values. One of c("fisher", "pearson", "stouffer", "tippet", "wilkinson"). default "fisher".</p></td>
    </tr>
    <tr>
      <th>plot</th>
      <td><p>Logical. Display a volcano plot using the combined values. default = FALSE</p></td>
    </tr>
    <tr>
      <th>plot_lfc</th>
      <td><p>Numeric. LogFC cutoff used when plotting. default = 0</p></td>
    </tr>
    <tr>
      <th>plot_pval</th>
      <td><p>Numeric. P-value cutoff used when plotting. default = 0.05.</p></td>
    </tr>
    <tr>
      <th>plot_labels</th>
      <td><p>Logical. Display labels for significant genes on plot. default = FALSE</p></td>
    </tr>
    <tr>
      <th>plot_count</th>
      <td><p>Logical. Display count of up/down genes on plot. default = TRUE</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments passed to lfc_fun.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>data.table with columns containing the combined logFC and p-values.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The following p-value combination methods are available:</p><ul>
<li><p>fisher: Uses the sum of the logarithms for the p-values. Sensitive to very small p-values therefore a single significant study can lead to a very small combined p-value.</p></li>
<li><p>pearson: Similar to Fisher's method but sensitive to large p-values; therefore more false negatives are obtained.</p></li>
<li><p>stouffer: (not yet implemented) Recommended when weights for each study can be calculated.</p></li>
<li><p>tippet: Uses the minimum of the p-values across all studies. Recommended when the aim is to discard genes.</p></li>
<li><p>wilkinson: Uses the maximum of the p-values across all studies. Recommended when the aim is to identify the most robust genes.</p></li>
</ul>


  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Gennaro Calendo.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


