<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Perform vote counting on differential expression results — meta_vote • coriell</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Perform vote counting on differential expression results — meta_vote"><meta property="og:description" content="Perform vote-counting on a list of differential expression results. This method
uses user defined p-value and log fold-change cutoffs to determine the number
of times a gene is differentially expressed across multiple studies. A gene is
then considered 'meta-differentially expressed' if it is differentially expressed
in the user defined proportion of studies."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coriell</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.12.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/REdiscoverTE-workflow.html">REdiscoverTE Workflow</a>
    </li>
    <li>
      <a href="../articles/RNA-seq-workflow.html">RNA-seq-workflow</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/coriell-research/coriell/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform vote counting on differential expression results</h1>
    <small class="dont-index">Source: <a href="https://github.com/coriell-research/coriell/blob/HEAD/R/meta-analysis.R" class="external-link"><code>R/meta-analysis.R</code></a></small>
    <div class="hidden name"><code>meta_vote.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Perform vote-counting on a list of differential expression results. This method
uses user defined p-value and log fold-change cutoffs to determine the number
of times a gene is differentially expressed across multiple studies. A gene is
then considered 'meta-differentially expressed' if it is differentially expressed
in the user defined proportion of studies.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">meta_vote</span><span class="op">(</span></span>
<span>  <span class="va">exp_list</span>,</span>
<span>  lfc_col <span class="op">=</span> <span class="st">"logFC"</span>,</span>
<span>  pval_col <span class="op">=</span> <span class="st">"FDR"</span>,</span>
<span>  gene_col <span class="op">=</span> <span class="st">"feature_id"</span>,</span>
<span>  lfc <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  pval <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  meta_prop <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  all_common <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>exp_list</dt>
<dd><p>list() of data.frames of differential expression results.</p></dd>


<dt>lfc_col</dt>
<dd><p>Column name in expression data.frames that contains the log fold-change estimates. default = "logFC".</p></dd>


<dt>pval_col</dt>
<dd><p>Column name in expression data.frames that contains the significance values. default = "FDR".</p></dd>


<dt>gene_col</dt>
<dd><p>Column name in the expression data.frames that contains the gene IDs. default = "feature_id".</p></dd>


<dt>lfc</dt>
<dd><p>Numeric. Log fold-change value used as cutoff for determining differential expression in each study. default = 0.</p></dd>


<dt>pval</dt>
<dd><p>Numeric. Significance value used as cutoff for determining differential expression in each study</p></dd>


<dt>meta_prop</dt>
<dd><p>Numeric. Proportion of studies a gene must be differentially expressed in to be considered as 'meta-differentially expressed'. default = 0.5.</p></dd>


<dt>all_common</dt>
<dd><p>Logical. Use only genes present in all experiments. default = FALSE.</p></dd>


<dt>plot</dt>
<dd><p>Logical. Display a metavolcano ggplot2 plot of the results. default = FALSE.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>data.table of meta-expression results with the following columns:</p><ul><li><p>gene_col: Genes included in the final voting procedure.</p></li>
<li><p>n_de: The total number of studies in which the gene was differentially expressed (either up or down-regulated).</p></li>
<li><p>prop_de: The proportion of studies in which the gene was differentially expressed (either up or down-regulated).</p></li>
<li><p>sign_consistency: The number of studies in which a gene was up-regulated minus the number of studies in which a gene was down-regulated.</p></li>
<li><p>vote: The final vote for the gene. based on <code>sign(sign consistency) &amp; abs(sign consistency) &gt;= meta_prop * N studies</code>.</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>There are certain caveats to consider when using the vote counting method. 1.
combining FDR and logFC cutoffs after the fact destroys FDR correction. See
<a href="https://academic.oup.com/bioinformatics/article/25/6/765/251641" class="external-link">McCarthy &amp; Smyth, 2009</a> for more details.
2. The direction of expression is not considered when the final vote is cast. For example, if
geneA is up-regulated in 4 studies but down-regulated in 1 then the final vote for
geneA will be 'up-regulated'.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Gennaro Calendo.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

