<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Perform GO analysis with PANTHER — panther_go • coriell</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Perform GO analysis with PANTHER — panther_go"><meta property="og:description" content="Sends a request to PANTHER db to perform over 
representation analysis. This function excludes the option to import a reference 
list and reference organism. By default, in this case, PANTHER will use all 
of the genes of the given organism as the reference list."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coriell</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.10.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/REdiscoverTE-workflow.html">REdiscoverTE Workflow</a>
    </li>
    <li>
      <a href="../articles/RNA-seq-workflow.html">RNA-seq-workflow</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/coriell-research/coriell/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform GO analysis with PANTHER</h1>
    <small class="dont-index">Source: <a href="https://github.com/coriell-research/coriell/blob/HEAD/R/panther-go.R" class="external-link"><code>R/panther-go.R</code></a></small>
    <div class="hidden name"><code>panther_go.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Sends a request to <a href="http://pantherdb.org/" class="external-link">PANTHER db</a> to perform over 
representation analysis. This function excludes the option to import a reference 
list and reference organism. By default, in this case, PANTHER will use all 
of the genes of the given organism as the reference list.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">panther_go</span><span class="op">(</span></span>
<span>  <span class="va">gene_list</span>,</span>
<span>  <span class="va">organism</span>,</span>
<span>  <span class="va">annot_dataset</span>,</span>
<span>  enrichment_test_type <span class="op">=</span> <span class="st">"fisher"</span>,</span>
<span>  correction <span class="op">=</span> <span class="st">"fdr"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>gene_list</dt>
<dd><p>character vector. Maximum of 100,000 identifiers. Can be any
of the following: Ensemble gene identifier, Ensemble protein identifier,
Ensemble transcript identifier, Entrez gene id, gene symbol, NCBI GI, HGNC Id,
International protein index id, NCBI UniGene id, UniProt accession and
UniProt id</p></dd>


<dt>organism</dt>
<dd><p>character string. Taxon ID (e.g. "9606" for HUMAN, "10090" for
MOUSE, "10116" for RAT). To get list of available taxon IDs see:</p><div class="sourceCode"><pre><code>curl -X GET "http://pantherdb.org/services/oai/pantherdb/supportedgenomes" -H  "accept: application/json"</code></pre></div></dd>


<dt>annot_dataset</dt>
<dd><p>character string. One of c("biological_process",
"molecular_function", "cellular_component", "panther_go_slim_mf", "panther_go_slim_bp",
"panther_go_slim_cc", "panther_pc", "panther_pathway", "panther_reactome_pathway"). see:</p><div class="sourceCode"><pre><code>curl -X POST "http://pantherdb.org/services/oai/pantherdb/supportedannotdatasets" -H "accept: application/json"</code></pre></div><p>for full descriptions.</p></dd>


<dt>enrichment_test_type</dt>
<dd><p>character string. One of c("fisher", "binomial").
Default "fisher"</p></dd>


<dt>correction</dt>
<dd><p>character string. One of c("fdr", "bonferroni", "none").
Default "fdr"</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>data.table of results from over representation analysis.
See <a href="http://www.pantherdb.org/help/PANTHER_user_manual.pdf" class="external-link">PANTHER user manual</a></p>
 

<p>for column descriptions in "table".</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="st">"CTNNB1"</span>, <span class="st">"ADAM17"</span>, <span class="st">"AXIN1"</span>, <span class="st">"AXIN2"</span>, <span class="st">"CCND2"</span>, <span class="st">"CSNK1E"</span>, <span class="st">"CTNNB1"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"CUL1"</span>, <span class="st">"DKK1"</span>, <span class="st">"DKK4"</span>, <span class="st">"DLL1"</span>, <span class="st">"DVL2"</span>, <span class="st">"FRAT1"</span>, <span class="st">"FZD1"</span>, <span class="st">"FZD8"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"GNAI1"</span>, <span class="st">"HDAC11"</span>, <span class="st">"HDAC2"</span>, <span class="st">"HDAC5"</span>, <span class="st">"HEY1"</span>, <span class="st">"HEY2"</span>, <span class="st">"JAG1"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"JAG2"</span>, <span class="st">"KAT2A"</span>, <span class="st">"LEF1"</span>, <span class="st">"MAML1"</span>, <span class="st">"MYC"</span>, <span class="st">"NCOR2"</span>, <span class="st">"NCSTN"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"NKD1"</span>, <span class="st">"NOTCH1"</span>, <span class="st">"NOTCH4"</span>, <span class="st">"NUMB"</span>, <span class="st">"PPARD"</span>, <span class="st">"PSEN2"</span>, <span class="st">"PTCH1"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"RBPJ"</span>, <span class="st">"SKP2"</span>, <span class="st">"TCF7"</span>, <span class="st">"TP53"</span>, <span class="st">"WNT1"</span>, <span class="st">"WNT5B"</span>, <span class="st">"WNT6"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">go_results</span> <span class="op">&lt;-</span> <span class="fu">panther_go</span><span class="op">(</span><span class="va">genes</span>, organism <span class="op">=</span> <span class="st">"9606"</span>, annot_dataset <span class="op">=</span> <span class="st">"biological_process"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># to view results access the 'table' item</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">go_results</span><span class="op">$</span><span class="va">table</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># to view the actual request (which allows you to parse results yourself) use</span></span></span>
<span class="r-in"><span><span class="va">req</span> <span class="op">&lt;-</span> <span class="va">go_results</span><span class="op">$</span><span class="va">request</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Gennaro Calendo.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

