<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>MD Plot Function — plot_md • coriell</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="MD Plot Function — plot_md" />
<meta property="og:description" content="Create an MD plot from the given dataframe" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coriell</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/RNA-seq-workflow.html">RNA-seq-workflow</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>MD Plot Function</h1>
    
    <div class="hidden name"><code>plot_md.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Create an MD plot from the given dataframe</p>
    </div>

    <pre class="usage"><span class='fu'>plot_md</span><span class='op'>(</span>
  <span class='va'>df</span>,
  x <span class='op'>=</span> <span class='va'>logCPM</span>,
  y <span class='op'>=</span> <span class='va'>logFC</span>,
  sig_col <span class='op'>=</span> <span class='va'>FDR</span>,
  fdr <span class='op'>=</span> <span class='fl'>0.05</span>,
  lfc <span class='op'>=</span> <span class='fl'>0</span>,
  annotate_counts <span class='op'>=</span> <span class='cn'>TRUE</span>,
  xmax_label_offset <span class='op'>=</span> <span class='fl'>0.8</span>,
  ymax_label_offset <span class='op'>=</span> <span class='fl'>0.5</span>,
  ymin_label_offset <span class='op'>=</span> <span class='fl'>0.5</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>df</th>
      <td><p>dataframe containing log-fold-change, p-value, and logCPM columns.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>column in dataframe containing the logCPM data. Default (logCPM)</p></td>
    </tr>
    <tr>
      <th>y</th>
      <td><p>column in dataframe containing the log-fold-change values. Default (logFC)</p></td>
    </tr>
    <tr>
      <th>sig_col</th>
      <td><p>column in dataframe containing the results from significance testing. Default (FDR)</p></td>
    </tr>
    <tr>
      <th>fdr</th>
      <td><p>numeric. Significance level cutoff for plotting. Values below the given fdr threshold are considered significant. Default (0.05)</p></td>
    </tr>
    <tr>
      <th>lfc</th>
      <td><p>numeric. Log-fold-change cutoff for plotting. Values greater than the abs(lfc) and less than fdr are displayed as differentially expressed. Default(0)</p></td>
    </tr>
    <tr>
      <th>annotate_counts</th>
      <td><p>TRUE/FALSE. Annotate the plot with the summarized gene counts</p></td>
    </tr>
    <tr>
      <th>xmax_label_offset</th>
      <td><p>numeric. Value between 0 and 1 inclusive. Controls the x-position of the count labels</p></td>
    </tr>
    <tr>
      <th>ymax_label_offset</th>
      <td><p>numeric. Value between 0 and 1 inclusive. Controls the y-position of the 'up' count label</p></td>
    </tr>
    <tr>
      <th>ymin_label_offset</th>
      <td><p>numeric. Value between 0 and 1 inclusive. Controls the y-position of the 'down' count label</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>ggplot MD plot</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'><a href='http://bioinf.wehi.edu.au/edgeR'>edgeR</a></span><span class='op'>)</span>
<span class='kw'><a href='https://rdrr.io/r/base/library.html'>library</a></span><span class='op'>(</span><span class='va'>coriell</span><span class='op'>)</span>

<span class='co'># create some fake data</span>
<span class='va'>x</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span><span class='op'>(</span>
  ctl1 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/NegBinomial.html'>rnbinom</a></span><span class='op'>(</span><span class='fl'>1000</span>, size <span class='op'>=</span> <span class='fl'>0.4</span>, prob <span class='op'>=</span> <span class='fl'>1e-5</span><span class='op'>)</span>,
  ctl2 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/NegBinomial.html'>rnbinom</a></span><span class='op'>(</span><span class='fl'>1000</span>, size <span class='op'>=</span> <span class='fl'>0.4</span>, prob <span class='op'>=</span> <span class='fl'>1e-5</span><span class='op'>)</span>,
  trt1 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/NegBinomial.html'>rnbinom</a></span><span class='op'>(</span><span class='fl'>1000</span>, size <span class='op'>=</span> <span class='fl'>0.4</span>, prob <span class='op'>=</span> <span class='fl'>1e-5</span><span class='op'>)</span>,
  trt2 <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/NegBinomial.html'>rnbinom</a></span><span class='op'>(</span><span class='fl'>1000</span>, size <span class='op'>=</span> <span class='fl'>0.4</span>, prob <span class='op'>=</span> <span class='fl'>1e-5</span><span class='op'>)</span>,
  row.names <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/paste.html'>paste0</a></span><span class='op'>(</span><span class='st'>"gene"</span>, <span class='fl'>1</span><span class='op'>:</span><span class='fl'>1000</span><span class='op'>)</span>
<span class='op'>)</span>

<span class='co'># run edger pipeline</span>
<span class='va'>group</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/factor.html'>factor</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>, <span class='fl'>1</span>, <span class='fl'>2</span>, <span class='fl'>2</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>y</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/DGEList.html'>DGEList</a></span><span class='op'>(</span>counts <span class='op'>=</span> <span class='va'>x</span>, group <span class='op'>=</span> <span class='va'>group</span><span class='op'>)</span>
<span class='va'>y</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/calcNormFactors.html'>calcNormFactors</a></span><span class='op'>(</span><span class='va'>y</span><span class='op'>)</span>
<span class='va'>design</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/stats/model.matrix.html'>model.matrix</a></span><span class='op'>(</span><span class='op'>~</span><span class='va'>group</span><span class='op'>)</span>
<span class='va'>y</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/estimateDisp.html'>estimateDisp</a></span><span class='op'>(</span><span class='va'>y</span>, <span class='va'>design</span><span class='op'>)</span>

<span class='co'># To perform quasi-likelihood F-tests:</span>
<span class='va'>fit</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/glmQLFTest.html'>glmQLFit</a></span><span class='op'>(</span><span class='va'>y</span>, <span class='va'>design</span><span class='op'>)</span>
<span class='va'>qlf</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/edgeR/man/glmQLFTest.html'>glmQLFTest</a></span><span class='op'>(</span><span class='va'>fit</span>, coef <span class='op'>=</span> <span class='fl'>2</span><span class='op'>)</span>

<span class='co'># convert the results object to a dataframe</span>
<span class='va'>res_df</span> <span class='op'>&lt;-</span> <span class='fu'><a href='edger_to_df.html'>edger_to_df</a></span><span class='op'>(</span><span class='va'>qlf</span><span class='op'>)</span>

<span class='co'># Create md plot</span>
<span class='fu'>plot_md</span><span class='op'>(</span><span class='va'>res_df</span><span class='op'>)</span>
<span class='op'>}</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Gennaro Calendo.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


