<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Heatmap with sensible defaults for RNA-seq expression data — quickmap • coriell</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Heatmap with sensible defaults for RNA-seq expression data — quickmap"><meta property="og:description" content="Generate a heatmap using pheatmap with sensible defaults for RNA-seq.
quickmap() will also attempt to perform vectorized scaling, clustering,
and distance calculations with functions from the Rfast,
fastcluster and rdist packages in order to speed up
calculations for large gene expression matrices."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coriell</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.10.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/REdiscoverTE-workflow.html">REdiscoverTE Workflow</a>
    </li>
    <li>
      <a href="../articles/RNA-seq-workflow.html">RNA-seq-workflow</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/coriell-research/coriell/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Heatmap with sensible defaults for RNA-seq expression data</h1>
    <small class="dont-index">Source: <a href="https://github.com/coriell-research/coriell/blob/HEAD/R/quickmap.R" class="external-link"><code>R/quickmap.R</code></a></small>
    <div class="hidden name"><code>quickmap.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generate a heatmap using <code>pheatmap</code> with sensible defaults for RNA-seq.
<code>quickmap()</code> will also attempt to perform vectorized scaling, clustering,
and distance calculations with functions from the <code>Rfast</code>,
<code>fastcluster</code> and <code>rdist</code> packages in order to speed up
calculations for large gene expression matrices.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">quickmap</span><span class="op">(</span></span>
<span>  <span class="va">mat</span>,</span>
<span>  diverging_palette <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  n_breaks <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  fix_extreme <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  thresh <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  removeVar <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>mat</dt>
<dd><p>numeric matrix to be passed onto pheatmap function</p></dd>


<dt>diverging_palette</dt>
<dd><p>logical. Default(TRUE). Sets the color scale to a diverging palette (blue -&gt; white -&gt; red). If FALSE, set the
color to a continuous color palette <code><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis::magma()</a></code>, useful for un-scaled expression data.</p></dd>


<dt>n_breaks</dt>
<dd><p>numeric. The number of breaks to use in the color palette. Default 50.</p></dd>


<dt>fix_extreme</dt>
<dd><p>logical. Should the extreme values at the ends of a diverging palette be fixed colors? Default FALSE.</p></dd>


<dt>removeVar</dt>
<dd><p>If not NULL remove this proportion of features based on the variance across rows. Default NULL.</p></dd>


<dt>...</dt>
<dd><p>args to be passed to <code>pheatmap()</code> function</p></dd>


<dt>thresh.</dt>
<dd><p>If fix_extreme is TRUE then what proportion of the extreme values should be filled. For example if the range
of the final scaled values is from -20 to 20 and thresh is set to 0.5 then -20 to -10 will be colored blue and 10 to 20
will be colored red. Default 0.5. See examples below.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>pheatmap object. See <code>?pheatmap::pheatmap()</code> for details</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The default arguments to <code><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap::pheatmap()</a></code> are:</p><ul><li><ul><li><p><code>scale = "row"</code></p></li>
<li><p><code>show_rownames = FALSE</code></p></li>
<li><p><code>border_col = NA</code></p></li>
<li><p><code>cluster_rows = TRUE</code></p></li>
<li><p><code>cluster_cols = TRUE</code></p></li>
<li><p><code>color = colorRampPalette(c("dodgerblue3", "grey99", "firebrick3"))(n_breaks)</code> for diverging_palette = TRUE</p></li>
<li><p><code>color = rev(viridisLite::magma(n = n_breaks))</code> for diverging_palette = FALSE</p></li>
<li><p><code>treeheight_row = 0</code></p></li>
<li><p><code>clustering_distance_rows = "correlation"</code></p></li>
<li><p><code>clustering_distance_cols = "euclidean"</code></p></li>
<li><p><code>clustering_method = "ward.D2"</code></p></li>
<li><p><code>angle_col = 315</code></p></li>
</ul><p>You can pass in additional arguments or simply override the defaults as well.</p></li>
</ul></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># display heatmap of scaled count data and add title to the plot</span></span></span>
<span class="r-in"><span><span class="fu">quickmap</span><span class="op">(</span><span class="va">GSE161650_lc</span>, main <span class="op">=</span> <span class="st">"THZ1 vs DMSO"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Remove 90% lowest variance features and fix color scale at ends</span></span></span>
<span class="r-in"><span><span class="fu">quickmap</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">GSE161650_lc</span>, </span></span>
<span class="r-in"><span>  removeVar <span class="op">=</span> <span class="fl">0.9</span>,</span></span>
<span class="r-in"><span>  main <span class="op">=</span> <span class="st">"THZ1 vs DMSO"</span>, </span></span>
<span class="r-in"><span>  fix_extreme <span class="op">=</span> <span class="cn">TRUE</span>, </span></span>
<span class="r-in"><span>  thresh <span class="op">=</span> <span class="fl">0.5</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="quickmap-1.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="quickmap-2.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Gennaro Calendo.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

