<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Heatmap with sensible defaults for RNA-seq expression data — quickmap • coriell</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Heatmap with sensible defaults for RNA-seq expression data — quickmap"><meta name="description" content="Generate a heatmap using pheatmap with sensible defaults for RNA-seq.
quickmap() will also attempt to perform vectorized scaling, clustering,
and distance calculations with functions from the Rfast,
fastcluster and rdist packages in order to speed up
calculations for large gene expression matrices."><meta property="og:description" content="Generate a heatmap using pheatmap with sensible defaults for RNA-seq.
quickmap() will also attempt to perform vectorized scaling, clustering,
and distance calculations with functions from the Rfast,
fastcluster and rdist packages in order to speed up
calculations for large gene expression matrices."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">coriell</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.15.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/Practical-analysis-with-SummarizedExperiments.html">Practical analysis with SummarizedExperiments</a></li>
    <li><a class="dropdown-item" href="../articles/rnaseq-workflow.html">RNAseq Workflow</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/coriell-research/coriell/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Heatmap with sensible defaults for RNA-seq expression data</h1>
      <small class="dont-index">Source: <a href="https://github.com/coriell-research/coriell/blob/HEAD/R/quickmap.R" class="external-link"><code>R/quickmap.R</code></a></small>
      <div class="d-none name"><code>quickmap.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Generate a heatmap using <code>pheatmap</code> with sensible defaults for RNA-seq.
<code>quickmap()</code> will also attempt to perform vectorized scaling, clustering,
and distance calculations with functions from the <code>Rfast</code>,
<code>fastcluster</code> and <code>rdist</code> packages in order to speed up
calculations for large gene expression matrices.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">quickmap</span><span class="op">(</span></span>
<span>  <span class="va">mat</span>,</span>
<span>  diverging_palette <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  fix_extreme <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  thresh <span class="op">=</span> <span class="fl">1.96</span>,</span>
<span>  removeVar <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-mat">mat<a class="anchor" aria-label="anchor" href="#arg-mat"></a></dt>
<dd><p>numeric matrix to be passed onto pheatmap function</p></dd>


<dt id="arg-diverging-palette">diverging_palette<a class="anchor" aria-label="anchor" href="#arg-diverging-palette"></a></dt>
<dd><p>logical. Default(TRUE). Sets the color scale to a
diverging palette (blue -&gt; white -&gt; red). If FALSE, set the color to a
continuous color palette <code><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis::magma()</a></code>, useful for un-scaled
expression data.</p></dd>


<dt id="arg-fix-extreme">fix_extreme<a class="anchor" aria-label="anchor" href="#arg-fix-extreme"></a></dt>
<dd><p>logical. Should the extreme values at the ends of a
diverging palette be fixed colors? Default FALSE.</p></dd>


<dt id="arg-thresh">thresh<a class="anchor" aria-label="anchor" href="#arg-thresh"></a></dt>
<dd><p>If fix_extreme=TRUE then at what value should the color scale
be fixed? Since scaling is implied by the use of this argument, the default
value is set to 1.96 (positive and negative).</p></dd>


<dt id="arg-removevar">removeVar<a class="anchor" aria-label="anchor" href="#arg-removevar"></a></dt>
<dd><p>If not NULL, remove this proportion of features based on
the variance across rows. Default NULL. Note, if removeVar is used the
lowest variance features are removed prior to scaling the input data</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>args to be passed to <code>pheatmap()</code> function</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>pheatmap object. See <code>?pheatmap::pheatmap()</code> for details</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The default arguments to <code><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap::pheatmap()</a></code> are:</p><ul><li><ul><li><p><code>scale = "row"</code></p></li>
<li><p><code>show_rownames = FALSE</code></p></li>
<li><p><code>border_col = NA</code></p></li>
<li><p><code>cluster_rows = TRUE</code></p></li>
<li><p><code>cluster_cols = TRUE</code></p></li>
<li><p><code>color = colorRampPalette(c("dodgerblue3", "grey99", "firebrick3"))(30)</code> for diverging_palette = TRUE</p></li>
<li><p><code>color = rev(viridisLite::magma(n = 30))</code> for diverging_palette = FALSE</p></li>
<li><p><code>treeheight_row = 0</code></p></li>
<li><p><code>clustering_distance_rows = "correlation"</code></p></li>
<li><p><code>clustering_distance_cols = "euclidean"</code></p></li>
<li><p><code>clustering_method = "ward.D2"</code></p></li>
<li><p><code>angle_col = 315</code></p></li>
</ul><p>You can pass in additional arguments or simply override the defaults as well.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># display heatmap of scaled count data and add title to the plot</span></span></span>
<span class="r-in"><span><span class="fu">quickmap</span><span class="op">(</span><span class="va">GSE161650_lc</span>, main <span class="op">=</span> <span class="st">"THZ1 vs DMSO"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="quickmap-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Remove 90% lowest variance features and fix color scale</span></span></span>
<span class="r-in"><span><span class="fu">quickmap</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">GSE161650_lc</span>,</span></span>
<span class="r-in"><span>  removeVar <span class="op">=</span> <span class="fl">0.9</span>,</span></span>
<span class="r-in"><span>  main <span class="op">=</span> <span class="st">"THZ1 vs DMSO"</span>,</span></span>
<span class="r-in"><span>  fix_extreme <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>  thresh <span class="op">=</span> <span class="fl">1</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing 90% lowest variance features...</span>
<span class="r-plt img"><img src="quickmap-2.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gennaro Calendo.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

