<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Generate simulated RNA-seq data for testing purposes — simulate_counts • coriell</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Generate simulated RNA-seq data for testing purposes — simulate_counts" />
<meta property="og:description" content="Simulate an RNASeq count matrix. Adapted from edgeR documentation.
This function is not meant for any serious modelling. It's really used mostly for testing
plotting and summarizing functions internally. It's good for generating test data." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coriell</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/RNA-seq-workflow.html">RNA-seq-workflow</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate simulated RNA-seq data for testing purposes</h1>
    
    <div class="hidden name"><code>simulate_counts.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Simulate an RNASeq count matrix. Adapted from edgeR documentation.
This function is not meant for any serious modelling. It's really used mostly for testing
plotting and summarizing functions internally. It's good for generating test data.</p>
    </div>

    <pre class="usage"><span class='fu'>simulate_counts</span><span class='op'>(</span>
  n_genes <span class='op'>=</span> <span class='fl'>1000</span>,
  n_up <span class='op'>=</span> <span class='fl'>50</span>,
  n_down <span class='op'>=</span> <span class='fl'>50</span>,
  n_samples <span class='op'>=</span> <span class='fl'>6</span>,
  groups <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"ctl"</span>, <span class='st'>"trt"</span><span class='op'>)</span>,
  de_group <span class='op'>=</span> <span class='st'>"trt"</span>,
  mu <span class='op'>=</span> <span class='fl'>10</span>,
  phi <span class='op'>=</span> <span class='fl'>0.1</span>,
  count_offset <span class='op'>=</span> <span class='fl'>10</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>n_genes</th>
      <td><p>integer. The number of gene to simulate</p></td>
    </tr>
    <tr>
      <th>n_up</th>
      <td><p>integer. The number of genes the count offset will be added to</p></td>
    </tr>
    <tr>
      <th>n_down</th>
      <td><p>integer. the number of genes the count offset will be subtracted from. Negative numbers will be coerced to zeros.</p></td>
    </tr>
    <tr>
      <th>n_samples</th>
      <td><p>integer. The number of samples to simulate</p></td>
    </tr>
    <tr>
      <th>groups</th>
      <td><p>character vector. Character vector specifying the names of the groups to simulate</p></td>
    </tr>
    <tr>
      <th>de_group</th>
      <td><p>character. The name of the group to be differentially expressed</p></td>
    </tr>
    <tr>
      <th>mu</th>
      <td><p>numeric. Mean value of counts</p></td>
    </tr>
    <tr>
      <th>phi</th>
      <td><p>numeric. Dispersion parameter</p></td>
    </tr>
    <tr>
      <th>count_offset</th>
      <td><p>numeric. Count offset to be applied to the up/down genes.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>list containing the count matrix and a vectors indicating which rows were modified.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># simulate counts using default parameters</span>
<span class='va'>sim</span> <span class='op'>&lt;-</span> <span class='fu'>simulate_counts</span><span class='op'>(</span><span class='op'>)</span>

<span class='co'># extract the count matrix from the simulation result</span>
<span class='va'>counts</span> <span class='op'>&lt;-</span> <span class='va'>sim</span><span class='op'>$</span><span class='va'>table</span>

<span class='co'># show rows where counts were modified</span>
<span class='va'>counts</span><span class='op'>[</span><span class='va'>sim</span><span class='op'>$</span><span class='va'>de_genes</span>, <span class='op'>]</span>
</div><div class='output co'>#&gt;      ctl.1 ctl.2 ctl.3 trt.1 trt.2 trt.3</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Gennaro Calendo.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


